# Practical session 1
In this practical session you will familiarise yourself with the data and their different aspects.

## Log into the Posit Cloud server

## Import 10X Visium data
In this tutorial we will be using some already prepared data from the [STexampleData](https://bioconductor.org/packages/STexampleData) package. These data are stored in the `SpatialExperiment` format (the format is described below).

The dataset a single sample from the dorsolateral prefrontal cortex (DLPFC) 10x Genomics Visium dataset, that was published by @Maynard2021Mar.

Here, we show how to load the data from the `STexampleData` package.

```{r 01_load-data, message=FALSE}
library(SpatialExperiment)
library(STexampleData)

# Load the object
spe <- Visium_humanDLPFC()
```

## Explore data types
The below discussion will help you familiarise with the different data types that we will be using.

### SpatialExperiment class

For the first part of this tutorial (practical sessions 1 and 2), we will be using the  [SpatialExperiment](https://bioconductor.org/packages/SpatialExperiment) S4 class from Bioconductor as the main data structure for storing and manipulating datasets.

`SpatialExperiment` is a specialized object class that supports the storing of spatially-resolved transcriptomics datasets within the Bioconductor framework. It builds on the [SingleCellExperiment](https://bioconductor.org/packages/SingleCellExperiment) class [@Amezquita2020Feb] for single-cell RNA sequencing data. More specifically, has extra customisations to store spatial information (i.e., spatial coordinates and images).

An overview of the `SpatialExperiment` object structure is is presented in Figure 2.1. In brief, the `SpatialExperiment` object consists of the below five parts:  

 1. `assays`: gene expression counts  
 2. `rowData`: information about features, usually genes  
 3. `colData`: information on spots (non-spatial and spatial metadata)  
 4. `spatialCoords`: spatial coordinates  
 5. `imgData`: image data 

**NOTE:** For spot-based STx data (i.e., 10x Genomics Visium), a single `assay` named `counts` is used.

```{r SpExp-overview, echo=FALSE, out.width = "100%", fig.align="center", fig.cap="Overview of the `SpatialExperiment` object class structure."}
knitr::include_graphics("images/SpatialExperiment.png")
```

For more details, see the related publication from Righelli et al., 2021 describing the `SpatialExperiment` [@Righelli2022Jun].

### Inspect the object
```{r 01_data-inspect, message=FALSE}
## Check the object's structure
spe

## Check number of features/genes (rows) and spots (columns)
dim(spe)

## Check names of 'assay' tables
assayNames(spe)
```

### Counts table and gene metadata
```{r 01_counts-inspect, message=FALSE}
## Have a look at the counts table
assay(spe)[1:6,1:4]
```

As you can see here the counts table is of class `dgTMatrix` which is a sparse matrix. This is because much like scRNA-seq data, STx data are sparse and include many zeros. As a result, to make the counts table as light as possible we resort to using sparse matrices. The next code chunk will demonstrate a part of the matrix that includes genes with some level of expression.

```{r 01_counts-chunks, message=FALSE}
assay(spe)[20:40, 2000:2010]

assay(spe)[33488:33508, 2000:2010]
```

As you can see, the levels of expression of different genes in the same spots differ significantly with many low values being present. We have to remember here that we look at un-normalised expression data which are affected by factors like library size. Nonetheless, this is a fact in STx data as it is for scRNA-seq data; many genes will show low expression in individual spots.

```{r 01_gene-metaData, message=FALSE}
## Have a look at the genes metadata
head(rowData(spe))

```

### Coordinates table and spot metadata
```{r 01_coordinates-inspect, message=FALSE}
## Check the spatial coordinates
head(spatialCoords(spe))

## spot-level metadata
head(colData(spe))

```

### Image metadata
```{r 01_image-inspect, message=FALSE}
## Have a look at the image metadata
imgData(spe)
```


