<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Page not found | Spatial transcriptomics data analysis: theory and practice</title>
  <meta name="description" content="This book will guide you through the practical steps of the in-person tutorial IP2 for the ISMB/ECCB 2023 conference in Lyon named: Spatial transcriptomics data analysis: theory and practice." />
  <meta name="generator" content="bookdown 0.33 and GitBook 2.6.7" />

  <meta property="og:title" content="Page not found | Spatial transcriptomics data analysis: theory and practice" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This book will guide you through the practical steps of the in-person tutorial IP2 for the ISMB/ECCB 2023 conference in Lyon named: Spatial transcriptomics data analysis: theory and practice." />
  <meta name="github-repo" content="https://github.com/ncl-icbam/ismb-tutorial-2023.git" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Page not found | Spatial transcriptomics data analysis: theory and practice" />
  
  <meta name="twitter:description" content="This book will guide you through the practical steps of the in-person tutorial IP2 for the ISMB/ECCB 2023 conference in Lyon named: Spatial transcriptomics data analysis: theory and practice." />
  

<meta name="author" content="Eleftherios Zormpas, Dr Simon J. Cockell" />


<meta name="date" content="2023-04-20" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  


<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>




<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Spatial transcriptomics data analysis: theory and practice</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> url: your book url like https://bookdown.org/yihui/bookdown</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#abstract"><i class="fa fa-check"></i>Abstract</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#learning-objectives"><i class="fa fa-check"></i>Learning objectives</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="practical-session-1.html"><a href="practical-session-1.html"><i class="fa fa-check"></i><b>2</b> Practical session 1</a>
<ul>
<li class="chapter" data-level="2.1" data-path="practical-session-1.html"><a href="practical-session-1.html#log-into-the-posit-cloud-server"><i class="fa fa-check"></i><b>2.1</b> Log into the Posit Cloud server</a></li>
<li class="chapter" data-level="2.2" data-path="practical-session-1.html"><a href="practical-session-1.html#import-10x-visium-data"><i class="fa fa-check"></i><b>2.2</b> Import 10X Visium data</a></li>
<li class="chapter" data-level="2.3" data-path="practical-session-1.html"><a href="practical-session-1.html#explore-data-types"><i class="fa fa-check"></i><b>2.3</b> Explore data types</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="practical-session-1.html"><a href="practical-session-1.html#spatialexperiment-class"><i class="fa fa-check"></i><b>2.3.1</b> SpatialExperiment class</a></li>
<li class="chapter" data-level="2.3.2" data-path="practical-session-1.html"><a href="practical-session-1.html#inspect-the-object"><i class="fa fa-check"></i><b>2.3.2</b> Inspect the object</a></li>
<li class="chapter" data-level="2.3.3" data-path="practical-session-1.html"><a href="practical-session-1.html#counts-table-and-gene-metadata"><i class="fa fa-check"></i><b>2.3.3</b> Counts table and gene metadata</a></li>
<li class="chapter" data-level="2.3.4" data-path="practical-session-1.html"><a href="practical-session-1.html#coordinates-table-and-spot-metadata"><i class="fa fa-check"></i><b>2.3.4</b> Coordinates table and spot metadata</a></li>
<li class="chapter" data-level="2.3.5" data-path="practical-session-1.html"><a href="practical-session-1.html#image-metadata"><i class="fa fa-check"></i><b>2.3.5</b> Image metadata</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="practical-session-2.html"><a href="practical-session-2.html"><i class="fa fa-check"></i><b>3</b> Practical session 2</a>
<ul>
<li class="chapter" data-level="" data-path="practical-session-2.html"><a href="practical-session-2.html#load-packages"><i class="fa fa-check"></i>Load packages</a></li>
<li class="chapter" data-level="3.1" data-path="practical-session-2.html"><a href="practical-session-2.html#spot-level-quality-control"><i class="fa fa-check"></i><b>3.1</b> Spot-level Quality Control</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="practical-session-2.html"><a href="practical-session-2.html#plot-tissue-map"><i class="fa fa-check"></i><b>3.1.1</b> Plot tissue map</a></li>
<li class="chapter" data-level="3.1.2" data-path="practical-session-2.html"><a href="practical-session-2.html#calculating-qc-metrics"><i class="fa fa-check"></i><b>3.1.2</b> Calculating QC metrics</a></li>
<li class="chapter" data-level="3.1.3" data-path="practical-session-2.html"><a href="practical-session-2.html#library-size-threshold-plot"><i class="fa fa-check"></i><b>3.1.3</b> Library size threshold plot</a></li>
<li class="chapter" data-level="3.1.4" data-path="practical-session-2.html"><a href="practical-session-2.html#number-of-expressed-genes"><i class="fa fa-check"></i><b>3.1.4</b> Number of expressed genes</a></li>
<li class="chapter" data-level="3.1.5" data-path="practical-session-2.html"><a href="practical-session-2.html#percentage-of-mitochondrial-expression"><i class="fa fa-check"></i><b>3.1.5</b> Percentage of mitochondrial expression</a></li>
<li class="chapter" data-level="3.1.6" data-path="practical-session-2.html"><a href="practical-session-2.html#number-of-cells-per-spot"><i class="fa fa-check"></i><b>3.1.6</b> Number of cells per spot</a></li>
<li class="chapter" data-level="3.1.7" data-path="practical-session-2.html"><a href="practical-session-2.html#remove-low-quality-spots"><i class="fa fa-check"></i><b>3.1.7</b> Remove low-quality spots</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="practical-session-2.html"><a href="practical-session-2.html#normalisation-of-counts"><i class="fa fa-check"></i><b>3.2</b> Normalisation of counts</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background"><i class="fa fa-check"></i><b>3.2.1</b> Background</a></li>
<li class="chapter" data-level="3.2.2" data-path="practical-session-2.html"><a href="practical-session-2.html#log-tranformation-of-counts"><i class="fa fa-check"></i><b>3.2.2</b> Log-tranformation of counts</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="practical-session-2.html"><a href="practical-session-2.html#selecting-genes"><i class="fa fa-check"></i><b>3.3</b> Selecting genes</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background-1"><i class="fa fa-check"></i><b>3.3.1</b> Background</a></li>
<li class="chapter" data-level="3.3.2" data-path="practical-session-2.html"><a href="practical-session-2.html#highly-variable-genes-hvgs"><i class="fa fa-check"></i><b>3.3.2</b> Highly Variable Genes (HVGs)</a></li>
<li class="chapter" data-level="3.3.3" data-path="practical-session-2.html"><a href="practical-session-2.html#spatially-variable-genes-svgs"><i class="fa fa-check"></i><b>3.3.3</b> Spatially variable genes (SVGs)</a></li>
<li class="chapter" data-level="3.3.4" data-path="practical-session-2.html"><a href="practical-session-2.html#integration-of-hvgs-and-svgs"><i class="fa fa-check"></i><b>3.3.4</b> Integration of HVGs and SVGs</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="practical-session-2.html"><a href="practical-session-2.html#dimensionality-reduction"><i class="fa fa-check"></i><b>3.4</b> Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background-2"><i class="fa fa-check"></i><b>3.4.1</b> Background</a></li>
<li class="chapter" data-level="3.4.2" data-path="practical-session-2.html"><a href="practical-session-2.html#pca-principal-component-analysis"><i class="fa fa-check"></i><b>3.4.2</b> PCA: Principal component analysis</a></li>
<li class="chapter" data-level="3.4.3" data-path="practical-session-2.html"><a href="practical-session-2.html#umap-uniform-manifold-approximation-and-projection"><i class="fa fa-check"></i><b>3.4.3</b> UMAP: Uniform Manifold Approximation and Projection</a></li>
<li class="chapter" data-level="3.4.4" data-path="practical-session-2.html"><a href="practical-session-2.html#umap-visualisations"><i class="fa fa-check"></i><b>3.4.4</b> UMAP visualisations</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="practical-session-2.html"><a href="practical-session-2.html#clustering"><i class="fa fa-check"></i><b>3.5</b> Clustering</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background-3"><i class="fa fa-check"></i><b>3.5.1</b> Background</a></li>
<li class="chapter" data-level="3.5.2" data-path="practical-session-2.html"><a href="practical-session-2.html#clustering-on-hvgs"><i class="fa fa-check"></i><b>3.5.2</b> Clustering on HVGs</a></li>
<li class="chapter" data-level="3.5.3" data-path="practical-session-2.html"><a href="practical-session-2.html#hvgs-clustering-visualisations"><i class="fa fa-check"></i><b>3.5.3</b> HVGs clustering visualisations</a></li>
<li class="chapter" data-level="3.5.4" data-path="practical-session-2.html"><a href="practical-session-2.html#spatially-aware-clustering"><i class="fa fa-check"></i><b>3.5.4</b> Spatially-aware clustering</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="practical-session-2.html"><a href="practical-session-2.html#inter-cluster-differentially-expressed-genes-dges"><i class="fa fa-check"></i><b>3.6</b> Inter-cluster differentially expressed genes (DGEs)</a>
<ul>
<li class="chapter" data-level="3.6.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background-4"><i class="fa fa-check"></i><b>3.6.1</b> Background</a></li>
<li class="chapter" data-level="3.6.2" data-path="practical-session-2.html"><a href="practical-session-2.html#dges-identification"><i class="fa fa-check"></i><b>3.6.2</b> DGEs identification</a></li>
<li class="chapter" data-level="3.6.3" data-path="practical-session-2.html"><a href="practical-session-2.html#dges-visualisation"><i class="fa fa-check"></i><b>3.6.3</b> DGEs visualisation</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="practical-session-2.html"><a href="practical-session-2.html#other-visualisations"><i class="fa fa-check"></i><b>3.7</b> Other visualisations</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="practical-session-3.html"><a href="practical-session-3.html"><i class="fa fa-check"></i><b>4</b> Practical session 3</a>
<ul>
<li class="chapter" data-level="4.1" data-path="practical-session-3.html"><a href="practical-session-3.html#load-packages-1"><i class="fa fa-check"></i><b>4.1</b> Load packages</a></li>
<li class="chapter" data-level="4.2" data-path="practical-session-3.html"><a href="practical-session-3.html#background-5"><i class="fa fa-check"></i><b>4.2</b> Background</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="practical-session-3.html"><a href="practical-session-3.html#main-geocomputatinal-data-structures"><i class="fa fa-check"></i><b>4.2.1</b> Main geocomputatinal data structures</a></li>
<li class="chapter" data-level="4.2.2" data-path="practical-session-3.html"><a href="practical-session-3.html#the-sf-objects"><i class="fa fa-check"></i><b>4.2.2</b> The <code>sf</code> objects</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="practical-session-3.html"><a href="practical-session-3.html#data-structures-preparation"><i class="fa fa-check"></i><b>4.3</b> Data structures preparation</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="practical-session-3.html"><a href="practical-session-3.html#load-new-dataset"><i class="fa fa-check"></i><b>4.3.1</b> Load new dataset</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="practical-session-3.html"><a href="practical-session-3.html#load-counts-table"><i class="fa fa-check"></i><b>4.4</b> Load counts table</a></li>
<li class="chapter" data-level="4.5" data-path="practical-session-3.html"><a href="practical-session-3.html#load-spot-metadata"><i class="fa fa-check"></i><b>4.5</b> Load spot metadata</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="practical-session-3.html"><a href="practical-session-3.html#create-point-geometries"><i class="fa fa-check"></i><b>4.5.1</b> Create point geometries</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="practical-session-3.html"><a href="practical-session-3.html#extract-coordinates"><i class="fa fa-check"></i><b>4.6</b> Extract coordinates</a></li>
<li class="chapter" data-level="4.7" data-path="practical-session-3.html"><a href="practical-session-3.html#convert-spots-to-centroids"><i class="fa fa-check"></i><b>4.7</b> Convert spots to centroids</a>
<ul>
<li class="chapter" data-level="4.7.1" data-path="practical-session-3.html"><a href="practical-session-3.html#tesselate-space"><i class="fa fa-check"></i><b>4.7.1</b> Tesselate space</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="practical-session-3.html"><a href="practical-session-3.html#combine-the-points-into-a-multipoint-geometry"><i class="fa fa-check"></i><b>4.8</b> Combine the points into a multipoint geometry:</a></li>
<li class="chapter" data-level="4.9" data-path="practical-session-3.html"><a href="practical-session-3.html#use-the-union-of-points-to-generate-a-voronoi-object"><i class="fa fa-check"></i><b>4.9</b> Use the union of points to generate a voronoi object</a></li>
<li class="chapter" data-level="4.10" data-path="practical-session-3.html"><a href="practical-session-3.html#create-an-enveloped-voronoi-tessellation-around-the-tissue"><i class="fa fa-check"></i><b>4.10</b> Create an enveloped voronoi tessellation around the tissue</a></li>
<li class="chapter" data-level="4.11" data-path="practical-session-3.html"><a href="practical-session-3.html#plot-tessellation-as-is"><i class="fa fa-check"></i><b>4.11</b> Plot tessellation as is</a></li>
<li class="chapter" data-level="4.12" data-path="practical-session-3.html"><a href="practical-session-3.html#plot-enveloped-tessellation"><i class="fa fa-check"></i><b>4.12</b> Plot enveloped tessellation</a>
<ul>
<li class="chapter" data-level="4.12.1" data-path="practical-session-3.html"><a href="practical-session-3.html#polygonise-the-tessellation"><i class="fa fa-check"></i><b>4.12.1</b> Polygonise the tessellation</a></li>
</ul></li>
<li class="chapter" data-level="4.13" data-path="practical-session-3.html"><a href="practical-session-3.html#generate-the-polygons-from-the-multilinestring"><i class="fa fa-check"></i><b>4.13</b> Generate the POLYGONS from the MULTILINESTRING</a></li>
<li class="chapter" data-level="4.14" data-path="practical-session-3.html"><a href="practical-session-3.html#plot-on-tissue-polygons"><i class="fa fa-check"></i><b>4.14</b> Plot on tissue polygons</a></li>
<li class="chapter" data-level="4.15" data-path="practical-session-3.html"><a href="practical-session-3.html#update-the-polygon-object-to-keep-the-centroid-geometries-as-well"><i class="fa fa-check"></i><b>4.15</b> Update the polygon object to keep the centroid geometries as well</a>
<ul>
<li class="chapter" data-level="4.15.1" data-path="practical-session-3.html"><a href="practical-session-3.html#identify-neighbours"><i class="fa fa-check"></i><b>4.15.1</b> Identify neighbours</a></li>
</ul></li>
<li class="chapter" data-level="4.16" data-path="practical-session-3.html"><a href="practical-session-3.html#get-contiguity-neighbours"><i class="fa fa-check"></i><b>4.16</b> Get contiguity neighbours</a></li>
<li class="chapter" data-level="4.17" data-path="practical-session-3.html"><a href="practical-session-3.html#set-centroids-as-default-geometry"><i class="fa fa-check"></i><b>4.17</b> Set centroids as default geometry</a></li>
<li class="chapter" data-level="4.18" data-path="practical-session-3.html"><a href="practical-session-3.html#get-the-neighbour-names"><i class="fa fa-check"></i><b>4.18</b> Get the neighbour names</a></li>
<li class="chapter" data-level="4.19" data-path="practical-session-3.html"><a href="practical-session-3.html#by-delaunay-triangulation"><i class="fa fa-check"></i><b>4.19</b> By Delaunay triangulation</a></li>
<li class="chapter" data-level="4.20" data-path="practical-session-3.html"><a href="practical-session-3.html#get-neighbours-by-soi"><i class="fa fa-check"></i><b>4.20</b> Get neighbours by SOI</a></li>
<li class="chapter" data-level="4.21" data-path="practical-session-3.html"><a href="practical-session-3.html#get-neighbours-by-gabriel-graph"><i class="fa fa-check"></i><b>4.21</b> Get neighbours by Gabriel graph</a></li>
<li class="chapter" data-level="4.22" data-path="practical-session-3.html"><a href="practical-session-3.html#get-relative-graph-neighbours"><i class="fa fa-check"></i><b>4.22</b> Get Relative graph neighbours</a></li>
<li class="chapter" data-level="4.23" data-path="practical-session-3.html"><a href="practical-session-3.html#set-centroids-as-default-geometry-1"><i class="fa fa-check"></i><b>4.23</b> Set centroids as default geometry</a></li>
<li class="chapter" data-level="4.24" data-path="practical-session-3.html"><a href="practical-session-3.html#get-distance-based-neighbours"><i class="fa fa-check"></i><b>4.24</b> Get distance-based neighbours</a></li>
<li class="chapter" data-level="4.25" data-path="practical-session-3.html"><a href="practical-session-3.html#get-distance-based-neighbours-for-k-1"><i class="fa fa-check"></i><b>4.25</b> Get distance-based neighbours for k = 1</a></li>
<li class="chapter" data-level="4.26" data-path="practical-session-3.html"><a href="practical-session-3.html#set-centroids-as-default-geometry-2"><i class="fa fa-check"></i><b>4.26</b> Set centroids as default geometry</a></li>
<li class="chapter" data-level="4.27" data-path="practical-session-3.html"><a href="practical-session-3.html#add-weights"><i class="fa fa-check"></i><b>4.27</b> Add weights</a></li>
<li class="chapter" data-level="4.28" data-path="practical-session-3.html"><a href="practical-session-3.html#have-a-look"><i class="fa fa-check"></i><b>4.28</b> Have a look</a>
<ul>
<li class="chapter" data-level="4.28.1" data-path="practical-session-3.html"><a href="practical-session-3.html#generate-distance-matrices"><i class="fa fa-check"></i><b>4.28.1</b> Generate distance matrices</a></li>
</ul></li>
<li class="chapter" data-level="4.29" data-path="practical-session-3.html"><a href="practical-session-3.html#set-centroids-as-default-geometry-3"><i class="fa fa-check"></i><b>4.29</b> Set centroids as default geometry</a></li>
<li class="chapter" data-level="4.30" data-path="practical-session-3.html"><a href="practical-session-3.html#get-euclidean-distances-between-spots"><i class="fa fa-check"></i><b>4.30</b> Get Euclidean distances between spots</a></li>
<li class="chapter" data-level="4.31" data-path="practical-session-3.html"><a href="practical-session-3.html#set-bandwidth"><i class="fa fa-check"></i><b>4.31</b> Set bandwidth</a></li>
<li class="chapter" data-level="4.32" data-path="practical-session-3.html"><a href="practical-session-3.html#select-a-kernel"><i class="fa fa-check"></i><b>4.32</b> Select a kernel</a></li>
<li class="chapter" data-level="4.33" data-path="practical-session-3.html"><a href="practical-session-3.html#calculate-w-distance-matrix"><i class="fa fa-check"></i><b>4.33</b> Calculate W distance matrix</a>
<ul>
<li class="chapter" data-level="4.33.1" data-path="practical-session-3.html"><a href="practical-session-3.html#putting-it-all-together"><i class="fa fa-check"></i><b>4.33.1</b> Putting it all together</a></li>
</ul></li>
<li class="chapter" data-level="4.34" data-path="practical-session-3.html"><a href="practical-session-3.html#load-data-and-metadata"><i class="fa fa-check"></i><b>4.34</b> Load data and metadata</a></li>
<li class="chapter" data-level="4.35" data-path="practical-session-3.html"><a href="practical-session-3.html#create-point-geometries-1"><i class="fa fa-check"></i><b>4.35</b> Create point geometries</a></li>
<li class="chapter" data-level="4.36" data-path="practical-session-3.html"><a href="practical-session-3.html#tessellate-space"><i class="fa fa-check"></i><b>4.36</b> Tessellate space</a></li>
<li class="chapter" data-level="4.37" data-path="practical-session-3.html"><a href="practical-session-3.html#polygonise-tessellation"><i class="fa fa-check"></i><b>4.37</b> Polygonise tessellation</a></li>
<li class="chapter" data-level="4.38" data-path="practical-session-3.html"><a href="practical-session-3.html#update-the-polygon-object-to-keep-the-centroid-geometries-as-well-1"><i class="fa fa-check"></i><b>4.38</b> Update the polygon object to keep the centroid geometries as well</a></li>
<li class="chapter" data-level="4.39" data-path="practical-session-3.html"><a href="practical-session-3.html#identify-neighbours-by-sphere-of-influence"><i class="fa fa-check"></i><b>4.39</b> Identify neighbours by Sphere Of Influence</a></li>
<li class="chapter" data-level="4.40" data-path="practical-session-3.html"><a href="practical-session-3.html#generate-distance-matrix"><i class="fa fa-check"></i><b>4.40</b> Generate distance matrix</a></li>
<li class="chapter" data-level="4.41" data-path="practical-session-3.html"><a href="practical-session-3.html#plot-on-tissue-polygons-1"><i class="fa fa-check"></i><b>4.41</b> Plot on tissue polygons</a></li>
<li class="chapter" data-level="4.42" data-path="practical-session-3.html"><a href="practical-session-3.html#plot-neighbour-graph"><i class="fa fa-check"></i><b>4.42</b> Plot neighbour graph</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="practical-session-2-1.html"><a href="practical-session-2-1.html"><i class="fa fa-check"></i><b>5</b> Practical session 2</a>
<ul>
<li class="chapter" data-level="5.1" data-path="practical-session-2-1.html"><a href="practical-session-2-1.html#geographically-weighted-principal-components-analysis-gwpca"><i class="fa fa-check"></i><b>5.1</b> Geographically Weighted Principal Components Analysis (GWPCA)</a></li>
<li class="chapter" data-level="5.2" data-path="practical-session-2-1.html"><a href="practical-session-2-1.html#load-quality-controled-and-normalised-data"><i class="fa fa-check"></i><b>5.2</b> Load Quality Controled and Normalised data</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="blocks.html"><a href="blocks.html"><i class="fa fa-check"></i><b>6</b> Blocks</a>
<ul>
<li class="chapter" data-level="6.1" data-path="blocks.html"><a href="blocks.html#equations"><i class="fa fa-check"></i><b>6.1</b> Equations</a></li>
<li class="chapter" data-level="6.2" data-path="blocks.html"><a href="blocks.html#theorems-and-proofs"><i class="fa fa-check"></i><b>6.2</b> Theorems and proofs</a></li>
<li class="chapter" data-level="6.3" data-path="blocks.html"><a href="blocks.html#callout-blocks"><i class="fa fa-check"></i><b>6.3</b> Callout blocks</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="sharing-your-book.html"><a href="sharing-your-book.html"><i class="fa fa-check"></i><b>7</b> Sharing your book</a>
<ul>
<li class="chapter" data-level="7.1" data-path="sharing-your-book.html"><a href="sharing-your-book.html#publishing"><i class="fa fa-check"></i><b>7.1</b> Publishing</a></li>
<li class="chapter" data-level="7.2" data-path="sharing-your-book.html"><a href="sharing-your-book.html#pages"><i class="fa fa-check"></i><b>7.2</b> 404 pages</a></li>
<li class="chapter" data-level="7.3" data-path="sharing-your-book.html"><a href="sharing-your-book.html#metadata-for-sharing"><i class="fa fa-check"></i><b>7.3</b> Metadata for sharing</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Spatial transcriptomics data analysis: theory and practice</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="page-not-found" class="section level1">
<h1>Page not found</h1>
<p>The page you requested cannot be found (perhaps it was moved or renamed).</p>
<p>You may want to try searching to find the page's new location, or use
the table of contents to find the page you are looking for.</p>
</div>
            </section>

          </div>
        </div>
      </div>


    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/%s",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
