<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Practical session 2 | Spatial transcriptomics data analysis: theory and practice</title>
  <meta name="description" content="This book will guide you through the practical steps of the in-person tutorial IP2 for the ISMB/ECCB 2023 conference in Lyon named: Spatial transcriptomics data analysis: theory and practice." />
  <meta name="generator" content="bookdown 0.33 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Practical session 2 | Spatial transcriptomics data analysis: theory and practice" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This book will guide you through the practical steps of the in-person tutorial IP2 for the ISMB/ECCB 2023 conference in Lyon named: Spatial transcriptomics data analysis: theory and practice." />
  <meta name="github-repo" content="https://github.com/ncl-icbam/ismb-tutorial-2023.git" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Practical session 2 | Spatial transcriptomics data analysis: theory and practice" />
  
  <meta name="twitter:description" content="This book will guide you through the practical steps of the in-person tutorial IP2 for the ISMB/ECCB 2023 conference in Lyon named: Spatial transcriptomics data analysis: theory and practice." />
  

<meta name="author" content="Eleftherios Zormpas, Dr Simon J. Cockell" />


<meta name="date" content="2023-04-10" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="practical-session-1.html"/>
<link rel="next" href="practical-session-2-1.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> url: your book url like https://bookdown.org/yihui/bookdown</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#abstract"><i class="fa fa-check"></i>Abstract</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#learning-objectives"><i class="fa fa-check"></i>Learning objectives</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="practical-session-1.html"><a href="practical-session-1.html"><i class="fa fa-check"></i><b>2</b> Practical session 1</a>
<ul>
<li class="chapter" data-level="2.1" data-path="practical-session-1.html"><a href="practical-session-1.html#log-into-the-posit-cloud-server"><i class="fa fa-check"></i><b>2.1</b> Log into the Posit Cloud server</a></li>
<li class="chapter" data-level="2.2" data-path="practical-session-1.html"><a href="practical-session-1.html#import-10x-visium-data"><i class="fa fa-check"></i><b>2.2</b> Import 10X Visium data</a></li>
<li class="chapter" data-level="2.3" data-path="practical-session-1.html"><a href="practical-session-1.html#explore-data-types"><i class="fa fa-check"></i><b>2.3</b> Explore data types</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="practical-session-1.html"><a href="practical-session-1.html#spatialexperiment-class"><i class="fa fa-check"></i><b>2.3.1</b> SpatialExperiment class</a></li>
<li class="chapter" data-level="2.3.2" data-path="practical-session-1.html"><a href="practical-session-1.html#inspect-the-object"><i class="fa fa-check"></i><b>2.3.2</b> Inspect the object</a></li>
<li class="chapter" data-level="2.3.3" data-path="practical-session-1.html"><a href="practical-session-1.html#counts-table-and-gene-metadata"><i class="fa fa-check"></i><b>2.3.3</b> Counts table and gene metadata</a></li>
<li class="chapter" data-level="2.3.4" data-path="practical-session-1.html"><a href="practical-session-1.html#coordinates-table-and-spot-metadata"><i class="fa fa-check"></i><b>2.3.4</b> Coordinates table and spot metadata</a></li>
<li class="chapter" data-level="2.3.5" data-path="practical-session-1.html"><a href="practical-session-1.html#image-metadata"><i class="fa fa-check"></i><b>2.3.5</b> Image metadata</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="practical-session-2.html"><a href="practical-session-2.html"><i class="fa fa-check"></i><b>3</b> Practical session 2</a>
<ul>
<li class="chapter" data-level="" data-path="practical-session-2.html"><a href="practical-session-2.html#load-packages"><i class="fa fa-check"></i>Load packages</a></li>
<li class="chapter" data-level="3.1" data-path="practical-session-2.html"><a href="practical-session-2.html#spot-level-quality-control"><i class="fa fa-check"></i><b>3.1</b> Spot-level Quality Control</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="practical-session-2.html"><a href="practical-session-2.html#plot-tissue-map"><i class="fa fa-check"></i><b>3.1.1</b> Plot tissue map</a></li>
<li class="chapter" data-level="3.1.2" data-path="practical-session-2.html"><a href="practical-session-2.html#calculating-qc-metrics"><i class="fa fa-check"></i><b>3.1.2</b> Calculating QC metrics</a></li>
<li class="chapter" data-level="3.1.3" data-path="practical-session-2.html"><a href="practical-session-2.html#library-size-threshold-plot"><i class="fa fa-check"></i><b>3.1.3</b> Library size threshold plot</a></li>
<li class="chapter" data-level="3.1.4" data-path="practical-session-2.html"><a href="practical-session-2.html#number-of-expressed-genes"><i class="fa fa-check"></i><b>3.1.4</b> Number of expressed genes</a></li>
<li class="chapter" data-level="3.1.5" data-path="practical-session-2.html"><a href="practical-session-2.html#percentage-of-mitochondrial-expression"><i class="fa fa-check"></i><b>3.1.5</b> Percentage of mitochondrial expression</a></li>
<li class="chapter" data-level="3.1.6" data-path="practical-session-2.html"><a href="practical-session-2.html#number-of-cells-per-spot"><i class="fa fa-check"></i><b>3.1.6</b> Number of cells per spot</a></li>
<li class="chapter" data-level="3.1.7" data-path="practical-session-2.html"><a href="practical-session-2.html#remove-low-quality-spots"><i class="fa fa-check"></i><b>3.1.7</b> Remove low-quality spots</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="practical-session-2.html"><a href="practical-session-2.html#normalisation-of-counts"><i class="fa fa-check"></i><b>3.2</b> Normalisation of counts</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background"><i class="fa fa-check"></i><b>3.2.1</b> Background</a></li>
<li class="chapter" data-level="3.2.2" data-path="practical-session-2.html"><a href="practical-session-2.html#log-tranformation-of-counts"><i class="fa fa-check"></i><b>3.2.2</b> Log-tranformation of counts</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="practical-session-2.html"><a href="practical-session-2.html#selecting-genes"><i class="fa fa-check"></i><b>3.3</b> Selecting genes</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background-1"><i class="fa fa-check"></i><b>3.3.1</b> Background</a></li>
<li class="chapter" data-level="3.3.2" data-path="practical-session-2.html"><a href="practical-session-2.html#highly-variable-genes-hvgs"><i class="fa fa-check"></i><b>3.3.2</b> Highly Variable Genes (HVGs)</a></li>
<li class="chapter" data-level="3.3.3" data-path="practical-session-2.html"><a href="practical-session-2.html#spatially-variable-genes-svgs"><i class="fa fa-check"></i><b>3.3.3</b> Spatially variable genes (SVGs)</a></li>
<li class="chapter" data-level="3.3.4" data-path="practical-session-2.html"><a href="practical-session-2.html#integration-of-hvgs-and-svgs"><i class="fa fa-check"></i><b>3.3.4</b> Integration of HVGs and SVGs</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="practical-session-2.html"><a href="practical-session-2.html#dimensionality-reduction"><i class="fa fa-check"></i><b>3.4</b> Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background-2"><i class="fa fa-check"></i><b>3.4.1</b> Background</a></li>
<li class="chapter" data-level="3.4.2" data-path="practical-session-2.html"><a href="practical-session-2.html#pca-principal-component-analysis"><i class="fa fa-check"></i><b>3.4.2</b> PCA: Principal component analysis</a></li>
<li class="chapter" data-level="3.4.3" data-path="practical-session-2.html"><a href="practical-session-2.html#umap-uniform-manifold-approximation-and-projection"><i class="fa fa-check"></i><b>3.4.3</b> UMAP: Uniform Manifold Approximation and Projection</a></li>
<li class="chapter" data-level="3.4.4" data-path="practical-session-2.html"><a href="practical-session-2.html#umap-visualisations"><i class="fa fa-check"></i><b>3.4.4</b> UMAP visualisations</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="practical-session-2.html"><a href="practical-session-2.html#clustering"><i class="fa fa-check"></i><b>3.5</b> Clustering</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background-3"><i class="fa fa-check"></i><b>3.5.1</b> Background</a></li>
<li class="chapter" data-level="3.5.2" data-path="practical-session-2.html"><a href="practical-session-2.html#clustering-on-hvgs"><i class="fa fa-check"></i><b>3.5.2</b> Clustering on HVGs</a></li>
<li class="chapter" data-level="3.5.3" data-path="practical-session-2.html"><a href="practical-session-2.html#hvgs-clustering-visualisations"><i class="fa fa-check"></i><b>3.5.3</b> HVGs clustering visualisations</a></li>
<li class="chapter" data-level="3.5.4" data-path="practical-session-2.html"><a href="practical-session-2.html#spatially-aware-clustering"><i class="fa fa-check"></i><b>3.5.4</b> Spatially-aware clustering</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="practical-session-2.html"><a href="practical-session-2.html#inter-cluster-differentially-expressed-genes-dges"><i class="fa fa-check"></i><b>3.6</b> Inter-cluster differentially expressed genes (DGEs)</a>
<ul>
<li class="chapter" data-level="3.6.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background-4"><i class="fa fa-check"></i><b>3.6.1</b> Background</a></li>
<li class="chapter" data-level="3.6.2" data-path="practical-session-2.html"><a href="practical-session-2.html#dges-identification"><i class="fa fa-check"></i><b>3.6.2</b> DGEs identification</a></li>
<li class="chapter" data-level="3.6.3" data-path="practical-session-2.html"><a href="practical-session-2.html#dges-visualisation"><i class="fa fa-check"></i><b>3.6.3</b> DGEs visualisation</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="practical-session-2.html"><a href="practical-session-2.html#other-visualisations"><i class="fa fa-check"></i><b>3.7</b> Other visualisations</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="practical-session-2-1.html"><a href="practical-session-2-1.html"><i class="fa fa-check"></i><b>4</b> Practical session 2</a>
<ul>
<li class="chapter" data-level="4.1" data-path="practical-session-2-1.html"><a href="practical-session-2-1.html#section"><i class="fa fa-check"></i><b>4.1</b> </a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="practical-session-2-2.html"><a href="practical-session-2-2.html"><i class="fa fa-check"></i><b>5</b> Practical session 2</a>
<ul>
<li class="chapter" data-level="5.1" data-path="practical-session-2-2.html"><a href="practical-session-2-2.html#section-1"><i class="fa fa-check"></i><b>5.1</b> </a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="blocks.html"><a href="blocks.html"><i class="fa fa-check"></i><b>6</b> Blocks</a>
<ul>
<li class="chapter" data-level="6.1" data-path="blocks.html"><a href="blocks.html#equations"><i class="fa fa-check"></i><b>6.1</b> Equations</a></li>
<li class="chapter" data-level="6.2" data-path="blocks.html"><a href="blocks.html#theorems-and-proofs"><i class="fa fa-check"></i><b>6.2</b> Theorems and proofs</a></li>
<li class="chapter" data-level="6.3" data-path="blocks.html"><a href="blocks.html#callout-blocks"><i class="fa fa-check"></i><b>6.3</b> Callout blocks</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="sharing-your-book.html"><a href="sharing-your-book.html"><i class="fa fa-check"></i><b>7</b> Sharing your book</a>
<ul>
<li class="chapter" data-level="7.1" data-path="sharing-your-book.html"><a href="sharing-your-book.html#publishing"><i class="fa fa-check"></i><b>7.1</b> Publishing</a></li>
<li class="chapter" data-level="7.2" data-path="sharing-your-book.html"><a href="sharing-your-book.html#pages"><i class="fa fa-check"></i><b>7.2</b> 404 pages</a></li>
<li class="chapter" data-level="7.3" data-path="sharing-your-book.html"><a href="sharing-your-book.html#metadata-for-sharing"><i class="fa fa-check"></i><b>7.3</b> Metadata for sharing</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Spatial transcriptomics data analysis: theory and practice</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="practical-session-2" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">Chapter 3</span> Practical session 2<a href="practical-session-2.html#practical-session-2" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>In this session we will demonstrate the implementation of the methods discussed earlier and will particularly focus on the most common analysis routines in STx: QC, data visualisation and clustering analysis always inside the interoperable Bioconductor environment.</p>
<div id="load-packages" class="section level2 unnumbered hasAnchor">
<h2>Load packages<a href="practical-session-2.html#load-packages" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="practical-session-2.html#cb1-1" tabindex="-1"></a><span class="fu">library</span>(ggspavis)</span>
<span id="cb1-2"><a href="practical-session-2.html#cb1-2" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-3"><a href="practical-session-2.html#cb1-3" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb1-4"><a href="practical-session-2.html#cb1-4" tabindex="-1"></a><span class="fu">library</span>(scran)</span>
<span id="cb1-5"><a href="practical-session-2.html#cb1-5" tabindex="-1"></a><span class="fu">library</span>(igraph)</span>
<span id="cb1-6"><a href="practical-session-2.html#cb1-6" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span></code></pre></div>
<ul>
<li><p><a href="https://bioconductor.org/packages/release/bioc/html/ggspavis.html"><code>ggspavis</code></a> is a Bioconductor package that includes visualization functions for spatially resolved transcriptomics datasets stored in <code>SpatialExperiment</code> format from spot-based (e.g., 10x Genomics Visium) platforms (<span class="citation">Weber and Crowell (<a href="#ref-ggspavis2023Apr">2022</a>)</span>).</p></li>
<li><p><a href="https://bioconductor.org/packages/release/bioc/html/scater.html"><code>scater</code></a> is also a Bioconductor package that is a selection of tools for doing various analyses of scRNA-seq gene expression data, with a focus on quality control and visualization which has extended applications to STx data too. It is based on the <code>SingleCellExperiment</code> and <code>SpatialExperiment</code> classes and thus is interoperable with many other Bioconductor packages such as <a href="https://bioconductor.org/packages/3.16/scran"><code>scran</code></a>, <a href="https://bioconductor.org/packages/3.16/scuttle"><code>scuttle</code></a> and <a href="https://bioconductor.org/packages/3.16/iSEE"><code>iSEE</code></a>.</p></li>
</ul>
</div>
<div id="spot-level-quality-control" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> Spot-level Quality Control<a href="practical-session-2.html#spot-level-quality-control" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Spot-level quality control (sQC) procedures are employed to eliminate low-quality spots before conducting further analyses. Low-quality spots may result from issues during library preparation or other experimental procedures, such as a high percentage of dead cells due to cell damage during library preparation, or low mRNA capture efficiency caused by ineffective reverse transcription or PCR amplification. Keeping these spots usually leads to creating problems during downstream analyses.</p>
<p>We can identify low-quality spots using several characteristics that are also used in QC for scRNA-sq data, including:</p>
<ol style="list-style-type: decimal">
<li><strong>library size</strong> (total of UMI counts per spot is going to be different due to sequencing <em>-like different samples in a bulk RNA-seq-</em> or due to number of cells in the spot)</li>
<li><strong>number of expressed genes</strong> (i.e. number of genes with non-zero UMI counts per spot)</li>
<li><strong>proportion of reads mapping to mitochondrial genes</strong> (a high proportion indicates putative cell damage)</li>
</ol>
<p>Low library size or low number of expressed features can indicate poor mRNA capture rates, e.g. due to cell damage and missing mRNAs, or low reaction efficiency. A high proportion of mitochondrial reads indicates cell damage, e.g. partial cell lysis leading to leakage and missing cytoplasmic mRNAs, with the resulting reads therefore concentrated on the remaining mitochondrial mRNAs that are relatively protected inside the mitochondrial membrane. Unusually high numbers of cells per spot can indicate problems during cell segmentation.</p>
<p>The idea of using scRNA-seq QC metrics in STx data comes from the fact that if we remove space and count each spot as a single cell, the two datasets share common features. However, the expected distributions for high-quality spots are different (compared to high-quality cells in scRNA-seq), since spots may contain zero, one, or multiple cells.</p>
<p>A few publications for further reading that can help you understand the quality controls: <span class="citation">McCarthy et al. (<a href="#ref-McCarthy2017Apr">2017</a>)</span> and <span class="citation">Amezquita et al. (<a href="#ref-Amezquita2020Feb">2020</a>)</span>.</p>
<div id="plot-tissue-map" class="section level3 hasAnchor" number="3.1.1">
<h3><span class="header-section-number">3.1.1</span> Plot tissue map<a href="practical-session-2.html#plot-tissue-map" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The DLPFC dataset we will be using comes with manual annotations by the authors <span class="citation">Maynard et al. (<a href="#ref-Maynard2021Mar">2021</a>)</span>. We can plot the tissue map with and without the annotations to get a complete view.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="practical-session-2.html#cb2-1" tabindex="-1"></a><span class="do">## Plot spatial coordinates without annotations</span></span>
<span id="cb2-2"><a href="practical-session-2.html#cb2-2" tabindex="-1"></a><span class="fu">plotSpots</span>(spe)</span>
<span id="cb2-3"><a href="practical-session-2.html#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="practical-session-2.html#cb2-4" tabindex="-1"></a><span class="do">## Plot spatial coordinates with annotations</span></span>
<span id="cb2-5"><a href="practical-session-2.html#cb2-5" tabindex="-1"></a><span class="fu">plotSpots</span>(spe,</span>
<span id="cb2-6"><a href="practical-session-2.html#cb2-6" tabindex="-1"></a>          <span class="at">annotate =</span> <span class="st">&quot;ground_truth&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/02_plot-maps-gTruth-1.png" width="50%" /><img src="_main_files/figure-html/02_plot-maps-gTruth-2.png" width="50%" /></p>
</div>
<div id="calculating-qc-metrics" class="section level3 hasAnchor" number="3.1.2">
<h3><span class="header-section-number">3.1.2</span> Calculating QC metrics<a href="practical-session-2.html#calculating-qc-metrics" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We will now calculate the three main QC metrics described above using methods from the <code>scater</code> <span class="citation">(<a href="#ref-McCarthy2017Apr">McCarthy et al. 2017</a>)</span> package and our own functions <strong>INSERT LINK OR CITATION HERE</strong>.</p>
<p>So far, the dataset contains both on- and off-tissue spots. For the analysis though we are only interested in the on-tissue spots. Therefore, before we run any calculations we want to remove the off-tissue spots.</p>
<p><strong><em>NOTE</em></strong>: the on- or off-tissue information for each spot can be found in the <code>colData</code> of the <code>spe</code> object and in the <code>in_tissue</code> column where <em>0 = off-tissue</em> and <em>1 = on-tissue</em>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="practical-session-2.html#cb3-1" tabindex="-1"></a><span class="do">## Dataset dimensions before the filtering</span></span>
<span id="cb3-2"><a href="practical-session-2.html#cb3-2" tabindex="-1"></a><span class="fu">dim</span>(spe)</span></code></pre></div>
<pre><code>## [1] 33538  4992</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="practical-session-2.html#cb5-1" tabindex="-1"></a><span class="do">## Subset to keep only on-tissue spots</span></span>
<span id="cb5-2"><a href="practical-session-2.html#cb5-2" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[, <span class="fu">colData</span>(spe)<span class="sc">$</span>in_tissue <span class="sc">==</span> <span class="dv">1</span>]</span>
<span id="cb5-3"><a href="practical-session-2.html#cb5-3" tabindex="-1"></a><span class="fu">dim</span>(spe)</span></code></pre></div>
<pre><code>## [1] 33538  3639</code></pre>
<p>The next thing we need to do before we make decisions on how to quality <em>“trim”</em> the dataset is to calculate the percentage per spot of mitochodrial gene expression and store this information inside the <code>colData</code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="practical-session-2.html#cb7-1" tabindex="-1"></a><span class="do">## Fetch mitochondrial gene names</span></span>
<span id="cb7-2"><a href="practical-session-2.html#cb7-2" tabindex="-1"></a>is_mito <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">&quot;(^MT-)|(^mt-)&quot;</span>, <span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_name)</span>
<span id="cb7-3"><a href="practical-session-2.html#cb7-3" tabindex="-1"></a><span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_name[is_mito]</span></code></pre></div>
<pre><code>##  [1] &quot;MT-ND1&quot;  &quot;MT-ND2&quot;  &quot;MT-CO1&quot;  &quot;MT-CO2&quot;  &quot;MT-ATP8&quot; &quot;MT-ATP6&quot; &quot;MT-CO3&quot;  &quot;MT-ND3&quot;  &quot;MT-ND4L&quot; &quot;MT-ND4&quot;  &quot;MT-ND5&quot;  &quot;MT-ND6&quot;  &quot;MT-CYB&quot;</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="practical-session-2.html#cb9-1" tabindex="-1"></a><span class="do">## Calculate per-spot QC metrics and store in colData</span></span>
<span id="cb9-2"><a href="practical-session-2.html#cb9-2" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">addPerCellQC</span>(spe, <span class="at">subsets =</span> <span class="fu">list</span>(<span class="at">mito =</span> is_mito))</span>
<span id="cb9-3"><a href="practical-session-2.html#cb9-3" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(spe))</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 13 columns
##                            barcode_id     sample_id in_tissue array_row array_col ground_truth cell_count       sum  detected
##                           &lt;character&gt;   &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;  &lt;character&gt;  &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt;
## AAACAAGTATCTCCCA-1 AAACAAGTATCTCCCA-1 sample_151673         1        50       102       Layer3          6      8458      3586
## AAACAATCTACTAGCA-1 AAACAATCTACTAGCA-1 sample_151673         1         3        43       Layer1         16      1667      1150
## AAACACCAATAACTGC-1 AAACACCAATAACTGC-1 sample_151673         1        59        19           WM          5      3769      1960
## AAACAGAGCGACTCCT-1 AAACAGAGCGACTCCT-1 sample_151673         1        14        94       Layer3          2      5433      2424
## AAACAGCTTTCAGAAG-1 AAACAGCTTTCAGAAG-1 sample_151673         1        43         9       Layer5          4      4278      2264
## AAACAGGGTCTATATT-1 AAACAGGGTCTATATT-1 sample_151673         1        47        13       Layer6          6      4004      2178
##                    subsets_mito_sum subsets_mito_detected subsets_mito_percent     total
##                           &lt;numeric&gt;             &lt;numeric&gt;            &lt;numeric&gt; &lt;numeric&gt;
## AAACAAGTATCTCCCA-1             1407                    13              16.6351      8458
## AAACAATCTACTAGCA-1              204                    11              12.2376      1667
## AAACACCAATAACTGC-1              430                    13              11.4089      3769
## AAACAGAGCGACTCCT-1             1316                    13              24.2223      5433
## AAACAGCTTTCAGAAG-1              651                    12              15.2174      4278
## AAACAGGGTCTATATT-1              621                    13              15.5095      4004</code></pre>
<p>After calculating the necessary metrics, we need to apply some cut-off thresholds for each metric to perform QC over each spot. What is important to remember here is that each dataset might need slightly different cut-off values to be applied. As a result we cannot rely on identifying a single value to use every time and we need to rely on plotting these metrics and making a decision on a dataset-by-dataset basis.</p>
</div>
<div id="library-size-threshold-plot" class="section level3 hasAnchor" number="3.1.3">
<h3><span class="header-section-number">3.1.3</span> Library size threshold plot<a href="practical-session-2.html#library-size-threshold-plot" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We can plot a histogram of the library sizes across spots. The library size is the number of UMI counts in each spot. We can find this information in the <code>sum</code> column in the <code>colData</code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="practical-session-2.html#cb11-1" tabindex="-1"></a><span class="do">## Density and histogram of library sizes</span></span>
<span id="cb11-2"><a href="practical-session-2.html#cb11-2" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(spe)),</span>
<span id="cb11-3"><a href="practical-session-2.html#cb11-3" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> sum)) <span class="sc">+</span></span>
<span id="cb11-4"><a href="practical-session-2.html#cb11-4" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="fu">after_stat</span>(density)), </span>
<span id="cb11-5"><a href="practical-session-2.html#cb11-5" tabindex="-1"></a>                 <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>, </span>
<span id="cb11-6"><a href="practical-session-2.html#cb11-6" tabindex="-1"></a>                 <span class="at">fill =</span> <span class="st">&quot;grey&quot;</span>) <span class="sc">+</span></span>
<span id="cb11-7"><a href="practical-session-2.html#cb11-7" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>,</span>
<span id="cb11-8"><a href="practical-session-2.html#cb11-8" tabindex="-1"></a>               <span class="at">adjust =</span> <span class="fl">0.5</span>,</span>
<span id="cb11-9"><a href="practical-session-2.html#cb11-9" tabindex="-1"></a>               <span class="at">fill =</span> <span class="st">&quot;#A0CBE8&quot;</span>,</span>
<span id="cb11-10"><a href="practical-session-2.html#cb11-10" tabindex="-1"></a>               <span class="at">colour =</span> <span class="st">&quot;#4E79A7&quot;</span>) <span class="sc">+</span></span>
<span id="cb11-11"><a href="practical-session-2.html#cb11-11" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb11-12"><a href="practical-session-2.html#cb11-12" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb11-13"><a href="practical-session-2.html#cb11-13" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Library size&quot;</span>) <span class="sc">+</span> </span>
<span id="cb11-14"><a href="practical-session-2.html#cb11-14" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>) <span class="sc">+</span> </span>
<span id="cb11-15"><a href="practical-session-2.html#cb11-15" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/02_plot-libSize-histo-1.png" width="672" /></p>
<p>As we can see there are no obvious issues with the library sizes. An example of an issue could be a high frequency of small libraries which would indicate poor experimental output. Generally we do not want to keep spots with too small libraries.</p>
<p>If the dataset we are analysing contains the number of cells that are present in each spot (this one does), then it makes sense to also plot the library sizes against the number of cells per spot. In that way we are making sure that we don’t remove any spots that may have biological meaning. In many cases though the datasets do not have such information unless we can generate it using a nuclei segmentation tool to extract this information from the H&amp;E images.</p>
<p>The horizontal red line (argument <code>threshold</code> in the <code>plotQC</code> function) shows a first guess at a possible filtering threshold for library size based on the above histogram. The below plot can also be plotted using ggplot. For the ggplot implementation have a look at the <a href="XXXX">appendix</a>.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="practical-session-2.html#cb12-1" tabindex="-1"></a><span class="do">## Scatter plot, library size against number of cells per spot</span></span>
<span id="cb12-2"><a href="practical-session-2.html#cb12-2" tabindex="-1"></a><span class="fu">plotQC</span>(spe, <span class="at">type =</span> <span class="st">&quot;scatter&quot;</span>, </span>
<span id="cb12-3"><a href="practical-session-2.html#cb12-3" tabindex="-1"></a>       <span class="at">metric_x =</span> <span class="st">&quot;cell_count&quot;</span>, <span class="at">metric_y =</span> <span class="st">&quot;sum&quot;</span>, </span>
<span id="cb12-4"><a href="practical-session-2.html#cb12-4" tabindex="-1"></a>       <span class="at">threshold_y =</span> <span class="dv">700</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/02_plot-libSizeVScelNo-1.png" width="576" /></p>
<p>We need to keep in mind here that the threshold is, up to an extend, arbitrary. It is crucial, then, to have a look at the number of spots that are left out of the dataset because of this cut-off value and also have a look at their putative spatial patterns. If, by any chance, we filtered out spots with biological relevance, then we will observe some patterns on the tissue map that are correlating with some of the known biological structures of the tissue. As a result, we probably have set our threshold too high.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="practical-session-2.html#cb13-1" tabindex="-1"></a><span class="do">## Select library size threshold</span></span>
<span id="cb13-2"><a href="practical-session-2.html#cb13-2" tabindex="-1"></a>qc_lib_size <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>sum <span class="sc">&lt;</span> <span class="dv">700</span></span>
<span id="cb13-3"><a href="practical-session-2.html#cb13-3" tabindex="-1"></a><span class="do">## Check how many spots are filtered out</span></span>
<span id="cb13-4"><a href="practical-session-2.html#cb13-4" tabindex="-1"></a><span class="fu">table</span>(qc_lib_size)</span></code></pre></div>
<pre><code>## qc_lib_size
## FALSE  TRUE 
##  3628    11</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="practical-session-2.html#cb15-1" tabindex="-1"></a><span class="do">## Apply threshold</span></span>
<span id="cb15-2"><a href="practical-session-2.html#cb15-2" tabindex="-1"></a><span class="fu">colData</span>(spe)<span class="sc">$</span>qc_lib_size <span class="ot">&lt;-</span> qc_lib_size</span>
<span id="cb15-3"><a href="practical-session-2.html#cb15-3" tabindex="-1"></a></span>
<span id="cb15-4"><a href="practical-session-2.html#cb15-4" tabindex="-1"></a><span class="do">## Check putative spatial patterns of removed spots</span></span>
<span id="cb15-5"><a href="practical-session-2.html#cb15-5" tabindex="-1"></a><span class="fu">plotQC</span>(spe, <span class="at">type =</span> <span class="st">&quot;spots&quot;</span>, </span>
<span id="cb15-6"><a href="practical-session-2.html#cb15-6" tabindex="-1"></a>       <span class="at">discard =</span> <span class="st">&quot;qc_lib_size&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/02_libSize-thresh-1.png" width="672" /></p>
<p>As an aside, try to illustrate what happens if we set the threshold too high (i.e., 2000 UMI counts).<br />
<strong>NOTE:</strong> For reference, remember the ground truth layers in this dataset that we <a href="practical-session-2.html#plot-tissue-map">plotted</a> at the beginning.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="practical-session-2.html#cb16-1" tabindex="-1"></a><span class="do">## Select library size threshold</span></span>
<span id="cb16-2"><a href="practical-session-2.html#cb16-2" tabindex="-1"></a>code...</span>
<span id="cb16-3"><a href="practical-session-2.html#cb16-3" tabindex="-1"></a><span class="do">## Check how many spots are filtered out</span></span>
<span id="cb16-4"><a href="practical-session-2.html#cb16-4" tabindex="-1"></a>code...</span>
<span id="cb16-5"><a href="practical-session-2.html#cb16-5" tabindex="-1"></a><span class="do">## Apply threshold</span></span>
<span id="cb16-6"><a href="practical-session-2.html#cb16-6" tabindex="-1"></a>code...</span>
<span id="cb16-7"><a href="practical-session-2.html#cb16-7" tabindex="-1"></a></span>
<span id="cb16-8"><a href="practical-session-2.html#cb16-8" tabindex="-1"></a><span class="do">## Check putative spatial patterns of removed spots</span></span>
<span id="cb16-9"><a href="practical-session-2.html#cb16-9" tabindex="-1"></a><span class="fu">plotQC</span>(...)</span></code></pre></div>
</div>
<div id="number-of-expressed-genes" class="section level3 hasAnchor" number="3.1.4">
<h3><span class="header-section-number">3.1.4</span> Number of expressed genes<a href="practical-session-2.html#number-of-expressed-genes" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>As we did with the library sizes, we can plot a histogram of the number of expressed genes across spots. A gene is expressed in a spot if it has at least one count in it. We can find this information in the <code>detected</code> column in the <code>colData</code>.</p>
<p>We will follow the same logic for the plots as we did for the library size earlier.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="practical-session-2.html#cb17-1" tabindex="-1"></a><span class="do">## Density and histogram of expressed genes</span></span>
<span id="cb17-2"><a href="practical-session-2.html#cb17-2" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(spe)),</span>
<span id="cb17-3"><a href="practical-session-2.html#cb17-3" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> detected)) <span class="sc">+</span></span>
<span id="cb17-4"><a href="practical-session-2.html#cb17-4" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="fu">after_stat</span>(density)), </span>
<span id="cb17-5"><a href="practical-session-2.html#cb17-5" tabindex="-1"></a>                 <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>, </span>
<span id="cb17-6"><a href="practical-session-2.html#cb17-6" tabindex="-1"></a>                 <span class="at">fill =</span> <span class="st">&quot;grey&quot;</span>) <span class="sc">+</span></span>
<span id="cb17-7"><a href="practical-session-2.html#cb17-7" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>,</span>
<span id="cb17-8"><a href="practical-session-2.html#cb17-8" tabindex="-1"></a>               <span class="at">adjust =</span> <span class="fl">0.5</span>,</span>
<span id="cb17-9"><a href="practical-session-2.html#cb17-9" tabindex="-1"></a>               <span class="at">fill =</span> <span class="st">&quot;#A0CBE8&quot;</span>,</span>
<span id="cb17-10"><a href="practical-session-2.html#cb17-10" tabindex="-1"></a>               <span class="at">colour =</span> <span class="st">&quot;#4E79A7&quot;</span>) <span class="sc">+</span></span>
<span id="cb17-11"><a href="practical-session-2.html#cb17-11" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb17-12"><a href="practical-session-2.html#cb17-12" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb17-13"><a href="practical-session-2.html#cb17-13" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Genes expressed in each spot&quot;</span>) <span class="sc">+</span> </span>
<span id="cb17-14"><a href="practical-session-2.html#cb17-14" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>) <span class="sc">+</span> </span>
<span id="cb17-15"><a href="practical-session-2.html#cb17-15" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/02_plot-genesInSpot-histo-1.png" width="672" /></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="practical-session-2.html#cb18-1" tabindex="-1"></a><span class="co"># plot number of expressed genes vs. number of cells per spot</span></span>
<span id="cb18-2"><a href="practical-session-2.html#cb18-2" tabindex="-1"></a><span class="fu">plotQC</span>(spe, <span class="at">type =</span> <span class="st">&quot;scatter&quot;</span>, </span>
<span id="cb18-3"><a href="practical-session-2.html#cb18-3" tabindex="-1"></a>       <span class="at">metric_x =</span> <span class="st">&quot;cell_count&quot;</span>, <span class="at">metric_y =</span> <span class="st">&quot;detected&quot;</span>, </span>
<span id="cb18-4"><a href="practical-session-2.html#cb18-4" tabindex="-1"></a>       <span class="at">threshold_y =</span> <span class="dv">500</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/02_genesInSpot-scatter-1.png" width="576" /></p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="practical-session-2.html#cb19-1" tabindex="-1"></a><span class="do">## Select expressed genes threshold</span></span>
<span id="cb19-2"><a href="practical-session-2.html#cb19-2" tabindex="-1"></a>qc_detected <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>detected <span class="sc">&lt;</span> <span class="dv">500</span></span>
<span id="cb19-3"><a href="practical-session-2.html#cb19-3" tabindex="-1"></a><span class="do">## Check how many spots are filtered out</span></span>
<span id="cb19-4"><a href="practical-session-2.html#cb19-4" tabindex="-1"></a><span class="fu">table</span>(qc_detected)</span></code></pre></div>
<pre><code>## qc_detected
## FALSE  TRUE 
##  3628    11</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="practical-session-2.html#cb21-1" tabindex="-1"></a><span class="do">## Apply threshold</span></span>
<span id="cb21-2"><a href="practical-session-2.html#cb21-2" tabindex="-1"></a><span class="fu">colData</span>(spe)<span class="sc">$</span>qc_detected <span class="ot">&lt;-</span> qc_detected</span>
<span id="cb21-3"><a href="practical-session-2.html#cb21-3" tabindex="-1"></a></span>
<span id="cb21-4"><a href="practical-session-2.html#cb21-4" tabindex="-1"></a><span class="do">## Check for putative spatial pattern of removed spots</span></span>
<span id="cb21-5"><a href="practical-session-2.html#cb21-5" tabindex="-1"></a><span class="fu">plotQC</span>(spe, <span class="at">type =</span> <span class="st">&quot;spots&quot;</span>, </span>
<span id="cb21-6"><a href="practical-session-2.html#cb21-6" tabindex="-1"></a>       <span class="at">discard =</span> <span class="st">&quot;qc_detected&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/02_genesInSpot-thresh-1.png" width="672" /></p>
<p>Again, try to illustrate what happens if we set the threshold too high (i.e., 1000 expressed genes).<br />
<strong>NOTE:</strong> For reference, remember the ground truth layers in this dataset that we <a href="practical-session-2.html#plot-tissue-map">plotted</a> at the beginning.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="practical-session-2.html#cb22-1" tabindex="-1"></a><span class="do">## Select library size threshold</span></span>
<span id="cb22-2"><a href="practical-session-2.html#cb22-2" tabindex="-1"></a>code...</span>
<span id="cb22-3"><a href="practical-session-2.html#cb22-3" tabindex="-1"></a><span class="do">## Check how many spots are filtered out</span></span>
<span id="cb22-4"><a href="practical-session-2.html#cb22-4" tabindex="-1"></a>code...</span>
<span id="cb22-5"><a href="practical-session-2.html#cb22-5" tabindex="-1"></a><span class="do">## Apply threshold</span></span>
<span id="cb22-6"><a href="practical-session-2.html#cb22-6" tabindex="-1"></a>code...</span>
<span id="cb22-7"><a href="practical-session-2.html#cb22-7" tabindex="-1"></a></span>
<span id="cb22-8"><a href="practical-session-2.html#cb22-8" tabindex="-1"></a><span class="do">## Check putative spatial patterns of removed spots</span></span>
<span id="cb22-9"><a href="practical-session-2.html#cb22-9" tabindex="-1"></a><span class="fu">plotQC</span>(...)</span></code></pre></div>
</div>
<div id="percentage-of-mitochondrial-expression" class="section level3 hasAnchor" number="3.1.5">
<h3><span class="header-section-number">3.1.5</span> Percentage of mitochondrial expression<a href="practical-session-2.html#percentage-of-mitochondrial-expression" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>As we briefly touched at the beginning, a high proportion of mitochondrial reads indicates low cell quality, probably due to cell damage.</p>
<p>To investigate the percentage of mitochondrial expression across spots we need to take a look in the column <code>subsets_mito_percent</code> in the <code>colData</code>.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="practical-session-2.html#cb23-1" tabindex="-1"></a><span class="do">## Density and histogram of percentage of mitochondrial expression</span></span>
<span id="cb23-2"><a href="practical-session-2.html#cb23-2" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(spe)),</span>
<span id="cb23-3"><a href="practical-session-2.html#cb23-3" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> subsets_mito_percent)) <span class="sc">+</span></span>
<span id="cb23-4"><a href="practical-session-2.html#cb23-4" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="fu">after_stat</span>(density)), </span>
<span id="cb23-5"><a href="practical-session-2.html#cb23-5" tabindex="-1"></a>                 <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>, </span>
<span id="cb23-6"><a href="practical-session-2.html#cb23-6" tabindex="-1"></a>                 <span class="at">fill =</span> <span class="st">&quot;grey&quot;</span>) <span class="sc">+</span></span>
<span id="cb23-7"><a href="practical-session-2.html#cb23-7" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>,</span>
<span id="cb23-8"><a href="practical-session-2.html#cb23-8" tabindex="-1"></a>               <span class="at">adjust =</span> <span class="fl">0.5</span>,</span>
<span id="cb23-9"><a href="practical-session-2.html#cb23-9" tabindex="-1"></a>               <span class="at">fill =</span> <span class="st">&quot;#A0CBE8&quot;</span>,</span>
<span id="cb23-10"><a href="practical-session-2.html#cb23-10" tabindex="-1"></a>               <span class="at">colour =</span> <span class="st">&quot;#4E79A7&quot;</span>) <span class="sc">+</span></span>
<span id="cb23-11"><a href="practical-session-2.html#cb23-11" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb23-12"><a href="practical-session-2.html#cb23-12" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb23-13"><a href="practical-session-2.html#cb23-13" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Percentage of mitochondrial expression&quot;</span>) <span class="sc">+</span> </span>
<span id="cb23-14"><a href="practical-session-2.html#cb23-14" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>) <span class="sc">+</span> </span>
<span id="cb23-15"><a href="practical-session-2.html#cb23-15" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/02_plot-mitoPercent-histo-1.png" width="672" /></p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="practical-session-2.html#cb24-1" tabindex="-1"></a><span class="co"># plot mitochondrial read proportion vs. number of cells per spot</span></span>
<span id="cb24-2"><a href="practical-session-2.html#cb24-2" tabindex="-1"></a><span class="fu">plotQC</span>(spe, <span class="at">type =</span> <span class="st">&quot;scatter&quot;</span>, </span>
<span id="cb24-3"><a href="practical-session-2.html#cb24-3" tabindex="-1"></a>       <span class="at">metric_x =</span> <span class="st">&quot;cell_count&quot;</span>, <span class="at">metric_y =</span> <span class="st">&quot;subsets_mito_percent&quot;</span>, </span>
<span id="cb24-4"><a href="practical-session-2.html#cb24-4" tabindex="-1"></a>       <span class="at">threshold_y =</span> <span class="dv">28</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/02_mitoPercent-scatter-1.png" width="576" /></p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="practical-session-2.html#cb25-1" tabindex="-1"></a><span class="do">## Select expressed genes threshold</span></span>
<span id="cb25-2"><a href="practical-session-2.html#cb25-2" tabindex="-1"></a>qc_mito <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>subsets_mito_percent <span class="sc">&gt;</span> <span class="dv">28</span></span>
<span id="cb25-3"><a href="practical-session-2.html#cb25-3" tabindex="-1"></a><span class="do">## Check how many spots are filtered out</span></span>
<span id="cb25-4"><a href="practical-session-2.html#cb25-4" tabindex="-1"></a><span class="fu">table</span>(qc_mito)</span></code></pre></div>
<pre><code>## qc_mito
## FALSE  TRUE 
##  3622    17</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="practical-session-2.html#cb27-1" tabindex="-1"></a><span class="do">## Apply threshold</span></span>
<span id="cb27-2"><a href="practical-session-2.html#cb27-2" tabindex="-1"></a><span class="fu">colData</span>(spe)<span class="sc">$</span>qc_mito <span class="ot">&lt;-</span> qc_mito</span>
<span id="cb27-3"><a href="practical-session-2.html#cb27-3" tabindex="-1"></a></span>
<span id="cb27-4"><a href="practical-session-2.html#cb27-4" tabindex="-1"></a><span class="do">## Check for putative spatial pattern of removed spots</span></span>
<span id="cb27-5"><a href="practical-session-2.html#cb27-5" tabindex="-1"></a><span class="fu">plotQC</span>(spe, <span class="at">type =</span> <span class="st">&quot;spots&quot;</span>, </span>
<span id="cb27-6"><a href="practical-session-2.html#cb27-6" tabindex="-1"></a>       <span class="at">discard =</span> <span class="st">&quot;qc_mito&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/02_mitoPercent-thresh-1.png" width="672" /></p>
<p>Again, try to illustrate what happens if we set the threshold too low (i.e., 20 0r 25%).<br />
<strong>NOTE:</strong> For reference, remember the ground truth layers in this dataset that we <a href="practical-session-2.html#plot-tissue-map">plotted</a> at the beginning.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="practical-session-2.html#cb28-1" tabindex="-1"></a><span class="do">## Select library size threshold</span></span>
<span id="cb28-2"><a href="practical-session-2.html#cb28-2" tabindex="-1"></a>code...</span>
<span id="cb28-3"><a href="practical-session-2.html#cb28-3" tabindex="-1"></a><span class="do">## Check how many spots are filtered out</span></span>
<span id="cb28-4"><a href="practical-session-2.html#cb28-4" tabindex="-1"></a>code...</span>
<span id="cb28-5"><a href="practical-session-2.html#cb28-5" tabindex="-1"></a><span class="do">## Apply threshold</span></span>
<span id="cb28-6"><a href="practical-session-2.html#cb28-6" tabindex="-1"></a>code...</span>
<span id="cb28-7"><a href="practical-session-2.html#cb28-7" tabindex="-1"></a></span>
<span id="cb28-8"><a href="practical-session-2.html#cb28-8" tabindex="-1"></a><span class="do">## Check putative spatial patterns of removed spots</span></span>
<span id="cb28-9"><a href="practical-session-2.html#cb28-9" tabindex="-1"></a><span class="fu">plotQC</span>(...)</span></code></pre></div>
</div>
<div id="number-of-cells-per-spot" class="section level3 hasAnchor" number="3.1.6">
<h3><span class="header-section-number">3.1.6</span> Number of cells per spot<a href="practical-session-2.html#number-of-cells-per-spot" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Number of cells per spot is an attribute that not all datasets include. Nonetheless, if it exists, we can use this information to further control the quality of the dataset prior to any downstream analysis. Ofcourse, the number of cells per spot depends on the tissue type and organism and according to <a href="https://kb.10xgenomics.com/hc/en-us/articles/360035487952-How-many-cells-are-captured-in-a-single-spot-">10X Genomics</a>, each spot can contain between 0 and 10 cells.</p>
<p>To investigate the number of cells in each spot looking for any outlier values that could indicate problems we need to take a look in the column <code>cell_count</code> in the <code>colData</code>.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="practical-session-2.html#cb29-1" tabindex="-1"></a><span class="do">## Density and histogram of the number of cells in each spot</span></span>
<span id="cb29-2"><a href="practical-session-2.html#cb29-2" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(spe)),</span>
<span id="cb29-3"><a href="practical-session-2.html#cb29-3" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> cell_count)) <span class="sc">+</span></span>
<span id="cb29-4"><a href="practical-session-2.html#cb29-4" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="fu">after_stat</span>(density)), </span>
<span id="cb29-5"><a href="practical-session-2.html#cb29-5" tabindex="-1"></a>                 <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>, </span>
<span id="cb29-6"><a href="practical-session-2.html#cb29-6" tabindex="-1"></a>                 <span class="at">fill =</span> <span class="st">&quot;grey&quot;</span>) <span class="sc">+</span></span>
<span id="cb29-7"><a href="practical-session-2.html#cb29-7" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>,</span>
<span id="cb29-8"><a href="practical-session-2.html#cb29-8" tabindex="-1"></a>               <span class="co">#adjust = 0.5,</span></span>
<span id="cb29-9"><a href="practical-session-2.html#cb29-9" tabindex="-1"></a>               <span class="at">fill =</span> <span class="st">&quot;#A0CBE8&quot;</span>,</span>
<span id="cb29-10"><a href="practical-session-2.html#cb29-10" tabindex="-1"></a>               <span class="at">colour =</span> <span class="st">&quot;#4E79A7&quot;</span>) <span class="sc">+</span></span>
<span id="cb29-11"><a href="practical-session-2.html#cb29-11" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb29-12"><a href="practical-session-2.html#cb29-12" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb29-13"><a href="practical-session-2.html#cb29-13" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Number of cells per spot&quot;</span>) <span class="sc">+</span> </span>
<span id="cb29-14"><a href="practical-session-2.html#cb29-14" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>) <span class="sc">+</span> </span>
<span id="cb29-15"><a href="practical-session-2.html#cb29-15" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/02_plot-cellsPerSpot-histo-1.png" width="672" /></p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="practical-session-2.html#cb30-1" tabindex="-1"></a><span class="do">## Have a look at the values</span></span>
<span id="cb30-2"><a href="practical-session-2.html#cb30-2" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">colData</span>(spe)<span class="sc">$</span>cell_count)</span></code></pre></div>
<pre><code>## 
##   0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20  21  22  23  25  26  27 
##  84 211 483 623 617 541 421 287 140  92  50  25  18  10   9   3   8   2   1   2   3   2   1   1   2   2   1</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="practical-session-2.html#cb32-1" tabindex="-1"></a><span class="co"># plot number of expressed genes vs. number of cells per spot</span></span>
<span id="cb32-2"><a href="practical-session-2.html#cb32-2" tabindex="-1"></a><span class="fu">plotQC</span>(spe, <span class="at">type =</span> <span class="st">&quot;scatter&quot;</span>, </span>
<span id="cb32-3"><a href="practical-session-2.html#cb32-3" tabindex="-1"></a>       <span class="at">metric_x =</span> <span class="st">&quot;cell_count&quot;</span>, <span class="at">metric_y =</span> <span class="st">&quot;detected&quot;</span>, </span>
<span id="cb32-4"><a href="practical-session-2.html#cb32-4" tabindex="-1"></a>       <span class="at">threshold_x =</span> <span class="dv">10</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/02_cellsPerSpot-scatter-1.png" width="576" /></p>
<p>As we can see from both the histogram and the scatter plot there is a tail of very high values, which could indicate problems for these spots. More specifically, we can see from the scatter plot that most of the spots with very high cell counts also have low numbers of expressed genes. This indicates problems with the experiment on these spots, and they should be removed.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="practical-session-2.html#cb33-1" tabindex="-1"></a><span class="do">## Select expressed genes threshold</span></span>
<span id="cb33-2"><a href="practical-session-2.html#cb33-2" tabindex="-1"></a>qc_cell_count <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>cell_count <span class="sc">&gt;</span> <span class="dv">10</span></span>
<span id="cb33-3"><a href="practical-session-2.html#cb33-3" tabindex="-1"></a><span class="do">## Check how many spots are filtered out</span></span>
<span id="cb33-4"><a href="practical-session-2.html#cb33-4" tabindex="-1"></a><span class="fu">table</span>(qc_cell_count)</span></code></pre></div>
<pre><code>## qc_cell_count
## FALSE  TRUE 
##  3549    90</code></pre>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="practical-session-2.html#cb35-1" tabindex="-1"></a><span class="do">## Apply threshold</span></span>
<span id="cb35-2"><a href="practical-session-2.html#cb35-2" tabindex="-1"></a><span class="fu">colData</span>(spe)<span class="sc">$</span>qc_cell_count <span class="ot">&lt;-</span> qc_cell_count</span>
<span id="cb35-3"><a href="practical-session-2.html#cb35-3" tabindex="-1"></a></span>
<span id="cb35-4"><a href="practical-session-2.html#cb35-4" tabindex="-1"></a><span class="do">## Check for putative spatial pattern of removed spots</span></span>
<span id="cb35-5"><a href="practical-session-2.html#cb35-5" tabindex="-1"></a><span class="fu">plotQC</span>(spe, <span class="at">type =</span> <span class="st">&quot;spots&quot;</span>, </span>
<span id="cb35-6"><a href="practical-session-2.html#cb35-6" tabindex="-1"></a>       <span class="at">discard =</span> <span class="st">&quot;qc_cell_count&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/02_cellsPerSpot-thresh-1.png" width="672" /></p>
<p>While there is a spatial pattern to the discarded spots, it does not appear to be correlated with the known biological features (cortical layers). The discarded spots are all on the edges of the tissue. It seems plausible that something has gone wrong with the cell segmentation on the edges of the images, so it makes sense to remove these spots.</p>
<p>The discarded spots are located at the tissue edges, indicating a potential issue with cell segmentation in those regions. Therefore, it is reasonable to remove these spots from the analysis.</p>
</div>
<div id="remove-low-quality-spots" class="section level3 hasAnchor" number="3.1.7">
<h3><span class="header-section-number">3.1.7</span> Remove low-quality spots<a href="practical-session-2.html#remove-low-quality-spots" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Since we have calculated different spot-level QC metrics and selected thresholds for each one, we can combine them to identify a set of low-quality spots, and remove them from our <code>spe</code> object.</p>
<p>We also check again that the combined set of discarded spots does not correspond to any obvious biologically relevant group of spots.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="practical-session-2.html#cb36-1" tabindex="-1"></a><span class="do">## Check the number of discarded spots for each metric</span></span>
<span id="cb36-2"><a href="practical-session-2.html#cb36-2" tabindex="-1"></a><span class="fu">apply</span>(<span class="fu">cbind</span>(qc_lib_size, qc_detected, qc_mito, qc_cell_count), <span class="dv">2</span>, sum)</span></code></pre></div>
<pre><code>##   qc_lib_size   qc_detected       qc_mito qc_cell_count 
##            11            11            17            90</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="practical-session-2.html#cb38-1" tabindex="-1"></a><span class="do">## Combine together the set of discarded spots</span></span>
<span id="cb38-2"><a href="practical-session-2.html#cb38-2" tabindex="-1"></a>discard <span class="ot">&lt;-</span> qc_lib_size <span class="sc">|</span> qc_detected <span class="sc">|</span> qc_mito <span class="sc">|</span> qc_cell_count</span>
<span id="cb38-3"><a href="practical-session-2.html#cb38-3" tabindex="-1"></a><span class="do">## Store the set in the object</span></span>
<span id="cb38-4"><a href="practical-session-2.html#cb38-4" tabindex="-1"></a><span class="fu">colData</span>(spe)<span class="sc">$</span>discard <span class="ot">&lt;-</span> discard</span>
<span id="cb38-5"><a href="practical-session-2.html#cb38-5" tabindex="-1"></a></span>
<span id="cb38-6"><a href="practical-session-2.html#cb38-6" tabindex="-1"></a><span class="do">## Check the spatial pattern of combined set of discarded spots</span></span>
<span id="cb38-7"><a href="practical-session-2.html#cb38-7" tabindex="-1"></a><span class="fu">plotQC</span>(spe, <span class="at">type =</span> <span class="st">&quot;spots&quot;</span>, </span>
<span id="cb38-8"><a href="practical-session-2.html#cb38-8" tabindex="-1"></a>       <span class="at">discard =</span> <span class="st">&quot;discard&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/02_checkQC-thresh-1.png" width="672" />
Since this dataset has also manual annotation (remember <a href="practical-session-2.html#plot-tissue-map">here</a>) we see that there are locations that are not annotated (marked with <code>NA</code>). We could further remove those locations to reduce noise and further increase the quality of the dataset.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="practical-session-2.html#cb39-1" tabindex="-1"></a><span class="do">## Select locations without annotation</span></span>
<span id="cb39-2"><a href="practical-session-2.html#cb39-2" tabindex="-1"></a>qc_NA_spots <span class="ot">&lt;-</span> <span class="fu">is.na</span>(<span class="fu">colData</span>(spe)<span class="sc">$</span>ground_truth)</span>
<span id="cb39-3"><a href="practical-session-2.html#cb39-3" tabindex="-1"></a><span class="do">## Combine together the set of discarded spots</span></span>
<span id="cb39-4"><a href="practical-session-2.html#cb39-4" tabindex="-1"></a>discard <span class="ot">&lt;-</span> qc_lib_size <span class="sc">|</span> qc_detected <span class="sc">|</span> qc_mito <span class="sc">|</span> qc_cell_count <span class="sc">|</span> qc_NA_spots</span>
<span id="cb39-5"><a href="practical-session-2.html#cb39-5" tabindex="-1"></a><span class="do">## Store the set in the object</span></span>
<span id="cb39-6"><a href="practical-session-2.html#cb39-6" tabindex="-1"></a><span class="fu">colData</span>(spe)<span class="sc">$</span>discard <span class="ot">&lt;-</span> discard</span>
<span id="cb39-7"><a href="practical-session-2.html#cb39-7" tabindex="-1"></a></span>
<span id="cb39-8"><a href="practical-session-2.html#cb39-8" tabindex="-1"></a><span class="do">## Check the spatial pattern of combined set of discarded spots</span></span>
<span id="cb39-9"><a href="practical-session-2.html#cb39-9" tabindex="-1"></a><span class="fu">plotQC</span>(spe, <span class="at">type =</span> <span class="st">&quot;spots&quot;</span>, </span>
<span id="cb39-10"><a href="practical-session-2.html#cb39-10" tabindex="-1"></a>       <span class="at">discard =</span> <span class="st">&quot;discard&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/02_notAnnotSpots-1.png" width="672" /></p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="practical-session-2.html#cb40-1" tabindex="-1"></a><span class="do">## remove combined set of low-quality spots</span></span>
<span id="cb40-2"><a href="practical-session-2.html#cb40-2" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[, <span class="sc">!</span><span class="fu">colData</span>(spe)<span class="sc">$</span>discard]</span></code></pre></div>
</div>
</div>
<div id="normalisation-of-counts" class="section level2 hasAnchor" number="3.2">
<h2><span class="header-section-number">3.2</span> Normalisation of counts<a href="practical-session-2.html#normalisation-of-counts" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="background" class="section level3 hasAnchor" number="3.2.1">
<h3><span class="header-section-number">3.2.1</span> Background<a href="practical-session-2.html#background" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The most common method (if not the only one so far) of normalising gene expression in STx data is the method used in scRNA-seq data. Namely, this is a log-transformation (logcounts). For this to be applied, we treat each spot as being a single cell.</p>
<p>It is clear from what we discussed <a href="practical-session-2.html#number-of-cells-per-spot">earlier</a> that each spot can contain more than one cells. This is a limitation to the technology itself and to our methods of analysing STx data so far. Nonetheless, since STx expression data look like scRNA-seq we apply scRNA-seq methods for normalising (not all methods can be applied though).</p>
<p>Here we will be using methods from the <code>scater</code> <span class="citation">(<a href="#ref-McCarthy2017Apr">McCarthy et al. 2017</a>)</span> and <code>scran</code> <span class="citation">(<a href="#ref-Lun2016Oct">Lun, McCarthy, and Marioni 2016</a>)</span> packages that calculate logcounts using library size factors. The library size factors approach is arguably the simplest approach for STx data. Other approaches used in scRNA-seq are more difficult to justify their use in STx because of two main reasons:</p>
<ol style="list-style-type: decimal">
<li>Spots can contain multiple cells of different cell-types.</li>
<li>Datasets can include multiple tissue samples which will lead to different clusterings.</li>
</ol>
</div>
<div id="log-tranformation-of-counts" class="section level3 hasAnchor" number="3.2.2">
<h3><span class="header-section-number">3.2.2</span> Log-tranformation of counts<a href="practical-session-2.html#log-tranformation-of-counts" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="practical-session-2.html#cb41-1" tabindex="-1"></a><span class="do">## Calculate library size factors</span></span>
<span id="cb41-2"><a href="practical-session-2.html#cb41-2" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">computeLibraryFactors</span>(spe)</span>
<span id="cb41-3"><a href="practical-session-2.html#cb41-3" tabindex="-1"></a><span class="do">## Have a look at the size factors</span></span>
<span id="cb41-4"><a href="practical-session-2.html#cb41-4" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">sizeFactors</span>(spe))</span></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##  0.1514  0.6326  0.9011  1.0000  1.2849  3.7500</code></pre>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="practical-session-2.html#cb43-1" tabindex="-1"></a><span class="do">## Density and histogram of library sizes</span></span>
<span id="cb43-2"><a href="practical-session-2.html#cb43-2" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">data.frame</span>(<span class="at">sFact =</span> <span class="fu">sizeFactors</span>(spe)), </span>
<span id="cb43-3"><a href="practical-session-2.html#cb43-3" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> sFact)) <span class="sc">+</span></span>
<span id="cb43-4"><a href="practical-session-2.html#cb43-4" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="fu">after_stat</span>(density)), </span>
<span id="cb43-5"><a href="practical-session-2.html#cb43-5" tabindex="-1"></a>                 <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>, </span>
<span id="cb43-6"><a href="practical-session-2.html#cb43-6" tabindex="-1"></a>                 <span class="at">fill =</span> <span class="st">&quot;grey&quot;</span>) <span class="sc">+</span></span>
<span id="cb43-7"><a href="practical-session-2.html#cb43-7" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>,</span>
<span id="cb43-8"><a href="practical-session-2.html#cb43-8" tabindex="-1"></a>               <span class="at">adjust =</span> <span class="fl">0.5</span>,</span>
<span id="cb43-9"><a href="practical-session-2.html#cb43-9" tabindex="-1"></a>               <span class="at">fill =</span> <span class="st">&quot;#A0CBE8&quot;</span>,</span>
<span id="cb43-10"><a href="practical-session-2.html#cb43-10" tabindex="-1"></a>               <span class="at">colour =</span> <span class="st">&quot;#4E79A7&quot;</span>) <span class="sc">+</span></span>
<span id="cb43-11"><a href="practical-session-2.html#cb43-11" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb43-12"><a href="practical-session-2.html#cb43-12" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb43-13"><a href="practical-session-2.html#cb43-13" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Library size&quot;</span>) <span class="sc">+</span> </span>
<span id="cb43-14"><a href="practical-session-2.html#cb43-14" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>) <span class="sc">+</span> </span>
<span id="cb43-15"><a href="practical-session-2.html#cb43-15" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/02_plot-labfact-histo-1.png" width="672" /></p>
<p>The log-transformation that takes place is a log2-transformation and in order to avoid <em>- Infinity</em> values we add a pseudo value of 1. Both the log2- transformation and the pseudocount of 1 are defaults in this method.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="practical-session-2.html#cb44-1" tabindex="-1"></a><span class="do">## Calculate logcounts and store in the spe object</span></span>
<span id="cb44-2"><a href="practical-session-2.html#cb44-2" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(spe)</span>
<span id="cb44-3"><a href="practical-session-2.html#cb44-3" tabindex="-1"></a></span>
<span id="cb44-4"><a href="practical-session-2.html#cb44-4" tabindex="-1"></a><span class="do">## Check that a new assay has been added</span></span>
<span id="cb44-5"><a href="practical-session-2.html#cb44-5" tabindex="-1"></a><span class="fu">assayNames</span>(spe)</span></code></pre></div>
<pre><code>## [1] &quot;counts&quot;    &quot;logcounts&quot;</code></pre>
</div>
</div>
<div id="selecting-genes" class="section level2 hasAnchor" number="3.3">
<h2><span class="header-section-number">3.3</span> Selecting genes<a href="practical-session-2.html#selecting-genes" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="background-1" class="section level3 hasAnchor" number="3.3.1">
<h3><span class="header-section-number">3.3.1</span> Background<a href="practical-session-2.html#background-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Gene selection -or alternatively “feature selection”- is applied to identify genes that are going to be informative and can produce valuable information form the downstream analyses. The most common way of applying gene selection is to select genes that are highly variable (HVGs). The assumption is that since we quality-controled and normalised our dataset, the genes with high variability are the ones that contain high levels of biological variability too. Since here we have a spatial dataset we can also try to identify spatially variable genes too (SVGs).</p>
<p>What is important to note here is that HVGs are identified using features like gene expression. Spatial information does not play a role in finding HVGs. STx data pose a dilemma; does the meaningful spatial information reflects only spatial distribution of major cell types or reflects additional important spatial features? If the first is true, then, relying on HVGs can be enough. If the second also holds true, then, it is important to identify SVGs as well.</p>
</div>
<div id="highly-variable-genes-hvgs" class="section level3 hasAnchor" number="3.3.2">
<h3><span class="header-section-number">3.3.2</span> Highly Variable Genes (HVGs)<a href="practical-session-2.html#highly-variable-genes-hvgs" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here we will be using methods from the <code>scran</code> package <span class="citation">(<a href="#ref-Lun2016Oct">Lun, McCarthy, and Marioni 2016</a>)</span> to identify a set of HVGs. Again, here we need to remember that <code>scran</code> methods were developed for scRNA-seq and we are performing the analysis under the assumption that the spots of an STx experiment can be treated as single cells.</p>
<p>In this dataset, the mitochondrial genes are too highly expressed and are not of major biological interest. As a result, if we are to identify true HVGs, we first need to remove the mitochondrial genes.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="practical-session-2.html#cb46-1" tabindex="-1"></a><span class="do">## Remove mitochondrial genes</span></span>
<span id="cb46-2"><a href="practical-session-2.html#cb46-2" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[<span class="sc">!</span>is_mito, ]</span></code></pre></div>
<p>Then, we apply methods from <code>scran</code> that give a list of HVGs, which can be used for further downstream analyses.</p>
<p>First we model the variance of the log-expression profiles for each gene, decomposing it into technical and biological components based on a fitted mean-variance trend.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="practical-session-2.html#cb47-1" tabindex="-1"></a><span class="do">## Fit mean-variance relationship</span></span>
<span id="cb47-2"><a href="practical-session-2.html#cb47-2" tabindex="-1"></a>dec <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(spe)</span>
<span id="cb47-3"><a href="practical-session-2.html#cb47-3" tabindex="-1"></a><span class="do">## Visualize mean-variance relationship</span></span>
<span id="cb47-4"><a href="practical-session-2.html#cb47-4" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">metadata</span>(dec)</span>
<span id="cb47-5"><a href="practical-session-2.html#cb47-5" tabindex="-1"></a>fit_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">mean =</span> fit<span class="sc">$</span>mean,</span>
<span id="cb47-6"><a href="practical-session-2.html#cb47-6" tabindex="-1"></a>                     <span class="at">var =</span> fit<span class="sc">$</span>var,</span>
<span id="cb47-7"><a href="practical-session-2.html#cb47-7" tabindex="-1"></a>                     <span class="at">trend =</span> fit<span class="sc">$</span><span class="fu">trend</span>(fit<span class="sc">$</span>mean))</span>
<span id="cb47-8"><a href="practical-session-2.html#cb47-8" tabindex="-1"></a></span>
<span id="cb47-9"><a href="practical-session-2.html#cb47-9" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> fit_df, </span>
<span id="cb47-10"><a href="practical-session-2.html#cb47-10" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> mean, <span class="at">y =</span> var)) <span class="sc">+</span> </span>
<span id="cb47-11"><a href="practical-session-2.html#cb47-11" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb47-12"><a href="practical-session-2.html#cb47-12" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> trend), <span class="at">colour =</span> <span class="st">&quot;dodgerblue&quot;</span>, <span class="at">linewidth =</span> <span class="fl">1.5</span>) <span class="sc">+</span> </span>
<span id="cb47-13"><a href="practical-session-2.html#cb47-13" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;mean of log-expression&quot;</span>,</span>
<span id="cb47-14"><a href="practical-session-2.html#cb47-14" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;variance of log-expression&quot;</span>) <span class="sc">+</span> </span>
<span id="cb47-15"><a href="practical-session-2.html#cb47-15" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/02_features_FitModel-1.png" width="672" />
The <code>trend</code> function that we used above is returned from the <code>modelGeneVar</code> function and returns the fitted value of the trend at any value of the mean.</p>
<p>We select the top 10% of genes based on their variability The parameter <code>prop</code> defines how many HVGs we want. For example <code>prop = 0.1</code> returns the top 10% of genes.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="practical-session-2.html#cb48-1" tabindex="-1"></a><span class="do">## Select top HVGs</span></span>
<span id="cb48-2"><a href="practical-session-2.html#cb48-2" tabindex="-1"></a>top_hvgs <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(dec, <span class="at">prop =</span> <span class="fl">0.1</span>)</span>
<span id="cb48-3"><a href="practical-session-2.html#cb48-3" tabindex="-1"></a></span>
<span id="cb48-4"><a href="practical-session-2.html#cb48-4" tabindex="-1"></a><span class="do">## How many are the HVGs?</span></span>
<span id="cb48-5"><a href="practical-session-2.html#cb48-5" tabindex="-1"></a><span class="fu">length</span>(top_hvgs)</span></code></pre></div>
<pre><code>## [1] 1429</code></pre>
</div>
<div id="spatially-variable-genes-svgs" class="section level3 hasAnchor" number="3.3.3">
<h3><span class="header-section-number">3.3.3</span> Spatially variable genes (SVGs)<a href="practical-session-2.html#spatially-variable-genes-svgs" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>SVGs are genes with a highly spatially correlated pattern of expression, which varies along with the spatial distribution of a tissue structure of interest. This phenomenon is also called <em>spatial autocorrelation</em> and is a phenomenon that underlies all types of spatial data as we will discuss later.</p>
<p>The field of geography has developed some statistical measures to calculate spatial autocorrelation. Examples of these are Moran’s <em>I</em> <span class="citation">(<a href="#ref-Moran1950Jun"><span>“<span class="nocase">Notes on Continuous Stochastic Phenomena on JSTOR</span>”</span> 1950</a>)</span> and Geary’s <em>C</em> <span class="citation">(<a href="#ref-Geary1954Nov"><span>“<span class="nocase">The Contiguity Ratio and Statistical Mapping on JSTOR</span>”</span> 1954</a>)</span> that can be used to rank genes by the observed spatial autocorrelation to identify SVGs.</p>
<p>Several sophisticated new statistical methods to identify SVGs in SRT data have also recently been developed. These include <a href="https://github.com/Teichlab/SpatialDE">SpatialDE</a> <span class="citation">(<a href="#ref-Svensson2018May">Svensson, Teichmann, and Stegle 2018</a>)</span>, <a href="https://xzhoulab.github.io/SPARK/">SPARK</a> <span class="citation">(<a href="#ref-Sun2020Feb">Sun, Zhu, and Zhou 2020</a>)</span>, and <a href="https://xzhoulab.github.io/SPARK/">SPARK-X</a> <span class="citation">(<a href="#ref-Zhu2021Dec">Zhu, Sun, and Zhou 2021</a>)</span>.</p>
</div>
<div id="integration-of-hvgs-and-svgs" class="section level3 hasAnchor" number="3.3.4">
<h3><span class="header-section-number">3.3.4</span> Integration of HVGs and SVGs<a href="practical-session-2.html#integration-of-hvgs-and-svgs" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>A recent benchmark paper <span class="citation">(<a href="#ref-Li2022Jan">Li et al. 2022</a>)</span> showed that integrating HVGs and SVGs to generate a combined set of features can improve downstream clustering performance in SRT data. This confirms that SVGs contain additional biologically relevant information that is not captured by HVGs in these datasets. For example, a simple way to combine these features is to concatenate columns of principal components (PCs) calculated on the set of HVGs and the set of SVGs (excluding overlapping HVGs), and then using the combined set of features for further downstream analyses <span class="citation">(<a href="#ref-Li2022Jan">Li et al. 2022</a>)</span>.</p>
</div>
</div>
<div id="dimensionality-reduction" class="section level2 hasAnchor" number="3.4">
<h2><span class="header-section-number">3.4</span> Dimensionality reduction<a href="practical-session-2.html#dimensionality-reduction" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="background-2" class="section level3 hasAnchor" number="3.4.1">
<h3><span class="header-section-number">3.4.1</span> Background<a href="practical-session-2.html#background-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Dimensionality reduction is an important step prior to any downstream clustering attempts. There are two main ways of reducing the dimensions of a dataset. The gold standard is Principal Components Analysis (PCA) and the other -a more recent one- Uniform manifold Approximation and Projection (UMAP) <span class="citation">(<a href="#ref-McInnes2018Feb">McInnes, Healy, and Melville 2018</a>)</span>. The main difference between the two is that the distances between the data points in PCA space are interpretable and can be used to cluster the data points while the distances in a UMAP embedding are not interpretable and thus, cannot be used to cluster the data points. As a result, we will be using PCA to reduce the dimensions of our dataset to assist clustering and UMAP to further reduce the principal components (PCs) in a two-dimensional space and produce better visualisations for the PCA.</p>
<p>Reducing the dimensions of the dataset is the output of a PCA. But, what are the main reasons we do so? One reason is to reduce the noise introduced by biologically uninteresting genes that their expression might show some random variation -and that is why they are in the HVGs list-. The other reason is to improve the computational efficiency of downstream analyses like clustering. In an STx experiment, like the one we are analysing here, we have more than 3000 spots and almost 1500 HVGs. As as result, each spot has 1500 attributes based on which the clustering will take place. This increase in the number of variables that differentiate or cluster together spots is leading to the <em>curse of dimensionality</em> <span class="citation">(<a href="#ref-Keogh2017Apr">Keogh and Mueen 2017</a>)</span> which makes the data points (spots) look equidistant in attribute space resulting in poor clustering output.</p>
</div>
<div id="pca-principal-component-analysis" class="section level3 hasAnchor" number="3.4.2">
<h3><span class="header-section-number">3.4.2</span> PCA: Principal component analysis<a href="practical-session-2.html#pca-principal-component-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here we will use an efficient implementation of PCA provided in the <code>scater</code> package <span class="citation">(<a href="#ref-McCarthy2017Apr">McCarthy et al. 2017</a>)</span> and retain the top 50 PCs for further downstream analyses. The random seed is required for reproducibility reasons because this implementation uses randomisation.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="practical-session-2.html#cb50-1" tabindex="-1"></a><span class="do">## Set seed</span></span>
<span id="cb50-2"><a href="practical-session-2.html#cb50-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">987</span>)</span>
<span id="cb50-3"><a href="practical-session-2.html#cb50-3" tabindex="-1"></a><span class="do">## Compute PCA</span></span>
<span id="cb50-4"><a href="practical-session-2.html#cb50-4" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(spe, <span class="at">subset_row =</span> top_hvgs)</span>
<span id="cb50-5"><a href="practical-session-2.html#cb50-5" tabindex="-1"></a><span class="do">## Check correctness - names</span></span>
<span id="cb50-6"><a href="practical-session-2.html#cb50-6" tabindex="-1"></a><span class="fu">reducedDimNames</span>(spe)</span></code></pre></div>
<pre><code>## [1] &quot;PCA&quot;</code></pre>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="practical-session-2.html#cb52-1" tabindex="-1"></a><span class="do">## Check correctness - dimensions</span></span>
<span id="cb52-2"><a href="practical-session-2.html#cb52-2" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">reducedDim</span>(spe, <span class="st">&quot;PCA&quot;</span>))</span></code></pre></div>
<pre><code>## [1] 3511   50</code></pre>
</div>
<div id="umap-uniform-manifold-approximation-and-projection" class="section level3 hasAnchor" number="3.4.3">
<h3><span class="header-section-number">3.4.3</span> UMAP: Uniform Manifold Approximation and Projection<a href="practical-session-2.html#umap-uniform-manifold-approximation-and-projection" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here we will also run UMAP -using <code>scater</code>’s implementation- on the 50 PCs generated above and retain the top 2 UMAP components to visualise results.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="practical-session-2.html#cb54-1" tabindex="-1"></a><span class="do">## Set seed</span></span>
<span id="cb54-2"><a href="practical-session-2.html#cb54-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">987</span>)</span>
<span id="cb54-3"><a href="practical-session-2.html#cb54-3" tabindex="-1"></a><span class="do">## Compute UMAP on top 50 PCs</span></span>
<span id="cb54-4"><a href="practical-session-2.html#cb54-4" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(spe, <span class="at">dimred =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb54-5"><a href="practical-session-2.html#cb54-5" tabindex="-1"></a><span class="do">## Check correctness - names</span></span>
<span id="cb54-6"><a href="practical-session-2.html#cb54-6" tabindex="-1"></a><span class="fu">reducedDimNames</span>(spe)</span></code></pre></div>
<pre><code>## [1] &quot;PCA&quot;  &quot;UMAP&quot;</code></pre>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="practical-session-2.html#cb56-1" tabindex="-1"></a><span class="do">## Check correctness - dimensions</span></span>
<span id="cb56-2"><a href="practical-session-2.html#cb56-2" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">reducedDim</span>(spe, <span class="st">&quot;UMAP&quot;</span>))</span></code></pre></div>
<pre><code>## [1] 3511    2</code></pre>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="practical-session-2.html#cb58-1" tabindex="-1"></a><span class="do">## Update column names for easier plotting</span></span>
<span id="cb58-2"><a href="practical-session-2.html#cb58-2" tabindex="-1"></a><span class="fu">colnames</span>(<span class="fu">reducedDim</span>(spe, <span class="st">&quot;UMAP&quot;</span>)) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;UMAP&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)</span></code></pre></div>
</div>
<div id="umap-visualisations" class="section level3 hasAnchor" number="3.4.4">
<h3><span class="header-section-number">3.4.4</span> UMAP visualisations<a href="practical-session-2.html#umap-visualisations" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We can generate plots either using plotting functions from the <a href="https://bioconductor.org/packages/ggspavis">ggspavis</a> package or <a href="https://ggplot2.tidyverse.org/"><code>ggplot2</code></a> package. Later on clustering, we will add cluster labels to these reduced dimension plots for an off-tissue visualisation.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="practical-session-2.html#cb59-1" tabindex="-1"></a><span class="do">## Plot top 2 PCA dimensions</span></span>
<span id="cb59-2"><a href="practical-session-2.html#cb59-2" tabindex="-1"></a><span class="fu">plotDimRed</span>(spe, <span class="at">type =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb59-3"><a href="practical-session-2.html#cb59-3" tabindex="-1"></a></span>
<span id="cb59-4"><a href="practical-session-2.html#cb59-4" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">as.data.frame</span>(spe<span class="sc">@</span>int_colData<span class="sc">@</span>listData<span class="sc">$</span>reducedDims<span class="sc">$</span>PCA),</span>
<span id="cb59-5"><a href="practical-session-2.html#cb59-5" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2)) <span class="sc">+</span> </span>
<span id="cb59-6"><a href="practical-session-2.html#cb59-6" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span> </span>
<span id="cb59-7"><a href="practical-session-2.html#cb59-7" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Reduced dimensions: PCA&quot;</span>,</span>
<span id="cb59-8"><a href="practical-session-2.html#cb59-8" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">&quot;PC1&quot;</span>,</span>
<span id="cb59-9"><a href="practical-session-2.html#cb59-9" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;PC2&quot;</span>) <span class="sc">+</span></span>
<span id="cb59-10"><a href="practical-session-2.html#cb59-10" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span>
<span id="cb59-11"><a href="practical-session-2.html#cb59-11" tabindex="-1"></a></span>
<span id="cb59-12"><a href="practical-session-2.html#cb59-12" tabindex="-1"></a><span class="do">## Plot top 2 UMAP dimensions</span></span>
<span id="cb59-13"><a href="practical-session-2.html#cb59-13" tabindex="-1"></a><span class="fu">plotDimRed</span>(spe, <span class="at">type =</span> <span class="st">&quot;UMAP&quot;</span>)</span>
<span id="cb59-14"><a href="practical-session-2.html#cb59-14" tabindex="-1"></a></span>
<span id="cb59-15"><a href="practical-session-2.html#cb59-15" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">as.data.frame</span>(spe<span class="sc">@</span>int_colData<span class="sc">@</span>listData<span class="sc">$</span>reducedDims<span class="sc">$</span>UMAP),</span>
<span id="cb59-16"><a href="practical-session-2.html#cb59-16" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> UMAP1, <span class="at">y =</span> UMAP2)) <span class="sc">+</span> </span>
<span id="cb59-17"><a href="practical-session-2.html#cb59-17" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span> </span>
<span id="cb59-18"><a href="practical-session-2.html#cb59-18" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Reduced dimensions: UMAP&quot;</span>,</span>
<span id="cb59-19"><a href="practical-session-2.html#cb59-19" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">&quot;UMAP1&quot;</span>,</span>
<span id="cb59-20"><a href="practical-session-2.html#cb59-20" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;UMAP2&quot;</span>) <span class="sc">+</span></span>
<span id="cb59-21"><a href="practical-session-2.html#cb59-21" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/02_dimRed_UMAP-vis-1.png" width="480" /><img src="_main_files/figure-html/02_dimRed_UMAP-vis-2.png" width="480" /><img src="_main_files/figure-html/02_dimRed_UMAP-vis-3.png" width="480" /><img src="_main_files/figure-html/02_dimRed_UMAP-vis-4.png" width="480" /></p>
</div>
</div>
<div id="clustering" class="section level2 hasAnchor" number="3.5">
<h2><span class="header-section-number">3.5</span> Clustering<a href="practical-session-2.html#clustering" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="background-3" class="section level3 hasAnchor" number="3.5.1">
<h3><span class="header-section-number">3.5.1</span> Background<a href="practical-session-2.html#background-3" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The clustering of observations into statistically similar groups is a well-established application in both bulk and single-cell RNA-Seq analysis. Clustering is a helpful tool because it structures and orders the data, allowing useful insights to be gained from complex, multivariate datasets and use those insights to classify the observed data or to generate hypotheses.</p>
<p>Common clustering methods are applied to ST data based on correlation or statistical distance of gene expression measurements. As we briefly touched above, the dimensionality of ST data means that sample distances in gene expression space tend to be small and not reliable for identifying clusters, so feature selection (gene selection) or dimensionality reduction approaches (i.e., PCA, UMAP) tend to be taken before clustering.</p>
<p>Common approaches to clustering gene expression data include k-means, hierarchical and Louvain algorithms, and all have been applied to the clustering of ST data. Some of these methods are implemented in some of the most popular single-cell analysis packages, such as <code>Seurat</code> <span class="citation">(<a href="#ref-Hao2021Jun">Hao et al. 2021</a>)</span> and <code>scran</code> <span class="citation">(<a href="#ref-Lun2016Oct">Lun, McCarthy, and Marioni 2016</a>)</span> and have been used for clustering in a number of ST studies.</p>
</div>
<div id="clustering-on-hvgs" class="section level3 hasAnchor" number="3.5.2">
<h3><span class="header-section-number">3.5.2</span> Clustering on HVGs<a href="practical-session-2.html#clustering-on-hvgs" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here, we apply graph-based clustering to the top 50 PCs calculated on the set of selected HVGs, using the Walktrap method implemented in <code>scran</code> <span class="citation">(<a href="#ref-Lun2016Oct">Lun, McCarthy, and Marioni 2016</a>)</span>. To do so, we assume that (i) each spot is equal to a cell and (ii) we can detect from the gene expression the biologically informative spatial distribution patterns of cell types.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="practical-session-2.html#cb60-1" tabindex="-1"></a><span class="do">## Set seed</span></span>
<span id="cb60-2"><a href="practical-session-2.html#cb60-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">987</span>)</span>
<span id="cb60-3"><a href="practical-session-2.html#cb60-3" tabindex="-1"></a><span class="do">## Set number of Nearest-Neighbours (NNs)</span></span>
<span id="cb60-4"><a href="practical-session-2.html#cb60-4" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb60-5"><a href="practical-session-2.html#cb60-5" tabindex="-1"></a><span class="do">## Build the k-NN graph</span></span>
<span id="cb60-6"><a href="practical-session-2.html#cb60-6" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">buildSNNGraph</span>(spe, <span class="at">k =</span> k, <span class="at">use.dimred =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb60-7"><a href="practical-session-2.html#cb60-7" tabindex="-1"></a><span class="do">## Run walktrap clustering</span></span>
<span id="cb60-8"><a href="practical-session-2.html#cb60-8" tabindex="-1"></a>g_walk <span class="ot">&lt;-</span> igraph<span class="sc">::</span><span class="fu">cluster_walktrap</span>(g)</span>
<span id="cb60-9"><a href="practical-session-2.html#cb60-9" tabindex="-1"></a><span class="do">## Get the cluster labels</span></span>
<span id="cb60-10"><a href="practical-session-2.html#cb60-10" tabindex="-1"></a>clus <span class="ot">&lt;-</span> g_walk<span class="sc">$</span>membership</span>
<span id="cb60-11"><a href="practical-session-2.html#cb60-11" tabindex="-1"></a><span class="do">## Check how many</span></span>
<span id="cb60-12"><a href="practical-session-2.html#cb60-12" tabindex="-1"></a><span class="fu">table</span>(clus)</span></code></pre></div>
<pre><code>## clus
##   1   2   3   4   5   6 
## 350 354 661 895 366 885</code></pre>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="practical-session-2.html#cb62-1" tabindex="-1"></a><span class="do">## Store cluster labels in column &#39;label&#39; in colData</span></span>
<span id="cb62-2"><a href="practical-session-2.html#cb62-2" tabindex="-1"></a><span class="fu">colLabels</span>(spe) <span class="ot">&lt;-</span> <span class="fu">factor</span>(clus)</span></code></pre></div>
</div>
<div id="hvgs-clustering-visualisations" class="section level3 hasAnchor" number="3.5.3">
<h3><span class="header-section-number">3.5.3</span> HVGs clustering visualisations<a href="practical-session-2.html#hvgs-clustering-visualisations" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We can visualise the clusters in two ways:<br />
1. plotting in spatial coordinates on the tissue map
2. plotting in the UMAP/PCA embeddings.</p>
<p>We can use plotting functions either from the <a href="https://bioconductor.org/packages/ggspavis">ggspavis</a>package.</p>
<p>For reference, we will also display the ground truth (manually annotated) labels available for this dataset.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="practical-session-2.html#cb63-1" tabindex="-1"></a><span class="do">## Plot in tissue map</span></span>
<span id="cb63-2"><a href="practical-session-2.html#cb63-2" tabindex="-1"></a><span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">&quot;label&quot;</span>, </span>
<span id="cb63-3"><a href="practical-session-2.html#cb63-3" tabindex="-1"></a>          <span class="at">palette =</span> <span class="st">&quot;libd_layer_colors&quot;</span>)</span>
<span id="cb63-4"><a href="practical-session-2.html#cb63-4" tabindex="-1"></a></span>
<span id="cb63-5"><a href="practical-session-2.html#cb63-5" tabindex="-1"></a><span class="do">## Plot ground truth in tissue map</span></span>
<span id="cb63-6"><a href="practical-session-2.html#cb63-6" tabindex="-1"></a><span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">&quot;ground_truth&quot;</span>, </span>
<span id="cb63-7"><a href="practical-session-2.html#cb63-7" tabindex="-1"></a>          <span class="at">palette =</span> <span class="st">&quot;libd_layer_colors&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/02_clust_vis-map-1.png" width="50%" /><img src="_main_files/figure-html/02_clust_vis-map-2.png" width="50%" /></p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="practical-session-2.html#cb64-1" tabindex="-1"></a><span class="do">## Plot clusters in PCA space</span></span>
<span id="cb64-2"><a href="practical-session-2.html#cb64-2" tabindex="-1"></a><span class="fu">plotDimRed</span>(spe, <span class="at">type =</span> <span class="st">&quot;PCA&quot;</span>, </span>
<span id="cb64-3"><a href="practical-session-2.html#cb64-3" tabindex="-1"></a>           <span class="at">annotate =</span> <span class="st">&quot;label&quot;</span>, <span class="at">palette =</span> <span class="st">&quot;libd_layer_colors&quot;</span>)</span>
<span id="cb64-4"><a href="practical-session-2.html#cb64-4" tabindex="-1"></a></span>
<span id="cb64-5"><a href="practical-session-2.html#cb64-5" tabindex="-1"></a><span class="do">## Plot clusters in UMAP space</span></span>
<span id="cb64-6"><a href="practical-session-2.html#cb64-6" tabindex="-1"></a><span class="fu">plotDimRed</span>(spe, <span class="at">type =</span> <span class="st">&quot;UMAP&quot;</span>, </span>
<span id="cb64-7"><a href="practical-session-2.html#cb64-7" tabindex="-1"></a>           <span class="at">annotate =</span> <span class="st">&quot;label&quot;</span>, <span class="at">palette =</span> <span class="st">&quot;libd_layer_colors&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/02_clust_vis-DimRed-1.png" width="50%" /><img src="_main_files/figure-html/02_clust_vis-DimRed-2.png" width="50%" /></p>
<p>From the visualizations, we can see that the clustering reproduces, up to an extend, the known biological structure of the tissue, but not perfectly. One reason for this could be the fact that each spot may comprise of many different cells whose gene expression profiles are diluted in the overall profile of the spot, thus leading to low-quality clustering.</p>
</div>
<div id="spatially-aware-clustering" class="section level3 hasAnchor" number="3.5.4">
<h3><span class="header-section-number">3.5.4</span> Spatially-aware clustering<a href="practical-session-2.html#spatially-aware-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In STx data, we can also perform clustering that takes spatial information into account, for example to identify spatially compact or spatially connected clusters.</p>
<p>A simple strategy is to perform graph-based clustering on a set of features (columns) that includes both molecular features (gene expression) and spatial features (x-y coordinates). In this case, a crucial tuning parameter is the relative amount of scaling between the two data modalities – if the scaling is chosen poorly, either the molecular or spatial features will dominate the clustering. Depending on data availability, further modalities could also be included. In this section, we will include some examples on this clustering approach.</p>
</div>
</div>
<div id="inter-cluster-differentially-expressed-genes-dges" class="section level2 hasAnchor" number="3.6">
<h2><span class="header-section-number">3.6</span> Inter-cluster differentially expressed genes (DGEs)<a href="practical-session-2.html#inter-cluster-differentially-expressed-genes-dges" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="background-4" class="section level3 hasAnchor" number="3.6.1">
<h3><span class="header-section-number">3.6.1</span> Background<a href="practical-session-2.html#background-4" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here, we will identify differentially expressed genes bewteen clusters.</p>
<p>We will use the <code>findMarkers</code> implementation from the <code>scran</code> <span class="citation">(<a href="#ref-Lun2016Oct">Lun, McCarthy, and Marioni 2016</a>)</span>. This implementation uses a binomial test, which tests for genes that differ in the proportion expressed vs. not expressed between clusters. This is a more stringent test than the default t-tests, and tends to select genes that are easier to interpret and validate experimentally.</p>
</div>
<div id="dges-identification" class="section level3 hasAnchor" number="3.6.2">
<h3><span class="header-section-number">3.6.2</span> DGEs identification<a href="practical-session-2.html#dges-identification" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="practical-session-2.html#cb65-1" tabindex="-1"></a><span class="do">## Set gene names as row names ease of plotting</span></span>
<span id="cb65-2"><a href="practical-session-2.html#cb65-2" tabindex="-1"></a><span class="fu">rownames</span>(spe) <span class="ot">&lt;-</span> <span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_name</span>
<span id="cb65-3"><a href="practical-session-2.html#cb65-3" tabindex="-1"></a><span class="do">## Test for DGEs</span></span>
<span id="cb65-4"><a href="practical-session-2.html#cb65-4" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">findMarkers</span>(spe, <span class="at">test =</span> <span class="st">&quot;binom&quot;</span>, <span class="at">direction =</span> <span class="st">&quot;up&quot;</span>)</span>
<span id="cb65-5"><a href="practical-session-2.html#cb65-5" tabindex="-1"></a><span class="do">## Check output</span></span>
<span id="cb65-6"><a href="practical-session-2.html#cb65-6" tabindex="-1"></a>markers</span></code></pre></div>
<pre><code>## List of length 6
## names(6): 1 2 3 4 5 6</code></pre>
<p>The output from the <code>findMarkers</code> implementation is a list of length equal to the number of clusters. Each element of the list contains the Log-Fold-Change (LogFC) of each gene between one cluster and all others.</p>
</div>
<div id="dges-visualisation" class="section level3 hasAnchor" number="3.6.3">
<h3><span class="header-section-number">3.6.3</span> DGEs visualisation<a href="practical-session-2.html#dges-visualisation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here we will plot LogFCs for cluster 1 against all other clusters</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="practical-session-2.html#cb67-1" tabindex="-1"></a><span class="do">## Select cluster 1 genes</span></span>
<span id="cb67-2"><a href="practical-session-2.html#cb67-2" tabindex="-1"></a>interesting <span class="ot">&lt;-</span> markers[[<span class="dv">1</span>]]</span>
<span id="cb67-3"><a href="practical-session-2.html#cb67-3" tabindex="-1"></a><span class="do">## Get the top genes</span></span>
<span id="cb67-4"><a href="practical-session-2.html#cb67-4" tabindex="-1"></a>best_set <span class="ot">&lt;-</span> interesting[interesting<span class="sc">$</span>Top <span class="sc">&lt;=</span> <span class="dv">5</span>, ]</span>
<span id="cb67-5"><a href="practical-session-2.html#cb67-5" tabindex="-1"></a><span class="do">## Calculate the effect</span></span>
<span id="cb67-6"><a href="practical-session-2.html#cb67-6" tabindex="-1"></a>logFCs <span class="ot">&lt;-</span> <span class="fu">getMarkerEffects</span>(best_set)</span>
<span id="cb67-7"><a href="practical-session-2.html#cb67-7" tabindex="-1"></a><span class="do">## Plot a heat map</span></span>
<span id="cb67-8"><a href="practical-session-2.html#cb67-8" tabindex="-1"></a><span class="fu">pheatmap</span>(logFCs, <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="dv">5</span>, <span class="dv">5</span>, <span class="at">length.out =</span> <span class="dv">101</span>))</span></code></pre></div>
<p><img src="_main_files/figure-html/02_dges_vis-clst1-1.png" width="50%" /></p>
<p>Below we will plot the log-transformed normalised expression of the top genes for one cluster alongside their expression in the other clusters.</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="practical-session-2.html#cb68-1" tabindex="-1"></a><span class="do">## Select genes</span></span>
<span id="cb68-2"><a href="practical-session-2.html#cb68-2" tabindex="-1"></a>top_genes <span class="ot">&lt;-</span> <span class="fu">head</span>(<span class="fu">rownames</span>(interesting))</span>
<span id="cb68-3"><a href="practical-session-2.html#cb68-3" tabindex="-1"></a><span class="do">## Plot expression</span></span>
<span id="cb68-4"><a href="practical-session-2.html#cb68-4" tabindex="-1"></a><span class="fu">plotExpression</span>(spe, <span class="at">x =</span> <span class="st">&quot;label&quot;</span>, <span class="at">features =</span> top_genes)</span></code></pre></div>
<p><img src="_main_files/figure-html/02_dges_vis-2-1.png" width="672" /></p>
</div>
</div>
<div id="other-visualisations" class="section level2 hasAnchor" number="3.7">
<h2><span class="header-section-number">3.7</span> Other visualisations<a href="practical-session-2.html#other-visualisations" class="anchor-section" aria-label="Anchor link to header"></a></h2>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Amezquita2020Feb" class="csl-entry">
Amezquita, Robert A., Aaron T. L. Lun, Etienne Becht, Vince J. Carey, Lindsay N. Carpp, Ludwig Geistlinger, Federico Marini, et al. 2020. <span>“<span class="nocase">Orchestrating single-cell analysis with Bioconductor</span>.”</span> <em>Nat Methods</em> 17 (February): 137–45. <a href="https://doi.org/10.1038/s41592-019-0654-x">https://doi.org/10.1038/s41592-019-0654-x</a>.
</div>
<div id="ref-Hao2021Jun" class="csl-entry">
Hao, Yuhan, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck, Shiwei Zheng, Andrew Butler, Maddie J. Lee, et al. 2021. <span>“<span class="nocase">Integrated analysis of multimodal single-cell data</span>.”</span> <em>Cell</em> 184 (13): 3573–3587.e29. <a href="https://doi.org/10.1016/j.cell.2021.04.048">https://doi.org/10.1016/j.cell.2021.04.048</a>.
</div>
<div id="ref-Keogh2017Apr" class="csl-entry">
Keogh, Eamonn, and Abdullah Mueen. 2017. <span>“<span class="nocase">Curse of Dimensionality</span>.”</span> In <em><span class="nocase">Encyclopedia of Machine Learning and Data Mining</span></em>, 314–15. Boston, MA, USA: Springer, Boston, MA. <a href="https://doi.org/10.1007/978-1-4899-7687-1_192">https://doi.org/10.1007/978-1-4899-7687-1_192</a>.
</div>
<div id="ref-Li2022Jan" class="csl-entry">
Li, Yijun, Stefan Stanojevic, Bing He, Zheng Jing, Qianhui Huang, Jian Kang, and Lana X. Garmire. 2022. <span>“<span class="nocase">Benchmarking Computational Integration Methods for Spatial Transcriptomics Data</span>.”</span> <em>bioRxiv</em>, January, 2021.08.27.457741. <a href="https://doi.org/10.1101/2021.08.27.457741">https://doi.org/10.1101/2021.08.27.457741</a>.
</div>
<div id="ref-Lun2016Oct" class="csl-entry">
Lun, Aaron T. L., Davis J. McCarthy, and John C. Marioni. 2016. <span>“<span class="nocase">A step-by-step workflow for low-level analysis of single-cell RNA-seq data with Bioconductor</span>.”</span> <em>F1000Research</em> 5 (2122): 2122. <a href="https://doi.org/10.12688/f1000research.9501.2">https://doi.org/10.12688/f1000research.9501.2</a>.
</div>
<div id="ref-Maynard2021Mar" class="csl-entry">
Maynard, Kristen R., Leonardo Collado-Torres, Lukas M. Weber, Cedric Uytingco, Brianna K. Barry, Stephen R. Williams, Joseph L. Catallini, et al. 2021. <span>“<span class="nocase">Transcriptome-scale spatial gene expression in the human dorsolateral prefrontal cortex</span>.”</span> <em>Nat Neurosci</em> 24 (March): 425–36. <a href="https://doi.org/10.1038/s41593-020-00787-0">https://doi.org/10.1038/s41593-020-00787-0</a>.
</div>
<div id="ref-McCarthy2017Apr" class="csl-entry">
McCarthy, Davis J., Kieran R. Campbell, Aaron T. L. Lun, and Quin F. Wills. 2017. <span>“<span class="nocase">Scater: pre-processing, quality control, normalization and visualization of single-cell RNA-seq data in R</span>.”</span> <em>Bioinformatics</em> 33 (8): 1179–86. <a href="https://doi.org/10.1093/bioinformatics/btw777">https://doi.org/10.1093/bioinformatics/btw777</a>.
</div>
<div id="ref-McInnes2018Feb" class="csl-entry">
McInnes, Leland, John Healy, and James Melville. 2018. <span>“<span class="nocase">UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction</span>.”</span> <em>arXiv</em>, February. <a href="https://doi.org/10.48550/arXiv.1802.03426">https://doi.org/10.48550/arXiv.1802.03426</a>.
</div>
<div id="ref-Moran1950Jun" class="csl-entry">
<span>“<span class="nocase">Notes on Continuous Stochastic Phenomena on JSTOR</span>.”</span> 1950. <em>Biometrika</em>. <a href="https://www.jstor.org/stable/2332142">https://www.jstor.org/stable/2332142</a>.
</div>
<div id="ref-Sun2020Feb" class="csl-entry">
Sun, Shiquan, Jiaqiang Zhu, and Xiang Zhou. 2020. <span>“<span class="nocase">Statistical analysis of spatial expression patterns for spatially resolved transcriptomic studies</span>.”</span> <em>Nat Methods</em> 17 (February): 193–200. <a href="https://doi.org/10.1038/s41592-019-0701-7">https://doi.org/10.1038/s41592-019-0701-7</a>.
</div>
<div id="ref-Svensson2018May" class="csl-entry">
Svensson, Valentine, Sarah A. Teichmann, and Oliver Stegle. 2018. <span>“<span class="nocase">SpatialDE: identification of spatially variable genes</span>.”</span> <em>Nat Methods</em> 15 (May): 343–46. <a href="https://doi.org/10.1038/nmeth.4636">https://doi.org/10.1038/nmeth.4636</a>.
</div>
<div id="ref-Geary1954Nov" class="csl-entry">
<span>“<span class="nocase">The Contiguity Ratio and Statistical Mapping on JSTOR</span>.”</span> 1954. <em>Incorporated Statistician</em>. <a href="https://www.jstor.org/stable/2986645">https://www.jstor.org/stable/2986645</a>.
</div>
<div id="ref-ggspavis2023Apr" class="csl-entry">
Weber, Lukas M., and Helena L. Crowell. 2022. <em>Ggspavis: Visualization Functions for Spatially Resolved Transcriptomics Data</em>. <a href="https://github.com/lmweber/ggspavis">https://github.com/lmweber/ggspavis</a>.
</div>
<div id="ref-Zhu2021Dec" class="csl-entry">
Zhu, Jiaqiang, Shiquan Sun, and Xiang Zhou. 2021. <span>“<span class="nocase">SPARK-X: non-parametric modeling enables scalable and robust detection of spatial expression patterns for large spatial transcriptomic studies</span>.”</span> <em>Genome Biol</em> 22 (1): 1–25. <a href="https://doi.org/10.1186/s13059-021-02404-0">https://doi.org/10.1186/s13059-021-02404-0</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="practical-session-1.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="practical-session-2-1.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/02-stx-analysis.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
