<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Practical session 3 | Spatial transcriptomics data analysis: theory and practice</title>
  <meta name="description" content="This book will guide you through the practical steps of the in-person tutorial IP2 for the ISMB/ECCB 2023 conference in Lyon named: Spatial transcriptomics data analysis: theory and practice." />
  <meta name="generator" content="bookdown 0.33 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Practical session 3 | Spatial transcriptomics data analysis: theory and practice" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This book will guide you through the practical steps of the in-person tutorial IP2 for the ISMB/ECCB 2023 conference in Lyon named: Spatial transcriptomics data analysis: theory and practice." />
  <meta name="github-repo" content="https://github.com/ncl-icbam/ismb-tutorial-2023.git" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Practical session 3 | Spatial transcriptomics data analysis: theory and practice" />
  
  <meta name="twitter:description" content="This book will guide you through the practical steps of the in-person tutorial IP2 for the ISMB/ECCB 2023 conference in Lyon named: Spatial transcriptomics data analysis: theory and practice." />
  

<meta name="author" content="Eleftherios Zormpas, Dr Simon J. Cockell" />


<meta name="date" content="2023-07-19" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="practical-session-2.html"/>
<link rel="next" href="practical-session-4.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Spatial transcriptomics data analysis: theory and practice</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#abstract"><i class="fa fa-check"></i>Abstract</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#learning-objectives"><i class="fa fa-check"></i>Learning objectives</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="practical-session-1.html"><a href="practical-session-1.html"><i class="fa fa-check"></i><b>1</b> Practical session 1</a>
<ul>
<li class="chapter" data-level="1.1" data-path="practical-session-1.html"><a href="practical-session-1.html#about-this-documentation"><i class="fa fa-check"></i><b>1.1</b> About this documentation</a></li>
<li class="chapter" data-level="1.2" data-path="practical-session-1.html"><a href="practical-session-1.html#posit-cloud"><i class="fa fa-check"></i><b>1.2</b> Posit Cloud</a></li>
<li class="chapter" data-level="1.3" data-path="practical-session-1.html"><a href="practical-session-1.html#import-10x-visium-data"><i class="fa fa-check"></i><b>1.3</b> Import 10X Visium data</a></li>
<li class="chapter" data-level="1.4" data-path="practical-session-1.html"><a href="practical-session-1.html#explore-data-types"><i class="fa fa-check"></i><b>1.4</b> Explore data types</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="practical-session-1.html"><a href="practical-session-1.html#spatialexperiment-class"><i class="fa fa-check"></i><b>1.4.1</b> SpatialExperiment class</a></li>
<li class="chapter" data-level="1.4.2" data-path="practical-session-1.html"><a href="practical-session-1.html#inspect-the-object"><i class="fa fa-check"></i><b>1.4.2</b> Inspect the object</a></li>
<li class="chapter" data-level="1.4.3" data-path="practical-session-1.html"><a href="practical-session-1.html#counts-table-and-gene-metadata"><i class="fa fa-check"></i><b>1.4.3</b> Counts table and gene metadata</a></li>
<li class="chapter" data-level="1.4.4" data-path="practical-session-1.html"><a href="practical-session-1.html#coordinates-table-and-spot-metadata"><i class="fa fa-check"></i><b>1.4.4</b> Coordinates table and spot metadata</a></li>
<li class="chapter" data-level="1.4.5" data-path="practical-session-1.html"><a href="practical-session-1.html#image-metadata"><i class="fa fa-check"></i><b>1.4.5</b> Image metadata</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="practical-session-1.html"><a href="practical-session-1.html#conclusion"><i class="fa fa-check"></i><b>1.5</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="practical-session-2.html"><a href="practical-session-2.html"><i class="fa fa-check"></i><b>2</b> Practical session 2</a>
<ul>
<li class="chapter" data-level="2.1" data-path="practical-session-2.html"><a href="practical-session-2.html#spot-level-quality-control"><i class="fa fa-check"></i><b>2.1</b> Spot-level Quality Control</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="practical-session-2.html"><a href="practical-session-2.html#plot-tissue-map"><i class="fa fa-check"></i><b>2.1.1</b> Plot tissue map</a></li>
<li class="chapter" data-level="2.1.2" data-path="practical-session-2.html"><a href="practical-session-2.html#calculating-qc-metrics"><i class="fa fa-check"></i><b>2.1.2</b> Calculating QC metrics</a></li>
<li class="chapter" data-level="2.1.3" data-path="practical-session-2.html"><a href="practical-session-2.html#library-size-threshold-plot"><i class="fa fa-check"></i><b>2.1.3</b> Library size threshold plot</a></li>
<li class="chapter" data-level="2.1.4" data-path="practical-session-2.html"><a href="practical-session-2.html#number-of-expressed-genes"><i class="fa fa-check"></i><b>2.1.4</b> Number of expressed genes</a></li>
<li class="chapter" data-level="2.1.5" data-path="practical-session-2.html"><a href="practical-session-2.html#percentage-of-mitochondrial-expression"><i class="fa fa-check"></i><b>2.1.5</b> Percentage of mitochondrial expression</a></li>
<li class="chapter" data-level="2.1.6" data-path="practical-session-2.html"><a href="practical-session-2.html#number-of-cells-per-spot"><i class="fa fa-check"></i><b>2.1.6</b> Number of cells per spot</a></li>
<li class="chapter" data-level="2.1.7" data-path="practical-session-2.html"><a href="practical-session-2.html#remove-low-quality-spots"><i class="fa fa-check"></i><b>2.1.7</b> Remove low-quality spots</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="practical-session-2.html"><a href="practical-session-2.html#normalisation-of-counts"><i class="fa fa-check"></i><b>2.2</b> Normalisation of counts</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background"><i class="fa fa-check"></i><b>2.2.1</b> Background</a></li>
<li class="chapter" data-level="2.2.2" data-path="practical-session-2.html"><a href="practical-session-2.html#log-tranformation-of-counts"><i class="fa fa-check"></i><b>2.2.2</b> Log-tranformation of counts</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="practical-session-2.html"><a href="practical-session-2.html#selecting-genes"><i class="fa fa-check"></i><b>2.3</b> Selecting genes</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background-1"><i class="fa fa-check"></i><b>2.3.1</b> Background</a></li>
<li class="chapter" data-level="2.3.2" data-path="practical-session-2.html"><a href="practical-session-2.html#highly-variable-genes-hvgs"><i class="fa fa-check"></i><b>2.3.2</b> Highly Variable Genes (HVGs)</a></li>
<li class="chapter" data-level="2.3.3" data-path="practical-session-2.html"><a href="practical-session-2.html#spatially-variable-genes-svgs"><i class="fa fa-check"></i><b>2.3.3</b> Spatially variable genes (SVGs)</a></li>
<li class="chapter" data-level="2.3.4" data-path="practical-session-2.html"><a href="practical-session-2.html#integration-of-hvgs-and-svgs"><i class="fa fa-check"></i><b>2.3.4</b> Integration of HVGs and SVGs</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="practical-session-2.html"><a href="practical-session-2.html#dimensionality-reduction"><i class="fa fa-check"></i><b>2.4</b> Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background-2"><i class="fa fa-check"></i><b>2.4.1</b> Background</a></li>
<li class="chapter" data-level="2.4.2" data-path="practical-session-2.html"><a href="practical-session-2.html#pca-principal-component-analysis"><i class="fa fa-check"></i><b>2.4.2</b> PCA: Principal component analysis</a></li>
<li class="chapter" data-level="2.4.3" data-path="practical-session-2.html"><a href="practical-session-2.html#umap-uniform-manifold-approximation-and-projection"><i class="fa fa-check"></i><b>2.4.3</b> UMAP: Uniform Manifold Approximation and Projection</a></li>
<li class="chapter" data-level="2.4.4" data-path="practical-session-2.html"><a href="practical-session-2.html#umap-visualisations"><i class="fa fa-check"></i><b>2.4.4</b> UMAP visualisations</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="practical-session-2.html"><a href="practical-session-2.html#clustering"><i class="fa fa-check"></i><b>2.5</b> Clustering</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background-3"><i class="fa fa-check"></i><b>2.5.1</b> Background</a></li>
<li class="chapter" data-level="2.5.2" data-path="practical-session-2.html"><a href="practical-session-2.html#clustering-on-hvgs"><i class="fa fa-check"></i><b>2.5.2</b> Clustering on HVGs</a></li>
<li class="chapter" data-level="2.5.3" data-path="practical-session-2.html"><a href="practical-session-2.html#hvgs-clustering-visualisations"><i class="fa fa-check"></i><b>2.5.3</b> HVGs clustering visualisations</a></li>
<li class="chapter" data-level="2.5.4" data-path="practical-session-2.html"><a href="practical-session-2.html#spatially-aware-clustering"><i class="fa fa-check"></i><b>2.5.4</b> Spatially-aware clustering</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="practical-session-2.html"><a href="practical-session-2.html#inter-cluster-differentially-expressed-genes-dges"><i class="fa fa-check"></i><b>2.6</b> Inter-cluster differentially expressed genes (DGEs)</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background-4"><i class="fa fa-check"></i><b>2.6.1</b> Background</a></li>
<li class="chapter" data-level="2.6.2" data-path="practical-session-2.html"><a href="practical-session-2.html#dges-identification"><i class="fa fa-check"></i><b>2.6.2</b> DGEs identification</a></li>
<li class="chapter" data-level="2.6.3" data-path="practical-session-2.html"><a href="practical-session-2.html#dges-visualisation"><i class="fa fa-check"></i><b>2.6.3</b> DGEs visualisation</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="practical-session-2.html"><a href="practical-session-2.html#other-visualisations"><i class="fa fa-check"></i><b>2.7</b> Other visualisations</a></li>
<li class="chapter" data-level="2.8" data-path="practical-session-2.html"><a href="practical-session-2.html#putting-it-all-together"><i class="fa fa-check"></i><b>2.8</b> Putting it all together</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="practical-session-3.html"><a href="practical-session-3.html"><i class="fa fa-check"></i><b>3</b> Practical session 3</a>
<ul>
<li class="chapter" data-level="3.1" data-path="practical-session-3.html"><a href="practical-session-3.html#background-5"><i class="fa fa-check"></i><b>3.1</b> Background</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="practical-session-3.html"><a href="practical-session-3.html#main-geocomputatinal-data-structures"><i class="fa fa-check"></i><b>3.1.1</b> Main geocomputatinal data structures</a></li>
<li class="chapter" data-level="3.1.2" data-path="practical-session-3.html"><a href="practical-session-3.html#the-sf-objects"><i class="fa fa-check"></i><b>3.1.2</b> The <code>sf</code> objects</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="practical-session-3.html"><a href="practical-session-3.html#data-structures-preparation"><i class="fa fa-check"></i><b>3.2</b> Data structures preparation</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="practical-session-3.html"><a href="practical-session-3.html#load-new-dataset"><i class="fa fa-check"></i><b>3.2.1</b> Load new dataset</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="practical-session-3.html"><a href="practical-session-3.html#spot-level-quality-control-1"><i class="fa fa-check"></i><b>3.3</b> Spot-level Quality Control</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="practical-session-3.html"><a href="practical-session-3.html#calculating-qc-metrics-1"><i class="fa fa-check"></i><b>3.3.1</b> Calculating QC metrics</a></li>
<li class="chapter" data-level="3.3.2" data-path="practical-session-3.html"><a href="practical-session-3.html#plot-manual-annotation"><i class="fa fa-check"></i><b>3.3.2</b> Plot manual annotation</a></li>
<li class="chapter" data-level="3.3.3" data-path="practical-session-3.html"><a href="practical-session-3.html#library-size-threshold"><i class="fa fa-check"></i><b>3.3.3</b> Library size threshold</a></li>
<li class="chapter" data-level="3.3.4" data-path="practical-session-3.html"><a href="practical-session-3.html#number-of-expressed-genes-1"><i class="fa fa-check"></i><b>3.3.4</b> Number of expressed genes</a></li>
<li class="chapter" data-level="3.3.5" data-path="practical-session-3.html"><a href="practical-session-3.html#percentage-of-mitochondrial-expression-1"><i class="fa fa-check"></i><b>3.3.5</b> Percentage of mitochondrial expression</a></li>
<li class="chapter" data-level="3.3.6" data-path="practical-session-3.html"><a href="practical-session-3.html#remove-low-quality-spots-1"><i class="fa fa-check"></i><b>3.3.6</b> Remove low-quality spots</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="practical-session-3.html"><a href="practical-session-3.html#normalisation-of-counts-1"><i class="fa fa-check"></i><b>3.4</b> Normalisation of counts</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="practical-session-3.html"><a href="practical-session-3.html#log-tranformation-of-counts-1"><i class="fa fa-check"></i><b>3.4.1</b> Log-tranformation of counts</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="practical-session-3.html"><a href="practical-session-3.html#gene-level-quality-control"><i class="fa fa-check"></i><b>3.5</b> Gene-level Quality Control</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="practical-session-3.html"><a href="practical-session-3.html#calculating-extra-qc-metrics"><i class="fa fa-check"></i><b>3.5.1</b> Calculating extra QC metrics</a></li>
<li class="chapter" data-level="3.5.2" data-path="practical-session-3.html"><a href="practical-session-3.html#set-and-apply-filters"><i class="fa fa-check"></i><b>3.5.2</b> Set and apply filters</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="practical-session-3.html"><a href="practical-session-3.html#selecting-genes-1"><i class="fa fa-check"></i><b>3.6</b> Selecting genes</a>
<ul>
<li class="chapter" data-level="3.6.1" data-path="practical-session-3.html"><a href="practical-session-3.html#highly-variable-genes-hvgs-1"><i class="fa fa-check"></i><b>3.6.1</b> Highly Variable Genes (HVGs)</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="practical-session-3.html"><a href="practical-session-3.html#neighbour-graph-and-distance-matrix"><i class="fa fa-check"></i><b>3.7</b> Neighbour graph and distance matrix</a>
<ul>
<li class="chapter" data-level="3.7.1" data-path="practical-session-3.html"><a href="practical-session-3.html#adding-spatial-weights"><i class="fa fa-check"></i><b>3.7.1</b> Adding spatial weights</a></li>
<li class="chapter" data-level="3.7.2" data-path="practical-session-3.html"><a href="practical-session-3.html#generate-distance-matrices"><i class="fa fa-check"></i><b>3.7.2</b> Generate distance matrices</a></li>
</ul></li>
<li class="chapter" data-level="3.8" data-path="practical-session-3.html"><a href="practical-session-3.html#putting-it-all-together-1"><i class="fa fa-check"></i><b>3.8</b> Putting it all together</a></li>
<li class="chapter" data-level="3.9" data-path="practical-session-3.html"><a href="practical-session-3.html#save-data"><i class="fa fa-check"></i><b>3.9</b> Save data</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="practical-session-4.html"><a href="practical-session-4.html"><i class="fa fa-check"></i><b>4</b> Practical session 4</a>
<ul>
<li class="chapter" data-level="4.1" data-path="practical-session-4.html"><a href="practical-session-4.html#geographically-weighted-principal-components-analysis-gwpca"><i class="fa fa-check"></i><b>4.1</b> Geographically Weighted Principal Components Analysis (GWPCA)</a></li>
<li class="chapter" data-level="4.2" data-path="practical-session-4.html"><a href="practical-session-4.html#load-packages"><i class="fa fa-check"></i><b>4.2</b> Load packages</a></li>
<li class="chapter" data-level="4.3" data-path="practical-session-4.html"><a href="practical-session-4.html#load-quality-controled-and-normalised-data"><i class="fa fa-check"></i><b>4.3</b> Load Quality Controled and Normalised data</a></li>
<li class="chapter" data-level="4.4" data-path="practical-session-4.html"><a href="practical-session-4.html#parameter-prearation-for-gwpca"><i class="fa fa-check"></i><b>4.4</b> Parameter prearation for GWPCA</a></li>
<li class="chapter" data-level="4.5" data-path="practical-session-4.html"><a href="practical-session-4.html#run-gwpca"><i class="fa fa-check"></i><b>4.5</b> Run GWPCA</a></li>
<li class="chapter" data-level="4.6" data-path="practical-session-4.html"><a href="practical-session-4.html#plot-global-pca-results"><i class="fa fa-check"></i><b>4.6</b> Plot global PCA results</a></li>
<li class="chapter" data-level="4.7" data-path="practical-session-4.html"><a href="practical-session-4.html#identify-the-leading-genes-in-each-location"><i class="fa fa-check"></i><b>4.7</b> Identify the leading genes in each location</a></li>
<li class="chapter" data-level="4.8" data-path="practical-session-4.html"><a href="practical-session-4.html#percentage-of-total-variation-ptv"><i class="fa fa-check"></i><b>4.8</b> Percentage of Total Variation (PTV)</a></li>
<li class="chapter" data-level="4.9" data-path="practical-session-4.html"><a href="practical-session-4.html#identify-discrepancies"><i class="fa fa-check"></i><b>4.9</b> Identify discrepancies</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Spatial transcriptomics data analysis: theory and practice</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="practical-session-3" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">Chapter 3</span> Practical session 3<a href="practical-session-3.html#practical-session-3" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>This practical session will demonstrate the application of the most commonly used spatial analysis tools to STx data, and how we work with coordinate data alongside expression data. ## Load packages</p>
<ul>
<li><p><a href="https://cran.r-project.org/web/packages/spdep/index.html"><code>spdep</code></a> is a collection of functions to create spatial weights matrix objects from polygon <em>‘contiguities’</em>, from point patterns by distance and tessellations, for summarizing these objects, and for permitting their use in spatial data analysis like regional aggregation and tests for spatial <em>‘autocorrelation’</em>.</p></li>
<li><p><a href="https://cran.r-project.org/web/packages/sf/index.html"><code>sf</code></a> (<em>Simple Features for R</em>) is a package that offers support for simple features, a standardized way to encode spatial vector data.</p></li>
<li><p><a href="https://cran.r-project.org/web/packages/GWmodel/index.html"><code>GWmodel</code></a> is a suit of models that fit situations when data are not described well by some global model, but where there are spatial regions where a suitably localised calibration provides a better description.</p></li>
</ul>
<div id="background-5" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> Background<a href="practical-session-3.html#background-5" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="main-geocomputatinal-data-structures" class="section level3 hasAnchor" number="3.1.1">
<h3><span class="header-section-number">3.1.1</span> Main geocomputatinal data structures<a href="practical-session-3.html#main-geocomputatinal-data-structures" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>There are three main data structures that we need to have ready before we undertake a geocomputational approach to STx data analysis. Namely these are; (1) geometries (point and polygon), (2) neighbours lists and (3) distance matrices.</p>
<ol style="list-style-type: decimal">
<li><p>Spatial geometries can be points, lines, polygons and pixels. Polygons consist of a multitude of points connected by lines and can have many forms like circle, hexagon, non-canonical polygon etc.</p></li>
<li><p>Neighbour lists are special types of lists that contain information about the neighbours of each polygon. The neighbours can be defined either by adjacency or by distance.</p></li>
<li><p>Distance matrices contain the distances between different points and can be either weighted or un-weighted. The weighted distances are usually objective to each point and its neighbours. Meaning that the closer or farther a neighbour is from the point of focus, the weight of their distance changes according to an applied kernel. Usually in the case of STx data, like the ones generated by the 10X Visium platform, the un-weighted distance between is two points is in pixels and we acquire it from the <code>spaceranger</code> output.</p></li>
</ol>
</div>
<div id="the-sf-objects" class="section level3 hasAnchor" number="3.1.2">
<h3><span class="header-section-number">3.1.2</span> The <code>sf</code> objects<a href="practical-session-3.html#the-sf-objects" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Package <code>sf</code> represents simple features as native R objects. All functions and methods in <code>sf</code> that operate on spatial data are prefixed by <em>st_</em>, which refers to <em>spatial type</em>. Simple features are implemented as R native data, using simple data structures (S3 classes, lists, matrix, vector). Typical use involves reading, manipulating and writing of sets of features, with attributes and geometries.</p>
<p>As attributes are typically stored in <code>data.frame</code> objects (or the very similar <code>tbl_df</code>), we will also store feature geometries in a <code>data.frame</code> column. Since geometries are not single-valued, they are put in a list-column, a list of length equal to the number of records in the <code>data.frame</code>, with each list element holding the simple feature geometry of that feature. The three classes used to represent simple features are:</p>
<ul>
<li><code>sf</code>, the table (<code>data.frame</code>) with feature attributes and feature geometries, which contains</li>
<li><code>sfc</code>, the list-column with the geometries for each feature (record), which is composed of</li>
<li><code>sfg</code>, the feature geometry of an individual simple feature.</li>
</ul>
<div id="simple-feature-geometry-types" class="section level4 hasAnchor" number="3.1.2.1">
<h4><span class="header-section-number">3.1.2.1</span> Simple feature geometry types<a href="practical-session-3.html#simple-feature-geometry-types" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The following seven simple feature types are the most common:</p>
<table>
<colgroup>
<col width="18%" />
<col width="81%" />
</colgroup>
<thead>
<tr class="header">
<th>type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>POINT</code></td>
<td>zero-dimensional geometry containing a single point</td>
</tr>
<tr class="even">
<td><code>LINESTRING</code></td>
<td>sequence of points connected by straight, non-self intersecting line pieces; one-dimensional geometry</td>
</tr>
<tr class="odd">
<td><code>POLYGON</code></td>
<td>geometry with a positive area (two-dimensional); sequence of points form a closed, non-self intersecting ring; the first ring denotes the exterior ring, zero or more subsequent rings denote holes in this exterior ring</td>
</tr>
<tr class="even">
<td><code>MULTIPOINT</code></td>
<td>set of points; a MULTIPOINT is simple if no two Points in the MULTIPOINT are equal</td>
</tr>
<tr class="odd">
<td><code>MULTILINESTRING</code></td>
<td>set of linestrings</td>
</tr>
<tr class="even">
<td><code>MULTIPOLYGON</code></td>
<td>set of polygons</td>
</tr>
<tr class="odd">
<td><code>GEOMETRYCOLLECTION</code></td>
<td>set of geometries of any type except GEOMETRYCOLLECTION</td>
</tr>
</tbody>
</table>
<p>Each of the geometry types can also be a (typed) empty set, containing zero coordinates (for <code>POINT</code> the standard is not clear how to represent the empty geometry). Empty geometries can be thought of being the analogue to missing (<code>NA</code>) attributes, NULL values or empty lists.</p>
</div>
<div id="sf-objects-with-simple-features" class="section level4 hasAnchor" number="3.1.2.2">
<h4><span class="header-section-number">3.1.2.2</span> sf: objects with simple features<a href="practical-session-3.html#sf-objects-with-simple-features" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>As we usually do not work with geometries of single <code>simple features</code>, but with datasets consisting of sets of features with attributes, the two are put together in <code>sf</code> (simple feature) objects. The following command reads a test dataset called <code>nc</code> from a file that is contained in the <code>sf</code> package:</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="practical-session-3.html#cb103-1" tabindex="-1"></a>nc <span class="ot">&lt;-</span> <span class="fu">st_read</span>(<span class="fu">system.file</span>(<span class="st">&quot;shape/nc.shp&quot;</span>, <span class="at">package =</span> <span class="st">&quot;sf&quot;</span>))</span></code></pre></div>
<pre><code>## Reading layer `nc&#39; from data source 
##   `/Users/b9047753/Library/R/arm64/4.2/library/sf/shape/nc.shp&#39; 
##   using driver `ESRI Shapefile&#39;
## Simple feature collection with 100 features and 14 fields
## Geometry type: MULTIPOLYGON
## Dimension:     XY
## Bounding box:  xmin: -84.32385 ymin: 33.88199 xmax: -75.45698 ymax: 36.58965
## Geodetic CRS:  NAD27</code></pre>
<p>The short report printed gives the file name, the driver (ESRI Shapefile), mentions that there are 100 features (records, represented as rows) and 14 fields (attributes, represented as columns).</p>
<p>This object is of class:</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="practical-session-3.html#cb105-1" tabindex="-1"></a><span class="fu">class</span>(nc)</span></code></pre></div>
<pre><code>## [1] &quot;sf&quot;         &quot;data.frame&quot;</code></pre>
<p>meaning it extends (and “is” a) <code>data.frame</code>, but with a single list-column with geometries, which is held in the column with name:</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="practical-session-3.html#cb107-1" tabindex="-1"></a><span class="fu">attr</span>(nc, <span class="st">&quot;sf_column&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;geometry&quot;</code></pre>
<p>If we print the first three features, we see their attribute values and an abridged version of the geometry</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="practical-session-3.html#cb109-1" tabindex="-1"></a><span class="fu">print</span>(nc[<span class="dv">9</span><span class="sc">:</span><span class="dv">15</span>], <span class="at">n =</span> <span class="dv">3</span>)</span></code></pre></div>
<p>which would give the following output:</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:Sf-overview"></span>
<img src="images/sf_xfig.png" alt="Overview of the `sf` object." width="100%" />
<p class="caption">
Figure 3.1: Overview of the <code>sf</code> object.
</p>
</div>
<p>In the output we see:</p>
<ul>
<li>in green a simple feature: a single record, or <code>data.frame</code> row, consisting of attributes and geometry</li>
<li>in blue a single simple feature geometry (an object of class <code>sfg</code>)</li>
<li>in red a simple feature list-column (an object of class <code>sfc</code>, which is a column in the <code>data.frame</code>)</li>
<li>that although geometries are native R objects, they are printed as <a href="#wkb">well-known text</a></li>
</ul>
<p>It is also possible to create <code>data.frame</code> objects with geometry list-columns that are not of class <code>sf</code>, e.g. by:</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="practical-session-3.html#cb110-1" tabindex="-1"></a>nc.no_sf <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(nc)</span>
<span id="cb110-2"><a href="practical-session-3.html#cb110-2" tabindex="-1"></a><span class="fu">class</span>(nc.no_sf)</span></code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<p>However, such objects:</p>
<ul>
<li>no longer register which column is the geometry list-column</li>
<li>no longer have a plot method, and</li>
<li>lack all of the other dedicated methods for class <code>sf</code></li>
</ul>
</div>
<div id="sfc-simple-feature-geometry-list-column" class="section level4 hasAnchor" number="3.1.2.3">
<h4><span class="header-section-number">3.1.2.3</span> sfc: simple feature geometry list-column<a href="practical-session-3.html#sfc-simple-feature-geometry-list-column" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The column in the <code>sf</code> data.frame that contains the geometries is a list, of class <code>sfc</code>. We can retrieve the geometry list-column in this case by using standard <code>data.frame</code> notation like <code>nc$geom</code> or <code>nc[[15]]</code>, but the more general way uses <code>st_geometry</code>:</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="practical-session-3.html#cb112-1" tabindex="-1"></a>(nc_geom <span class="ot">&lt;-</span> <span class="fu">st_geometry</span>(nc))</span></code></pre></div>
<pre><code>## Geometry set for 100 features 
## Geometry type: MULTIPOLYGON
## Dimension:     XY
## Bounding box:  xmin: -84.32385 ymin: 33.88199 xmax: -75.45698 ymax: 36.58965
## Geodetic CRS:  NAD27
## First 5 geometries:</code></pre>
<pre><code>## MULTIPOLYGON (((-81.47276 36.23436, -81.54084 3...</code></pre>
<pre><code>## MULTIPOLYGON (((-81.23989 36.36536, -81.24069 3...</code></pre>
<pre><code>## MULTIPOLYGON (((-80.45634 36.24256, -80.47639 3...</code></pre>
<pre><code>## MULTIPOLYGON (((-76.00897 36.3196, -76.01735 36...</code></pre>
<pre><code>## MULTIPOLYGON (((-77.21767 36.24098, -77.23461 3...</code></pre>
<p>Geometries are printed in abbreviated form, but we can view a complete geometry by selecting it, e.g. the first one by:</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="practical-session-3.html#cb119-1" tabindex="-1"></a>nc_geom[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>## MULTIPOLYGON (((-81.47276 36.23436, -81.54084 36.27251, -81.56198 36.27359, -81.63306 36.34069, -81.74107 36.39178, -81.69828 36.47178, -81.7028 36.51934, -81.67 36.58965, -81.3453 36.57286, -81.34754 36.53791, -81.32478 36.51368, -81.31332 36.4807, -81.26624 36.43721, -81.26284 36.40504, -81.24069 36.37942, -81.23989 36.36536, -81.26424 36.35241, -81.32899 36.3635, -81.36137 36.35316, -81.36569 36.33905, -81.35413 36.29972, -81.36745 36.2787, -81.40639 36.28505, -81.41233 36.26729, -81.43104 36.26072, -81.45289 36.23959, -81.47276 36.23436)))</code></pre>
<p>The way this is printed is called <em>well-known text</em>, and is part of the standards. The word <code>MULTIPOLYGON</code> is followed by three parentheses, because it can consist of multiple polygons, in the form of <code>MULTIPOLYGON(POL1,POL2)</code>, where <code>POL1</code> might consist of an exterior ring and zero or more interior rings, as of <code>(EXT1,HOLE1,HOLE2)</code>. Sets of coordinates are held together with parentheses, so we get <code>((crds_ext)(crds_hole1)(crds_hole2))</code> where <code>crds_</code> is a comma-separated set of coordinates of a ring. This leads to the case above, where <code>MULTIPOLYGON(((crds_ext)))</code> refers to the exterior ring (1), without holes (2), of the first polygon (3) - hence three parentheses.</p>
<p>We can see there is a single polygon with no rings:</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="practical-session-3.html#cb121-1" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">0</span>))</span>
<span id="cb121-2"><a href="practical-session-3.html#cb121-2" tabindex="-1"></a><span class="fu">plot</span>(nc[<span class="dv">1</span>], <span class="at">reset =</span> <span class="cn">FALSE</span>) <span class="co"># reset = FALSE: we want to add to a plot with a legend</span></span>
<span id="cb121-3"><a href="practical-session-3.html#cb121-3" tabindex="-1"></a><span class="fu">plot</span>(nc[<span class="dv">1</span>,<span class="dv">1</span>], <span class="at">col =</span> <span class="st">&#39;grey&#39;</span>, <span class="at">add =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/03_sfc_Test3-1.png" width="672" /></p>
<p>Following the <code>MULTIPOLYGON</code> data structure, in R we have a list of lists of lists of matrices. For instance, we get the first 3 coordinate pairs of the second exterior ring (first ring is always exterior) for the geometry of feature 4 by:</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="practical-session-3.html#cb122-1" tabindex="-1"></a>nc_geom[[<span class="dv">4</span>]][[<span class="dv">2</span>]][[<span class="dv">1</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,]</span></code></pre></div>
<pre><code>##           [,1]     [,2]
## [1,] -76.02717 36.55672
## [2,] -75.99866 36.55665
## [3,] -75.91192 36.54253</code></pre>
<p>Geometry columns have their own class,</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="practical-session-3.html#cb124-1" tabindex="-1"></a><span class="fu">class</span>(nc_geom)</span></code></pre></div>
<pre><code>## [1] &quot;sfc_MULTIPOLYGON&quot; &quot;sfc&quot;</code></pre>
</div>
<div id="sfg-simple-feature-geometry" class="section level4 hasAnchor" number="3.1.2.4">
<h4><span class="header-section-number">3.1.2.4</span> sfg: simple feature geometry<a href="practical-session-3.html#sfg-simple-feature-geometry" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Simple feature geometry (<code>sfg</code>) objects carry the geometry for a single feature, e.g. a point, linestring or polygon.</p>
<p>Simple feature geometries are implemented as R native data, using the following rules</p>
<ol style="list-style-type: decimal">
<li>a single POINT is a numeric vector</li>
<li>a set of points, e.g. in a LINESTRING or ring of a POLYGON is a <code>matrix</code>, each row containing a point</li>
<li>any other set is a <code>list</code></li>
</ol>
<p>The below figure illustrates the different types of geometries:</p>
<p><img src="_main_files/figure-html/03_sf_Test7-1.png" width="672" /></p>
<p>Geometries can also be empty, as in</p>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="practical-session-3.html#cb126-1" tabindex="-1"></a>(x <span class="ot">&lt;-</span> <span class="fu">st_geometrycollection</span>())</span>
<span id="cb126-2"><a href="practical-session-3.html#cb126-2" tabindex="-1"></a><span class="do">## GEOMETRYCOLLECTION EMPTY</span></span>
<span id="cb126-3"><a href="practical-session-3.html#cb126-3" tabindex="-1"></a><span class="fu">length</span>(x)</span>
<span id="cb126-4"><a href="practical-session-3.html#cb126-4" tabindex="-1"></a><span class="do">## [1] 0</span></span></code></pre></div>
<p><em>The above are taken from the very well written, well-descriptive and thorough <code>sf</code> package <a href="https://cran.r-project.org/web/packages/sf/vignettes/sf1.html">vignette</a>.</em></p>
</div>
</div>
</div>
<div id="data-structures-preparation" class="section level2 hasAnchor" number="3.2">
<h2><span class="header-section-number">3.2</span> Data structures preparation<a href="practical-session-3.html#data-structures-preparation" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>For this practical we will be using a human steatotic kidney dataset from the <a href="https://livercellatlas.org/index.php">Liver Atlas</a> <span class="citation">(<a href="#ref-GUILLIAMS2022379">Guilliams et al. 2022</a>)</span>. Specifically we will use the JBO019 sample.</p>
<div id="load-new-dataset" class="section level3 hasAnchor" number="3.2.1">
<h3><span class="header-section-number">3.2.1</span> Load new dataset<a href="practical-session-3.html#load-new-dataset" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>First we generate the <code>SpatialFeaturesExperiment</code> object which is an extension of the <code>SpatialExperiment</code> (SPE) object that we used in the 2nd practical session. The difference is that the SFE object has incorporated the <code>sf</code> object structure and thus can accommodate the use of geocomputational tools.</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="practical-session-3.html#cb127-1" tabindex="-1"></a>sampleDir <span class="ot">&lt;-</span> <span class="st">&quot;./data/spaceranger_outs/Human_Liver_Steatotic/JBO019_Results&quot;</span></span>
<span id="cb127-2"><a href="practical-session-3.html#cb127-2" tabindex="-1"></a>sampleNames <span class="ot">&lt;-</span> <span class="st">&quot;JBO019&quot;</span></span>
<span id="cb127-3"><a href="practical-session-3.html#cb127-3" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> <span class="fu">read10xVisiumSFE</span>(<span class="at">samples =</span> sampleDir, </span>
<span id="cb127-4"><a href="practical-session-3.html#cb127-4" tabindex="-1"></a>                        <span class="at">sample_id =</span> sampleNames, </span>
<span id="cb127-5"><a href="practical-session-3.html#cb127-5" tabindex="-1"></a>                        <span class="at">type =</span> <span class="st">&quot;sparse&quot;</span>, </span>
<span id="cb127-6"><a href="practical-session-3.html#cb127-6" tabindex="-1"></a>                        <span class="at">data =</span> <span class="st">&quot;filtered&quot;</span>, </span>
<span id="cb127-7"><a href="practical-session-3.html#cb127-7" tabindex="-1"></a>                        <span class="at">images =</span> <span class="st">&quot;lowres&quot;</span>, </span>
<span id="cb127-8"><a href="practical-session-3.html#cb127-8" tabindex="-1"></a>                        <span class="at">style =</span> <span class="st">&quot;W&quot;</span>, </span>
<span id="cb127-9"><a href="practical-session-3.html#cb127-9" tabindex="-1"></a>                        <span class="at">zero.policy =</span> <span class="cn">TRUE</span>)</span>
<span id="cb127-10"><a href="practical-session-3.html#cb127-10" tabindex="-1"></a></span>
<span id="cb127-11"><a href="practical-session-3.html#cb127-11" tabindex="-1"></a>ground_truth <span class="ot">&lt;-</span> <span class="fu">read_table</span>(<span class="st">&quot;./data/to_load/spotzonationGroup.txt&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="spot-level-quality-control-1" class="section level2 hasAnchor" number="3.3">
<h2><span class="header-section-number">3.3</span> Spot-level Quality Control<a href="practical-session-3.html#spot-level-quality-control-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="calculating-qc-metrics-1" class="section level3 hasAnchor" number="3.3.1">
<h3><span class="header-section-number">3.3.1</span> Calculating QC metrics<a href="practical-session-3.html#calculating-qc-metrics-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here we are populating the dataset with per-spot, per-gene features and geometries required to run downstream GWPCA…</p>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="practical-session-3.html#cb128-1" tabindex="-1"></a>is_mito <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">&quot;(^MT-)|(^mt-)&quot;</span>, <span class="fu">rowData</span>(sfe)<span class="sc">$</span>symbol)</span>
<span id="cb128-2"><a href="practical-session-3.html#cb128-2" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> <span class="fu">addPerLocQC</span>(sfe, <span class="at">gTruth =</span> ground_truth, <span class="at">assay =</span> <span class="st">&quot;counts&quot;</span>, <span class="dv">2</span>, <span class="at">subsets =</span> <span class="fu">list</span>(<span class="at">mito =</span> is_mito))</span>
<span id="cb128-3"><a href="practical-session-3.html#cb128-3" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> <span class="fu">addGeometries</span>(sfe, <span class="at">samples =</span> sampleDir, <span class="at">sample_id =</span> sampleNames, <span class="at">res =</span> <span class="st">&quot;fullres&quot;</span>)</span>
<span id="cb128-4"><a href="practical-session-3.html#cb128-4" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> <span class="fu">addPerGeneQC</span>(sfe, <span class="at">assay =</span> <span class="st">&quot;counts&quot;</span>, <span class="at">version =</span> <span class="cn">NULL</span>, <span class="at">mirror =</span> <span class="cn">NULL</span>)</span>
<span id="cb128-5"><a href="practical-session-3.html#cb128-5" tabindex="-1"></a></span>
<span id="cb128-6"><a href="practical-session-3.html#cb128-6" tabindex="-1"></a><span class="fu">colData</span>(sfe)</span></code></pre></div>
<pre><code>## DataFrame with 1185 rows and 15 columns
##                    in_tissue array_row array_col   sample_id            Barcode
##                    &lt;logical&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;        &lt;character&gt;
## AAACAAGTATCTCCCA-1      TRUE        50       102      JBO019 AAACAAGTATCTCCCA-1
## AAACATTTCCCGGATT-1      TRUE        61        97      JBO019 AAACATTTCCCGGATT-1
## AAACCCGAACGAAATC-1      TRUE        45       115      JBO019 AAACCCGAACGAAATC-1
## AAACGAGACGGTTGAT-1      TRUE        35        79      JBO019 AAACGAGACGGTTGAT-1
## AAACTAACGTGGCGAC-1      TRUE         8       110      JBO019 AAACTAACGTGGCGAC-1
## ...                      ...       ...       ...         ...                ...
## TTGTAATCCGTACTCG-1      TRUE        35        55      JBO019 TTGTAATCCGTACTCG-1
## TTGTGAACCTAATCCG-1      TRUE        56        90      JBO019 TTGTGAACCTAATCCG-1
## TTGTGCAGCCACGTCA-1      TRUE        60        74      JBO019 TTGTGCAGCCACGTCA-1
## TTGTGTTTCCCGAAAG-1      TRUE        51        59      JBO019 TTGTGTTTCCCGAAAG-1
## TTGTTGTGTGTCAAGA-1      TRUE        31        77      JBO019 TTGTTGTGTGTCAAGA-1
##                      Capt_area  annotation       index  sparsity       sum
##                    &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt;
## AAACAAGTATCTCCCA-1           1          NA      spot_1  0.910410     13443
## AAACATTTCCCGGATT-1           1          NA      spot_2  0.967805      2648
## AAACCCGAACGAAATC-1           1         Mid      spot_3  0.864958     27733
## AAACGAGACGGTTGAT-1           1     Central      spot_4  0.835818     32973
## AAACTAACGTGGCGAC-1           1          NA      spot_5  0.995418       400
## ...                        ...         ...         ...       ...       ...
## TTGTAATCCGTACTCG-1           1          NA   spot_1181  0.933716      7612
## TTGTGAACCTAATCCG-1           1          NA   spot_1182  0.955831      4299
## TTGTGCAGCCACGTCA-1           1          NA   spot_1183  0.978252      1452
## TTGTGTTTCCCGAAAG-1           1          NA   spot_1184  0.956778      3831
## TTGTTGTGTGTCAAGA-1           1         Mid   spot_1185  0.852160     27755
##                     detected subsets_mito_sum subsets_mito_detected
##                    &lt;integer&gt;        &lt;numeric&gt;             &lt;integer&gt;
## AAACAAGTATCTCCCA-1      2933             1021                    12
## AAACATTTCCCGGATT-1      1054              285                    12
## AAACCCGAACGAAATC-1      4421             2087                    12
## AAACGAGACGGTTGAT-1      5375              821                    12
## AAACTAACGTGGCGAC-1       150              182                    11
## ...                      ...              ...                   ...
## TTGTAATCCGTACTCG-1      2170              733                    11
## TTGTGAACCTAATCCG-1      1446              515                    12
## TTGTGCAGCCACGTCA-1       712               54                    10
## TTGTGTTTCCCGAAAG-1      1415              422                    11
## TTGTTGTGTGTCAAGA-1      4840              906                    12
##                    subsets_mito_percent     total
##                               &lt;numeric&gt; &lt;numeric&gt;
## AAACAAGTATCTCCCA-1              7.59503     13443
## AAACATTTCCCGGATT-1             10.76284      2648
## AAACCCGAACGAAATC-1              7.52533     27733
## AAACGAGACGGTTGAT-1              2.48992     32973
## AAACTAACGTGGCGAC-1             45.50000       400
## ...                                 ...       ...
## TTGTAATCCGTACTCG-1              9.62953      7612
## TTGTGAACCTAATCCG-1             11.97953      4299
## TTGTGCAGCCACGTCA-1              3.71901      1452
## TTGTGTTTCCCGAAAG-1             11.01540      3831
## TTGTTGTGTGTCAAGA-1              3.26428     27755</code></pre>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="practical-session-3.html#cb130-1" tabindex="-1"></a><span class="fu">rowData</span>(sfe)</span></code></pre></div>
<pre><code>## DataFrame with 32738 rows and 18 columns
##                    gene_name              id       mean  detected     total
##                  &lt;character&gt;     &lt;character&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## ENSG00000243485   MIR1302-10 ENSG00000243485 0.00000000  0.000000         0
## ENSG00000237613      FAM138A ENSG00000237613 0.00000000  0.000000         0
## ENSG00000186092        OR4F5 ENSG00000186092 0.00000000  0.000000         0
## ENSG00000238009 RP11-34P13.7 ENSG00000238009 0.00590717  0.590717         7
## ENSG00000239945 RP11-34P13.8 ENSG00000239945 0.00000000  0.000000         0
## ...                      ...             ...        ...       ...       ...
## ENSG00000215635   AC145205.1 ENSG00000215635          0         0         0
## ENSG00000268590        BAGE5 ENSG00000268590          0         0         0
## ENSG00000251180   CU459201.1 ENSG00000251180          0         0         0
## ENSG00000215616   AC002321.2 ENSG00000215616          0         0         0
## ENSG00000215611   AC002321.1 ENSG00000215611          0         0         0
##                 JBO019.sparsity JBO019.total JBO019.nLocations JBO019.s_min
##                       &lt;numeric&gt;    &lt;numeric&gt;         &lt;integer&gt;    &lt;numeric&gt;
## ENSG00000243485        1.000000            0                 0          Inf
## ENSG00000237613        1.000000            0                 0          Inf
## ENSG00000186092        1.000000            0                 0          Inf
## ENSG00000238009        0.994093            7                 7            1
## ENSG00000239945        1.000000            0                 0          Inf
## ...                         ...          ...               ...          ...
## ENSG00000215635               1            0                 0          Inf
## ENSG00000268590               1            0                 0          Inf
## ENSG00000251180               1            0                 0          Inf
## ENSG00000215616               1            0                 0          Inf
## ENSG00000215611               1            0                 0          Inf
##                 JBO019.max JBO019.s_mean JBO019.s_median JBO019.s_SD
##                  &lt;numeric&gt;     &lt;numeric&gt;       &lt;numeric&gt;   &lt;numeric&gt;
## ENSG00000243485          0           NaN              NA          NA
## ENSG00000237613          0           NaN              NA          NA
## ENSG00000186092          0           NaN              NA          NA
## ENSG00000238009          1             1               1           0
## ENSG00000239945          0           NaN              NA          NA
## ...                    ...           ...             ...         ...
## ENSG00000215635          0           NaN              NA          NA
## ENSG00000268590          0           NaN              NA          NA
## ENSG00000251180          0           NaN              NA          NA
## ENSG00000215616          0           NaN              NA          NA
## ENSG00000215611          0           NaN              NA          NA
##                 JBO019.p_mean JBO019.p_median JBO019.p_SD JBO019.s_CV
##                     &lt;numeric&gt;       &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;
## ENSG00000243485    0.00000000               0   0.0000000          NA
## ENSG00000237613    0.00000000               0   0.0000000          NA
## ENSG00000186092    0.00000000               0   0.0000000          NA
## ENSG00000238009    0.00590717               0   0.0766631           0
## ENSG00000239945    0.00000000               0   0.0000000          NA
## ...                       ...             ...         ...         ...
## ENSG00000215635             0               0           0          NA
## ENSG00000268590             0               0           0          NA
## ENSG00000251180             0               0           0          NA
## ENSG00000215616             0               0           0          NA
## ENSG00000215611             0               0           0          NA
##                 JBO019.p_CV
##                   &lt;numeric&gt;
## ENSG00000243485         NaN
## ENSG00000237613         NaN
## ENSG00000186092         NaN
## ENSG00000238009      1297.8
## ENSG00000239945         NaN
## ...                     ...
## ENSG00000215635         NaN
## ENSG00000268590         NaN
## ENSG00000251180         NaN
## ENSG00000215616         NaN
## ENSG00000215611         NaN</code></pre>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="practical-session-3.html#cb132-1" tabindex="-1"></a><span class="fu">colGeometries</span>(sfe)</span></code></pre></div>
<pre><code>## List of length 3
## names(3): spotPoly spotCntd spotHex</code></pre>
</div>
<div id="plot-manual-annotation" class="section level3 hasAnchor" number="3.3.2">
<h3><span class="header-section-number">3.3.2</span> Plot manual annotation<a href="practical-session-3.html#plot-manual-annotation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="practical-session-3.html#cb134-1" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb134-2"><a href="practical-session-3.html#cb134-2" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">geometry =</span> <span class="fu">colGeometries</span>(sfe)<span class="sc">$</span>spotHex<span class="sc">$</span>geometry, <span class="at">fill =</span> <span class="fu">colData</span>(sfe)<span class="sc">$</span>annotation)) <span class="sc">+</span> </span>
<span id="cb134-3"><a href="practical-session-3.html#cb134-3" tabindex="-1"></a>  <span class="fu">theme_void</span>() <span class="sc">+</span> </span>
<span id="cb134-4"><a href="practical-session-3.html#cb134-4" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;right&quot;</span>) <span class="sc">+</span> </span>
<span id="cb134-5"><a href="practical-session-3.html#cb134-5" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">&quot;Annotation&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/03_QC_sfe2-1.png" width="672" /></p>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="practical-session-3.html#cb135-1" tabindex="-1"></a><span class="co"># &gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;possible break point to continue on Pract 4&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt; #</span></span>
<span id="cb135-2"><a href="practical-session-3.html#cb135-2" tabindex="-1"></a></span>
<span id="cb135-3"><a href="practical-session-3.html#cb135-3" tabindex="-1"></a><span class="co"># Above we load the data and calculate some statistics, generate some </span></span>
<span id="cb135-4"><a href="practical-session-3.html#cb135-4" tabindex="-1"></a><span class="co"># neighbour graphs and we can explore the data structures. Below we </span></span>
<span id="cb135-5"><a href="practical-session-3.html#cb135-5" tabindex="-1"></a><span class="co"># apply the QC.</span></span>
<span id="cb135-6"><a href="practical-session-3.html#cb135-6" tabindex="-1"></a></span>
<span id="cb135-7"><a href="practical-session-3.html#cb135-7" tabindex="-1"></a><span class="co"># There is a catch in the above. The get.spatialNeighGraphs is going to be used</span></span>
<span id="cb135-8"><a href="practical-session-3.html#cb135-8" tabindex="-1"></a><span class="co"># here to showcase the neighbour graphs but actually it is better to be called </span></span>
<span id="cb135-9"><a href="practical-session-3.html#cb135-9" tabindex="-1"></a><span class="co"># after we remove unwanted locations through QC application</span></span>
<span id="cb135-10"><a href="practical-session-3.html#cb135-10" tabindex="-1"></a></span>
<span id="cb135-11"><a href="practical-session-3.html#cb135-11" tabindex="-1"></a><span class="co"># Ensembl version 92 does not match the dataset. Need to find out which</span></span>
<span id="cb135-12"><a href="practical-session-3.html#cb135-12" tabindex="-1"></a><span class="co"># ensembl version they used with SpaceRanger. Otherwise some genes will not have</span></span>
<span id="cb135-13"><a href="practical-session-3.html#cb135-13" tabindex="-1"></a><span class="co"># proper annotation.</span></span>
<span id="cb135-14"><a href="practical-session-3.html#cb135-14" tabindex="-1"></a></span>
<span id="cb135-15"><a href="practical-session-3.html#cb135-15" tabindex="-1"></a><span class="co"># &gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;possible break point to continue on Pract 4&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt; #</span></span></code></pre></div>
</div>
<div id="library-size-threshold" class="section level3 hasAnchor" number="3.3.3">
<h3><span class="header-section-number">3.3.3</span> Library size threshold<a href="practical-session-3.html#library-size-threshold" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb136-1"><a href="practical-session-3.html#cb136-1" tabindex="-1"></a><span class="co"># ----------------------------------------------- #</span></span>
<span id="cb136-2"><a href="practical-session-3.html#cb136-2" tabindex="-1"></a><span class="do">## Density and histogram of library sizes</span></span>
<span id="cb136-3"><a href="practical-session-3.html#cb136-3" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(sfe)),</span>
<span id="cb136-4"><a href="practical-session-3.html#cb136-4" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> sum)) <span class="sc">+</span></span>
<span id="cb136-5"><a href="practical-session-3.html#cb136-5" tabindex="-1"></a>    <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="fu">after_stat</span>(density)), </span>
<span id="cb136-6"><a href="practical-session-3.html#cb136-6" tabindex="-1"></a>                   <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>, </span>
<span id="cb136-7"><a href="practical-session-3.html#cb136-7" tabindex="-1"></a>                   <span class="at">fill =</span> <span class="st">&quot;grey&quot;</span>,</span>
<span id="cb136-8"><a href="practical-session-3.html#cb136-8" tabindex="-1"></a>                   <span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb136-9"><a href="practical-session-3.html#cb136-9" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>,</span>
<span id="cb136-10"><a href="practical-session-3.html#cb136-10" tabindex="-1"></a>                 <span class="at">adjust =</span> <span class="fl">0.5</span>,</span>
<span id="cb136-11"><a href="practical-session-3.html#cb136-11" tabindex="-1"></a>                 <span class="at">fill =</span> <span class="st">&quot;#A0CBE8&quot;</span>,</span>
<span id="cb136-12"><a href="practical-session-3.html#cb136-12" tabindex="-1"></a>                 <span class="at">colour =</span> <span class="st">&quot;#4E79A7&quot;</span>) <span class="sc">+</span></span>
<span id="cb136-13"><a href="practical-session-3.html#cb136-13" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="dv">1000</span>, <span class="cn">NA</span>),</span>
<span id="cb136-14"><a href="practical-session-3.html#cb136-14" tabindex="-1"></a>               <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>, </span>
<span id="cb136-15"><a href="practical-session-3.html#cb136-15" tabindex="-1"></a>               <span class="at">linetype =</span> <span class="st">&quot;dashed&quot;</span>) <span class="sc">+</span> </span>
<span id="cb136-16"><a href="practical-session-3.html#cb136-16" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb136-17"><a href="practical-session-3.html#cb136-17" tabindex="-1"></a>    <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb136-18"><a href="practical-session-3.html#cb136-18" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">&quot;Library size&quot;</span>) <span class="sc">+</span> </span>
<span id="cb136-19"><a href="practical-session-3.html#cb136-19" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>) <span class="sc">+</span> </span>
<span id="cb136-20"><a href="practical-session-3.html#cb136-20" tabindex="-1"></a>    <span class="fu">theme_classic</span>()</span></code></pre></div>
<pre><code>## Warning: Removed 1 rows containing missing values (`geom_vline()`).</code></pre>
<p><img src="_main_files/figure-html/03_QC_sfe3-1.png" width="672" /></p>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb138-1"><a href="practical-session-3.html#cb138-1" tabindex="-1"></a><span class="do">## Select library size threshold</span></span>
<span id="cb138-2"><a href="practical-session-3.html#cb138-2" tabindex="-1"></a>qc_lib_size <span class="ot">&lt;-</span> <span class="fu">colData</span>(sfe)<span class="sc">$</span>sum <span class="sc">&lt;</span> <span class="dv">1000</span> <span class="co">#| colData(sfe)$sum &gt; 45000</span></span>
<span id="cb138-3"><a href="practical-session-3.html#cb138-3" tabindex="-1"></a><span class="do">## Check how many spots are filtered out</span></span>
<span id="cb138-4"><a href="practical-session-3.html#cb138-4" tabindex="-1"></a><span class="fu">table</span>(qc_lib_size)</span></code></pre></div>
<pre><code>## qc_lib_size
## FALSE  TRUE 
##  1166    19</code></pre>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb140-1"><a href="practical-session-3.html#cb140-1" tabindex="-1"></a><span class="do">## Add threshold in colData</span></span>
<span id="cb140-2"><a href="practical-session-3.html#cb140-2" tabindex="-1"></a><span class="fu">colData</span>(sfe)<span class="sc">$</span>qc_lib_size <span class="ot">&lt;-</span> qc_lib_size</span>
<span id="cb140-3"><a href="practical-session-3.html#cb140-3" tabindex="-1"></a><span class="do">## Check putative spatial patterns of removed spots</span></span>
<span id="cb140-4"><a href="practical-session-3.html#cb140-4" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb140-5"><a href="practical-session-3.html#cb140-5" tabindex="-1"></a>    <span class="fu">geom_sf</span>(<span class="at">data =</span> <span class="fu">colGeometry</span>(sfe, <span class="st">&quot;spotHex&quot;</span>),</span>
<span id="cb140-6"><a href="practical-session-3.html#cb140-6" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">geometry =</span> geometry)) <span class="sc">+</span> </span>
<span id="cb140-7"><a href="practical-session-3.html#cb140-7" tabindex="-1"></a>    <span class="fu">geom_sf</span>(<span class="at">data =</span> <span class="fu">colGeometry</span>(sfe, <span class="st">&quot;spotHex&quot;</span>),</span>
<span id="cb140-8"><a href="practical-session-3.html#cb140-8" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">geometry =</span> geometry, <span class="at">fill =</span> <span class="fu">colData</span>(sfe)<span class="sc">$</span>qc_lib_size)) <span class="sc">+</span></span>
<span id="cb140-9"><a href="practical-session-3.html#cb140-9" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;grey95&quot;</span>, <span class="st">&quot;red&quot;</span>)) <span class="sc">+</span> </span>
<span id="cb140-10"><a href="practical-session-3.html#cb140-10" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">&quot;Discarded&quot;</span>) <span class="sc">+</span> </span>
<span id="cb140-11"><a href="practical-session-3.html#cb140-11" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/03_QC_sfe3-2.png" width="672" /></p>
</div>
<div id="number-of-expressed-genes-1" class="section level3 hasAnchor" number="3.3.4">
<h3><span class="header-section-number">3.3.4</span> Number of expressed genes<a href="practical-session-3.html#number-of-expressed-genes-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="practical-session-3.html#cb141-1" tabindex="-1"></a><span class="co"># ----------------------------------------------- #</span></span>
<span id="cb141-2"><a href="practical-session-3.html#cb141-2" tabindex="-1"></a><span class="do">## Density and histogram of expressed genes</span></span>
<span id="cb141-3"><a href="practical-session-3.html#cb141-3" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(sfe)),</span>
<span id="cb141-4"><a href="practical-session-3.html#cb141-4" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> detected)) <span class="sc">+</span></span>
<span id="cb141-5"><a href="practical-session-3.html#cb141-5" tabindex="-1"></a>    <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="fu">after_stat</span>(density)), </span>
<span id="cb141-6"><a href="practical-session-3.html#cb141-6" tabindex="-1"></a>                   <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>, </span>
<span id="cb141-7"><a href="practical-session-3.html#cb141-7" tabindex="-1"></a>                   <span class="at">fill =</span> <span class="st">&quot;grey&quot;</span>,</span>
<span id="cb141-8"><a href="practical-session-3.html#cb141-8" tabindex="-1"></a>                   <span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb141-9"><a href="practical-session-3.html#cb141-9" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>,</span>
<span id="cb141-10"><a href="practical-session-3.html#cb141-10" tabindex="-1"></a>                 <span class="at">adjust =</span> <span class="fl">0.5</span>,</span>
<span id="cb141-11"><a href="practical-session-3.html#cb141-11" tabindex="-1"></a>                 <span class="at">fill =</span> <span class="st">&quot;#A0CBE8&quot;</span>,</span>
<span id="cb141-12"><a href="practical-session-3.html#cb141-12" tabindex="-1"></a>                 <span class="at">colour =</span> <span class="st">&quot;#4E79A7&quot;</span>) <span class="sc">+</span> </span>
<span id="cb141-13"><a href="practical-session-3.html#cb141-13" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="dv">550</span>, <span class="cn">NA</span>),</span>
<span id="cb141-14"><a href="practical-session-3.html#cb141-14" tabindex="-1"></a>               <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>, </span>
<span id="cb141-15"><a href="practical-session-3.html#cb141-15" tabindex="-1"></a>               <span class="at">linetype =</span> <span class="st">&quot;dashed&quot;</span>) <span class="sc">+</span></span>
<span id="cb141-16"><a href="practical-session-3.html#cb141-16" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb141-17"><a href="practical-session-3.html#cb141-17" tabindex="-1"></a>    <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb141-18"><a href="practical-session-3.html#cb141-18" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">&quot;Genes expressed in each spot&quot;</span>) <span class="sc">+</span> </span>
<span id="cb141-19"><a href="practical-session-3.html#cb141-19" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>) <span class="sc">+</span> </span>
<span id="cb141-20"><a href="practical-session-3.html#cb141-20" tabindex="-1"></a>    <span class="fu">theme_classic</span>()</span></code></pre></div>
<pre><code>## Warning: Removed 1 rows containing missing values (`geom_vline()`).</code></pre>
<p><img src="_main_files/figure-html/03_QC_sfe4-1.png" width="672" /></p>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="practical-session-3.html#cb143-1" tabindex="-1"></a><span class="do">## Select expressed genes threshold</span></span>
<span id="cb143-2"><a href="practical-session-3.html#cb143-2" tabindex="-1"></a>qc_detected <span class="ot">&lt;-</span> <span class="fu">colData</span>(sfe)<span class="sc">$</span>detected <span class="sc">&lt;</span> <span class="dv">550</span> <span class="co">#| colData(sfe)$detected &gt; 6000</span></span>
<span id="cb143-3"><a href="practical-session-3.html#cb143-3" tabindex="-1"></a><span class="do">## Check how many spots are filtered out</span></span>
<span id="cb143-4"><a href="practical-session-3.html#cb143-4" tabindex="-1"></a><span class="fu">table</span>(qc_detected)</span></code></pre></div>
<pre><code>## qc_detected
## FALSE  TRUE 
##  1165    20</code></pre>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="practical-session-3.html#cb145-1" tabindex="-1"></a><span class="do">## Add threshold in colData</span></span>
<span id="cb145-2"><a href="practical-session-3.html#cb145-2" tabindex="-1"></a><span class="fu">colData</span>(sfe)<span class="sc">$</span>qc_detected <span class="ot">&lt;-</span> qc_detected</span>
<span id="cb145-3"><a href="practical-session-3.html#cb145-3" tabindex="-1"></a><span class="do">## Check for putative spatial pattern of removed spots</span></span>
<span id="cb145-4"><a href="practical-session-3.html#cb145-4" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb145-5"><a href="practical-session-3.html#cb145-5" tabindex="-1"></a>    <span class="fu">geom_sf</span>(<span class="at">data =</span> <span class="fu">colGeometry</span>(sfe, <span class="st">&quot;spotHex&quot;</span>),</span>
<span id="cb145-6"><a href="practical-session-3.html#cb145-6" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">geometry =</span> geometry)) <span class="sc">+</span> </span>
<span id="cb145-7"><a href="practical-session-3.html#cb145-7" tabindex="-1"></a>    <span class="fu">geom_sf</span>(<span class="at">data =</span> <span class="fu">colGeometry</span>(sfe, <span class="st">&quot;spotHex&quot;</span>),</span>
<span id="cb145-8"><a href="practical-session-3.html#cb145-8" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">geometry =</span> geometry, <span class="at">fill =</span> <span class="fu">colData</span>(sfe)<span class="sc">$</span>qc_detected)) <span class="sc">+</span></span>
<span id="cb145-9"><a href="practical-session-3.html#cb145-9" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;grey95&quot;</span>, <span class="st">&quot;red&quot;</span>)) <span class="sc">+</span> </span>
<span id="cb145-10"><a href="practical-session-3.html#cb145-10" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">&quot;Discarded&quot;</span>) <span class="sc">+</span> </span>
<span id="cb145-11"><a href="practical-session-3.html#cb145-11" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/03_QC_sfe4-2.png" width="672" /></p>
</div>
<div id="percentage-of-mitochondrial-expression-1" class="section level3 hasAnchor" number="3.3.5">
<h3><span class="header-section-number">3.3.5</span> Percentage of mitochondrial expression<a href="practical-session-3.html#percentage-of-mitochondrial-expression-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="practical-session-3.html#cb146-1" tabindex="-1"></a><span class="co"># ----------------------------------------------- #</span></span>
<span id="cb146-2"><a href="practical-session-3.html#cb146-2" tabindex="-1"></a><span class="do">## Density and histogram of percentage of mitochondrial expression</span></span>
<span id="cb146-3"><a href="practical-session-3.html#cb146-3" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(sfe)),</span>
<span id="cb146-4"><a href="practical-session-3.html#cb146-4" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> subsets_mito_percent)) <span class="sc">+</span></span>
<span id="cb146-5"><a href="practical-session-3.html#cb146-5" tabindex="-1"></a>    <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="fu">after_stat</span>(density)), </span>
<span id="cb146-6"><a href="practical-session-3.html#cb146-6" tabindex="-1"></a>                   <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>, </span>
<span id="cb146-7"><a href="practical-session-3.html#cb146-7" tabindex="-1"></a>                   <span class="at">fill =</span> <span class="st">&quot;grey&quot;</span>,</span>
<span id="cb146-8"><a href="practical-session-3.html#cb146-8" tabindex="-1"></a>                   <span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb146-9"><a href="practical-session-3.html#cb146-9" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>,</span>
<span id="cb146-10"><a href="practical-session-3.html#cb146-10" tabindex="-1"></a>                 <span class="at">adjust =</span> <span class="fl">0.5</span>,</span>
<span id="cb146-11"><a href="practical-session-3.html#cb146-11" tabindex="-1"></a>                 <span class="at">fill =</span> <span class="st">&quot;#A0CBE8&quot;</span>,</span>
<span id="cb146-12"><a href="practical-session-3.html#cb146-12" tabindex="-1"></a>                 <span class="at">colour =</span> <span class="st">&quot;#4E79A7&quot;</span>) <span class="sc">+</span> </span>
<span id="cb146-13"><a href="practical-session-3.html#cb146-13" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="dv">22</span>, <span class="cn">NA</span>),</span>
<span id="cb146-14"><a href="practical-session-3.html#cb146-14" tabindex="-1"></a>               <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>, </span>
<span id="cb146-15"><a href="practical-session-3.html#cb146-15" tabindex="-1"></a>               <span class="at">linetype =</span> <span class="st">&quot;dashed&quot;</span>) <span class="sc">+</span></span>
<span id="cb146-16"><a href="practical-session-3.html#cb146-16" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb146-17"><a href="practical-session-3.html#cb146-17" tabindex="-1"></a>    <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb146-18"><a href="practical-session-3.html#cb146-18" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">&quot;Percentage of mitochondrial expression&quot;</span>) <span class="sc">+</span> </span>
<span id="cb146-19"><a href="practical-session-3.html#cb146-19" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>) <span class="sc">+</span> </span>
<span id="cb146-20"><a href="practical-session-3.html#cb146-20" tabindex="-1"></a>    <span class="fu">theme_classic</span>()</span></code></pre></div>
<pre><code>## Warning: Removed 1 rows containing missing values (`geom_vline()`).</code></pre>
<p><img src="_main_files/figure-html/03_QC_sfe5-1.png" width="672" /></p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="practical-session-3.html#cb148-1" tabindex="-1"></a><span class="do">## Select mitochondrial percentage threshold</span></span>
<span id="cb148-2"><a href="practical-session-3.html#cb148-2" tabindex="-1"></a>qc_mito <span class="ot">&lt;-</span> <span class="fu">colData</span>(sfe)<span class="sc">$</span>subsets_mito_percent <span class="sc">&gt;</span> <span class="dv">22</span></span>
<span id="cb148-3"><a href="practical-session-3.html#cb148-3" tabindex="-1"></a><span class="do">## Check how many spots are filtered out</span></span>
<span id="cb148-4"><a href="practical-session-3.html#cb148-4" tabindex="-1"></a><span class="fu">table</span>(qc_mito)</span></code></pre></div>
<pre><code>## qc_mito
## FALSE  TRUE 
##  1180     5</code></pre>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="practical-session-3.html#cb150-1" tabindex="-1"></a><span class="do">## Add threshold in colData</span></span>
<span id="cb150-2"><a href="practical-session-3.html#cb150-2" tabindex="-1"></a><span class="fu">colData</span>(sfe)<span class="sc">$</span>qc_mito <span class="ot">&lt;-</span> qc_mito</span>
<span id="cb150-3"><a href="practical-session-3.html#cb150-3" tabindex="-1"></a><span class="do">## Check for putative spatial pattern of removed spots</span></span>
<span id="cb150-4"><a href="practical-session-3.html#cb150-4" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb150-5"><a href="practical-session-3.html#cb150-5" tabindex="-1"></a>    <span class="fu">geom_sf</span>(<span class="at">data =</span> <span class="fu">colGeometry</span>(sfe, <span class="st">&quot;spotHex&quot;</span>),</span>
<span id="cb150-6"><a href="practical-session-3.html#cb150-6" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">geometry =</span> geometry)) <span class="sc">+</span> </span>
<span id="cb150-7"><a href="practical-session-3.html#cb150-7" tabindex="-1"></a>    <span class="fu">geom_sf</span>(<span class="at">data =</span> <span class="fu">colGeometry</span>(sfe, <span class="st">&quot;spotHex&quot;</span>),</span>
<span id="cb150-8"><a href="practical-session-3.html#cb150-8" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">geometry =</span> geometry, <span class="at">fill =</span> <span class="fu">colData</span>(sfe)<span class="sc">$</span>qc_mito)) <span class="sc">+</span></span>
<span id="cb150-9"><a href="practical-session-3.html#cb150-9" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;grey95&quot;</span>, <span class="st">&quot;red&quot;</span>)) <span class="sc">+</span> </span>
<span id="cb150-10"><a href="practical-session-3.html#cb150-10" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">&quot;Discarded&quot;</span>) <span class="sc">+</span> </span>
<span id="cb150-11"><a href="practical-session-3.html#cb150-11" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/03_QC_sfe5-2.png" width="672" /></p>
</div>
<div id="remove-low-quality-spots-1" class="section level3 hasAnchor" number="3.3.6">
<h3><span class="header-section-number">3.3.6</span> Remove low-quality spots<a href="practical-session-3.html#remove-low-quality-spots-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="practical-session-3.html#cb151-1" tabindex="-1"></a><span class="co"># ----------------------------------------------- #</span></span>
<span id="cb151-2"><a href="practical-session-3.html#cb151-2" tabindex="-1"></a><span class="do">## Check the number of discarded spots for each metric</span></span>
<span id="cb151-3"><a href="practical-session-3.html#cb151-3" tabindex="-1"></a><span class="fu">apply</span>(<span class="fu">cbind</span>(qc_lib_size, qc_detected, qc_mito), <span class="dv">2</span>, sum)</span></code></pre></div>
<pre><code>## qc_lib_size qc_detected     qc_mito 
##          19          20           5</code></pre>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="practical-session-3.html#cb153-1" tabindex="-1"></a><span class="do">## Combine together the set of discarded spots</span></span>
<span id="cb153-2"><a href="practical-session-3.html#cb153-2" tabindex="-1"></a>discard <span class="ot">&lt;-</span> qc_lib_size <span class="sc">|</span> qc_detected <span class="sc">|</span> qc_mito</span>
<span id="cb153-3"><a href="practical-session-3.html#cb153-3" tabindex="-1"></a><span class="fu">table</span>(discard)</span></code></pre></div>
<pre><code>## discard
## FALSE  TRUE 
##  1161    24</code></pre>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="practical-session-3.html#cb155-1" tabindex="-1"></a><span class="do">## Store the set in the object</span></span>
<span id="cb155-2"><a href="practical-session-3.html#cb155-2" tabindex="-1"></a><span class="fu">colData</span>(sfe)<span class="sc">$</span>discard <span class="ot">&lt;-</span> discard</span>
<span id="cb155-3"><a href="practical-session-3.html#cb155-3" tabindex="-1"></a><span class="do">## Check for putative spatial pattern of removed spots</span></span>
<span id="cb155-4"><a href="practical-session-3.html#cb155-4" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb155-5"><a href="practical-session-3.html#cb155-5" tabindex="-1"></a>    <span class="fu">geom_sf</span>(<span class="at">data =</span> <span class="fu">colGeometry</span>(sfe, <span class="st">&quot;spotHex&quot;</span>),</span>
<span id="cb155-6"><a href="practical-session-3.html#cb155-6" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">geometry =</span> geometry)) <span class="sc">+</span> </span>
<span id="cb155-7"><a href="practical-session-3.html#cb155-7" tabindex="-1"></a>    <span class="fu">geom_sf</span>(<span class="at">data =</span> <span class="fu">colGeometry</span>(sfe, <span class="st">&quot;spotHex&quot;</span>),</span>
<span id="cb155-8"><a href="practical-session-3.html#cb155-8" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">geometry =</span> geometry, <span class="at">fill =</span> <span class="fu">colData</span>(sfe)<span class="sc">$</span>discard)) <span class="sc">+</span></span>
<span id="cb155-9"><a href="practical-session-3.html#cb155-9" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;grey95&quot;</span>, <span class="st">&quot;red&quot;</span>)) <span class="sc">+</span> </span>
<span id="cb155-10"><a href="practical-session-3.html#cb155-10" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">&quot;Discarded&quot;</span>) <span class="sc">+</span> </span>
<span id="cb155-11"><a href="practical-session-3.html#cb155-11" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/03_QC_sfe6-1.png" width="672" /></p>
<div class="sourceCode" id="cb156"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb156-1"><a href="practical-session-3.html#cb156-1" tabindex="-1"></a><span class="co"># ----------------------------------------------- #</span></span>
<span id="cb156-2"><a href="practical-session-3.html#cb156-2" tabindex="-1"></a><span class="do">## remove combined set of low-quality spots</span></span>
<span id="cb156-3"><a href="practical-session-3.html#cb156-3" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> sfe[, <span class="sc">!</span><span class="fu">colData</span>(sfe)<span class="sc">$</span>discard]</span></code></pre></div>
</div>
</div>
<div id="normalisation-of-counts-1" class="section level2 hasAnchor" number="3.4">
<h2><span class="header-section-number">3.4</span> Normalisation of counts<a href="practical-session-3.html#normalisation-of-counts-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="log-tranformation-of-counts-1" class="section level3 hasAnchor" number="3.4.1">
<h3><span class="header-section-number">3.4.1</span> Log-tranformation of counts<a href="practical-session-3.html#log-tranformation-of-counts-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="practical-session-3.html#cb157-1" tabindex="-1"></a><span class="do">## Calculate library size factors</span></span>
<span id="cb157-2"><a href="practical-session-3.html#cb157-2" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> <span class="fu">computeLibraryFactors</span>(sfe)</span>
<span id="cb157-3"><a href="practical-session-3.html#cb157-3" tabindex="-1"></a><span class="do">## Have a look at the size factors</span></span>
<span id="cb157-4"><a href="practical-session-3.html#cb157-4" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">sizeFactors</span>(sfe))</span></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.07961 0.36902 0.95469 1.00000 1.54936 2.77256</code></pre>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="practical-session-3.html#cb159-1" tabindex="-1"></a><span class="do">## Density and histogram of library sizes</span></span>
<span id="cb159-2"><a href="practical-session-3.html#cb159-2" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">data.frame</span>(<span class="at">sFact =</span> <span class="fu">sizeFactors</span>(sfe)), </span>
<span id="cb159-3"><a href="practical-session-3.html#cb159-3" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> sFact)) <span class="sc">+</span></span>
<span id="cb159-4"><a href="practical-session-3.html#cb159-4" tabindex="-1"></a>    <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="fu">after_stat</span>(density)), </span>
<span id="cb159-5"><a href="practical-session-3.html#cb159-5" tabindex="-1"></a>                   <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>, </span>
<span id="cb159-6"><a href="practical-session-3.html#cb159-6" tabindex="-1"></a>                   <span class="at">fill =</span> <span class="st">&quot;grey&quot;</span>,</span>
<span id="cb159-7"><a href="practical-session-3.html#cb159-7" tabindex="-1"></a>                   <span class="at">bins =</span> <span class="dv">40</span>) <span class="sc">+</span></span>
<span id="cb159-8"><a href="practical-session-3.html#cb159-8" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>,</span>
<span id="cb159-9"><a href="practical-session-3.html#cb159-9" tabindex="-1"></a>                 <span class="at">adjust =</span> <span class="fl">0.5</span>,</span>
<span id="cb159-10"><a href="practical-session-3.html#cb159-10" tabindex="-1"></a>                 <span class="at">fill =</span> <span class="st">&quot;#A0CBE8&quot;</span>,</span>
<span id="cb159-11"><a href="practical-session-3.html#cb159-11" tabindex="-1"></a>                 <span class="at">colour =</span> <span class="st">&quot;#4E79A7&quot;</span>) <span class="sc">+</span></span>
<span id="cb159-12"><a href="practical-session-3.html#cb159-12" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb159-13"><a href="practical-session-3.html#cb159-13" tabindex="-1"></a>    <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">pretty_breaks</span>(<span class="at">n =</span> <span class="dv">10</span>)) <span class="sc">+</span> </span>
<span id="cb159-14"><a href="practical-session-3.html#cb159-14" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">&quot;Library size&quot;</span>) <span class="sc">+</span> </span>
<span id="cb159-15"><a href="practical-session-3.html#cb159-15" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>) <span class="sc">+</span> </span>
<span id="cb159-16"><a href="practical-session-3.html#cb159-16" tabindex="-1"></a>    <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/03_LogNorm_sfe-1.png" width="672" /></p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="practical-session-3.html#cb160-1" tabindex="-1"></a><span class="co"># calculate logcounts using library size factors</span></span>
<span id="cb160-2"><a href="practical-session-3.html#cb160-2" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(sfe)</span></code></pre></div>
</div>
</div>
<div id="gene-level-quality-control" class="section level2 hasAnchor" number="3.5">
<h2><span class="header-section-number">3.5</span> Gene-level Quality Control<a href="practical-session-3.html#gene-level-quality-control" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="calculating-extra-qc-metrics" class="section level3 hasAnchor" number="3.5.1">
<h3><span class="header-section-number">3.5.1</span> Calculating extra QC metrics<a href="practical-session-3.html#calculating-extra-qc-metrics" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="practical-session-3.html#cb161-1" tabindex="-1"></a><span class="fu">rowData</span>(sfe)[[<span class="st">&quot;JBO019.s_logMean&quot;</span>]] <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">assay</span>(sfe, <span class="st">&quot;logcounts&quot;</span>)) <span class="sc">/</span> <span class="fu">rowData</span>(sfe)[[<span class="st">&quot;JBO019.nLocations&quot;</span>]]</span></code></pre></div>
</div>
<div id="set-and-apply-filters" class="section level3 hasAnchor" number="3.5.2">
<h3><span class="header-section-number">3.5.2</span> Set and apply filters<a href="practical-session-3.html#set-and-apply-filters" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="practical-session-3.html#cb162-1" tabindex="-1"></a>is_zero <span class="ot">&lt;-</span> <span class="fu">rowData</span>(sfe)<span class="sc">$</span>total <span class="sc">==</span> <span class="dv">0</span></span>
<span id="cb162-2"><a href="practical-session-3.html#cb162-2" tabindex="-1"></a>is_logLow <span class="ot">&lt;-</span> <span class="fu">rowData</span>(sfe)[[<span class="st">&quot;JBO019.s_logMean&quot;</span>]] <span class="sc">&lt;=</span> <span class="dv">1</span></span>
<span id="cb162-3"><a href="practical-session-3.html#cb162-3" tabindex="-1"></a>discard_gs <span class="ot">&lt;-</span> is_zero <span class="sc">|</span> is_mito <span class="sc">|</span> is_logLow</span>
<span id="cb162-4"><a href="practical-session-3.html#cb162-4" tabindex="-1"></a><span class="fu">table</span>(discard_gs)</span></code></pre></div>
<pre><code>## discard_gs
## FALSE  TRUE 
##  8535 24203</code></pre>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="practical-session-3.html#cb164-1" tabindex="-1"></a><span class="fu">rowData</span>(sfe)<span class="sc">$</span>discard <span class="ot">&lt;-</span> discard_gs</span>
<span id="cb164-2"><a href="practical-session-3.html#cb164-2" tabindex="-1"></a></span>
<span id="cb164-3"><a href="practical-session-3.html#cb164-3" tabindex="-1"></a><span class="do">## FEATURE SELECTION</span></span>
<span id="cb164-4"><a href="practical-session-3.html#cb164-4" tabindex="-1"></a><span class="do">## remove mitochondrial and other genes</span></span>
<span id="cb164-5"><a href="practical-session-3.html#cb164-5" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> sfe[<span class="sc">!</span><span class="fu">rowData</span>(sfe)<span class="sc">$</span>discard, ]</span></code></pre></div>
</div>
</div>
<div id="selecting-genes-1" class="section level2 hasAnchor" number="3.6">
<h2><span class="header-section-number">3.6</span> Selecting genes<a href="practical-session-3.html#selecting-genes-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="highly-variable-genes-hvgs-1" class="section level3 hasAnchor" number="3.6.1">
<h3><span class="header-section-number">3.6.1</span> Highly Variable Genes (HVGs)<a href="practical-session-3.html#highly-variable-genes-hvgs-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="practical-session-3.html#cb165-1" tabindex="-1"></a><span class="do">## Fit mean-variance relationship</span></span>
<span id="cb165-2"><a href="practical-session-3.html#cb165-2" tabindex="-1"></a>dec <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(sfe,</span>
<span id="cb165-3"><a href="practical-session-3.html#cb165-3" tabindex="-1"></a>                    <span class="at">assay.type =</span> <span class="st">&quot;logcounts&quot;</span>)</span>
<span id="cb165-4"><a href="practical-session-3.html#cb165-4" tabindex="-1"></a></span>
<span id="cb165-5"><a href="practical-session-3.html#cb165-5" tabindex="-1"></a><span class="do">## Visualize mean-variance relationship</span></span>
<span id="cb165-6"><a href="practical-session-3.html#cb165-6" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">metadata</span>(dec)</span>
<span id="cb165-7"><a href="practical-session-3.html#cb165-7" tabindex="-1"></a>fit_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">mean =</span> fit<span class="sc">$</span>mean,</span>
<span id="cb165-8"><a href="practical-session-3.html#cb165-8" tabindex="-1"></a>                     <span class="at">var =</span> fit<span class="sc">$</span>var,</span>
<span id="cb165-9"><a href="practical-session-3.html#cb165-9" tabindex="-1"></a>                     <span class="at">trend =</span> fit<span class="sc">$</span><span class="fu">trend</span>(fit<span class="sc">$</span>mean))</span>
<span id="cb165-10"><a href="practical-session-3.html#cb165-10" tabindex="-1"></a></span>
<span id="cb165-11"><a href="practical-session-3.html#cb165-11" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> fit_df, </span>
<span id="cb165-12"><a href="practical-session-3.html#cb165-12" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> mean, <span class="at">y =</span> var)) <span class="sc">+</span> </span>
<span id="cb165-13"><a href="practical-session-3.html#cb165-13" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb165-14"><a href="practical-session-3.html#cb165-14" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> trend), <span class="at">colour =</span> <span class="st">&quot;dodgerblue&quot;</span>, <span class="at">linewidth =</span> <span class="fl">1.5</span>) <span class="sc">+</span> </span>
<span id="cb165-15"><a href="practical-session-3.html#cb165-15" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;mean of log-expression&quot;</span>,</span>
<span id="cb165-16"><a href="practical-session-3.html#cb165-16" tabindex="-1"></a>         <span class="at">y =</span> <span class="st">&quot;variance of log-expression&quot;</span>) <span class="sc">+</span> </span>
<span id="cb165-17"><a href="practical-session-3.html#cb165-17" tabindex="-1"></a>    <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/03_HVGs_sfe-1.png" width="672" /></p>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="practical-session-3.html#cb166-1" tabindex="-1"></a><span class="do">## Select top HVGs</span></span>
<span id="cb166-2"><a href="practical-session-3.html#cb166-2" tabindex="-1"></a>top_hvgs <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(dec, </span>
<span id="cb166-3"><a href="practical-session-3.html#cb166-3" tabindex="-1"></a>                       <span class="at">var.field =</span> <span class="st">&quot;bio&quot;</span>, </span>
<span id="cb166-4"><a href="practical-session-3.html#cb166-4" tabindex="-1"></a>                       <span class="at">prop =</span> <span class="fl">0.5</span>,</span>
<span id="cb166-5"><a href="practical-session-3.html#cb166-5" tabindex="-1"></a>                       <span class="at">var.threshold =</span> <span class="dv">0</span>,</span>
<span id="cb166-6"><a href="practical-session-3.html#cb166-6" tabindex="-1"></a>                       <span class="at">fdr.threshold =</span> <span class="fl">0.1</span>)</span></code></pre></div>
</div>
</div>
<div id="neighbour-graph-and-distance-matrix" class="section level2 hasAnchor" number="3.7">
<h2><span class="header-section-number">3.7</span> Neighbour graph and distance matrix<a href="practical-session-3.html#neighbour-graph-and-distance-matrix" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="adding-spatial-weights" class="section level3 hasAnchor" number="3.7.1">
<h3><span class="header-section-number">3.7.1</span> Adding spatial weights<a href="practical-session-3.html#adding-spatial-weights" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The neighbour lists can be supplemented with spatial weights using the <code>nb2listw</code> and <code>nb2listwdist</code> function from <code>spdep</code> package for the chosen type and coding scheme style. There are 6 different coding scheme styles that can be used to weigh neighbour relationships:</p>
<ol style="list-style-type: decimal">
<li><strong>B</strong>: is the basic binary coding (1 for neighbour, 0 for no neighbour).</li>
<li><strong>W</strong>: is row standardised (sums over all links to n).</li>
<li><strong>C</strong>: is globally standardised (sums over all links to n).</li>
<li><strong>U</strong>: is equal to C divided by the number of neighbours (sums over all links to unity).</li>
<li><strong>S</strong>: is the variance-stabilizing coding scheme (sums over all links to n).</li>
<li><strong>minmax</strong>: divides the weights by the minimum of the maximum row sums and maximum column sums of the input weights; It is similar to the C and U styles.</li>
</ol>
<p>The coding scheme style is practically the value each neighbour will get. For example, in a binary coding scheme style (<strong>B</strong>) if a spot is a neighbour of the spot in focus then gets the value of <strong>1</strong>, else gets <strong>0</strong>. Another example, in a row standardised coding scheme style (<strong>W</strong>) if the spot in focus has a total of 10 neighbours and each neighbour has a weight of 1, then the sum of all neighbour weights is 10, and each neighbour will get a normalised weight of 1/10 = 0.1. As a result, in the row standardised coding scheme, spots with many neighbours will have neighbours with lower weights and thus will not be over-emphasised.</p>
<p>Starting from a binary neighbours list, in which regions are either listed as neighbours or are absent (thus not in the set of neighbours for some definition), we can add a distance-based weights list. The <code>nb2listwdist</code> function supplements a neighbours list with spatial weights for the chosen types of distance modelling and coding scheme. While the offered coding schemes parallel those of the <code>nb2listw</code> function above, three distance-based types of weights are available: inverse distance weighting (IDW), double-power distance weights (DPD), and exponential distance decay (EXP). The three types of distance weight calculations are based on pairwise distances 𝑑𝑖𝑗, all of which are controlled by parameter <em>“alpha”</em> (𝛼 below):</p>
<ol style="list-style-type: decimal">
<li><strong>idw</strong>: 𝑤𝑖𝑗=𝑑−𝛼𝑖𝑗,</li>
<li><strong>exp</strong>: 𝑤𝑖𝑗=exp(−𝛼⋅𝑑𝑖𝑗),</li>
<li><strong>dpd</strong>: 𝑤𝑖𝑗=[1−(𝑑𝑖𝑗/𝑑max)𝛼]𝛼,</li>
</ol>
<p>the latter of which leads to 𝑤𝑖𝑗=0 for all 𝑑𝑖𝑗&gt;𝑑max. Note that <em>IDW</em> weights show extreme behaviour close to 0 and can take on the value infinity. In such cases, the infinite values are replaced by the largest finite weight present in the weights list.</p>
</div>
<div id="generate-distance-matrices" class="section level3 hasAnchor" number="3.7.2">
<h3><span class="header-section-number">3.7.2</span> Generate distance matrices<a href="practical-session-3.html#generate-distance-matrices" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>A distance matrix is a mirrored matrix that contains the distance between a spot and every other spot. This distance can be a simple Euclidean distance based on the coordinates of the spots or a weighted distance according to a bandwidth around each spot using a kernel that gives higher scores to distances between spots that are closer together compared to the ones that are farther away. These weighted distance matrices are later used to run geographically weighted (GW) models.</p>
<p>There are 6 different kernels that can be used to weight the distances between spots. The next two figures are from the <code>GWmodel</code>’s publication <span class="citation">(<a href="#ref-Gollini2015Feb">Gollini et al. 2015</a>)</span> and provide a bit more description on that.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:GWmodelFig1"></span>
<img src="images/gwmodel_kernel_math.png" alt="The math equations that define the kernels." width="100%" />
<p class="caption">
Figure 3.2: The math equations that define the kernels.
</p>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:GWmodelFig2"></span>
<img src="images/gwmodel_kernel_graphs.png" alt="Examples from using each kernel." width="100%" />
<p class="caption">
Figure 3.3: Examples from using each kernel.
</p>
</div>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="practical-session-3.html#cb167-1" tabindex="-1"></a><span class="do">## add a neighbour graph using a weighted distance matrix</span></span>
<span id="cb167-2"><a href="practical-session-3.html#cb167-2" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> <span class="fu">addSpatialNeighGraphs</span>(sfe, <span class="st">&quot;JBO019&quot;</span>, <span class="at">type =</span> <span class="st">&quot;knearneigh&quot;</span>, <span class="at">style =</span> <span class="st">&quot;W&quot;</span>, <span class="at">distMod =</span> <span class="st">&quot;raw&quot;</span>, <span class="at">k =</span> <span class="dv">6</span>)</span>
<span id="cb167-3"><a href="practical-session-3.html#cb167-3" tabindex="-1"></a></span>
<span id="cb167-4"><a href="practical-session-3.html#cb167-4" tabindex="-1"></a><span class="fu">colGraphs</span>(sfe)</span></code></pre></div>
<pre><code>## $col
## Characteristics of weights list object:
## Neighbour list object:
## Number of regions: 1161 
## Number of nonzero links: 6966 
## Percentage nonzero weights: 0.5167959 
## Average number of links: 6 
## Non-symmetric neighbours list
## 
## Weights style: W 
## Weights constants summary:
##      n      nn   S0       S1       S2
## W 1161 1347921 1161 376.8333 4674.667</code></pre>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="practical-session-3.html#cb169-1" tabindex="-1"></a><span class="do">## Calculate a simple distance matrix</span></span>
<span id="cb169-2"><a href="practical-session-3.html#cb169-2" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> <span class="fu">addDistMat</span>(sfe, <span class="at">p =</span> <span class="dv">2</span>)</span></code></pre></div>
</div>
</div>
<div id="putting-it-all-together-1" class="section level2 hasAnchor" number="3.8">
<h2><span class="header-section-number">3.8</span> Putting it all together<a href="practical-session-3.html#putting-it-all-together-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The below code puts all these steps in order by selecting one of the options at each step.</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="practical-session-3.html#cb170-1" tabindex="-1"></a><span class="do">## Import data</span></span>
<span id="cb170-2"><a href="practical-session-3.html#cb170-2" tabindex="-1"></a>sampleDir <span class="ot">&lt;-</span> <span class="st">&quot;./data/spaceranger_outs/Human_Liver_Steatotic/JBO019_Results&quot;</span></span>
<span id="cb170-3"><a href="practical-session-3.html#cb170-3" tabindex="-1"></a>sampleNames <span class="ot">&lt;-</span> <span class="st">&quot;JBO019&quot;</span></span>
<span id="cb170-4"><a href="practical-session-3.html#cb170-4" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> <span class="fu">read10xVisiumSFE</span>(<span class="at">samples =</span> sampleDir, </span>
<span id="cb170-5"><a href="practical-session-3.html#cb170-5" tabindex="-1"></a>                        <span class="at">sample_id =</span> sampleNames, </span>
<span id="cb170-6"><a href="practical-session-3.html#cb170-6" tabindex="-1"></a>                        <span class="at">type =</span> <span class="st">&quot;sparse&quot;</span>, </span>
<span id="cb170-7"><a href="practical-session-3.html#cb170-7" tabindex="-1"></a>                        <span class="at">data =</span> <span class="st">&quot;filtered&quot;</span>, </span>
<span id="cb170-8"><a href="practical-session-3.html#cb170-8" tabindex="-1"></a>                        <span class="at">images =</span> <span class="st">&quot;lowres&quot;</span>, </span>
<span id="cb170-9"><a href="practical-session-3.html#cb170-9" tabindex="-1"></a>                        <span class="at">style =</span> <span class="st">&quot;W&quot;</span>, </span>
<span id="cb170-10"><a href="practical-session-3.html#cb170-10" tabindex="-1"></a>                        <span class="at">zero.policy =</span> <span class="cn">TRUE</span>)</span>
<span id="cb170-11"><a href="practical-session-3.html#cb170-11" tabindex="-1"></a><span class="co"># ----------------------------------------------- #</span></span>
<span id="cb170-12"><a href="practical-session-3.html#cb170-12" tabindex="-1"></a>ground_truth <span class="ot">&lt;-</span> <span class="fu">read_table</span>(<span class="st">&quot;./data/to_load/spotzonationGroup.txt&quot;</span>)</span>
<span id="cb170-13"><a href="practical-session-3.html#cb170-13" tabindex="-1"></a><span class="do">## Add QC metrics</span></span>
<span id="cb170-14"><a href="practical-session-3.html#cb170-14" tabindex="-1"></a>is_mito <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">&quot;(^MT-)|(^mt-)&quot;</span>, <span class="fu">rowData</span>(sfe)<span class="sc">$</span>symbol)</span>
<span id="cb170-15"><a href="practical-session-3.html#cb170-15" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> <span class="fu">addPerLocQC</span>(sfe, <span class="at">gTruth =</span> ground_truth, <span class="at">assay =</span> <span class="st">&quot;counts&quot;</span>, <span class="dv">2</span>, <span class="at">subsets =</span> <span class="fu">list</span>(<span class="at">mito =</span> is_mito))</span>
<span id="cb170-16"><a href="practical-session-3.html#cb170-16" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> <span class="fu">addGeometries</span>(sfe, <span class="at">samples =</span> sampleDir, <span class="at">sample_id =</span> sampleNames, <span class="at">res =</span> <span class="st">&quot;fullres&quot;</span>)</span>
<span id="cb170-17"><a href="practical-session-3.html#cb170-17" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> <span class="fu">addPerGeneQC</span>(sfe, <span class="at">assay =</span> <span class="st">&quot;counts&quot;</span>, <span class="at">version =</span> <span class="cn">NULL</span>, <span class="at">mirror =</span> <span class="cn">NULL</span>)</span>
<span id="cb170-18"><a href="practical-session-3.html#cb170-18" tabindex="-1"></a><span class="co"># ----------------------------------------------- #</span></span>
<span id="cb170-19"><a href="practical-session-3.html#cb170-19" tabindex="-1"></a><span class="do">## SPOT SELECTION</span></span>
<span id="cb170-20"><a href="practical-session-3.html#cb170-20" tabindex="-1"></a><span class="do">## Select library size threshold</span></span>
<span id="cb170-21"><a href="practical-session-3.html#cb170-21" tabindex="-1"></a>qc_lib_size <span class="ot">&lt;-</span> <span class="fu">colData</span>(sfe)<span class="sc">$</span>sum <span class="sc">&lt;</span> <span class="dv">1000</span></span>
<span id="cb170-22"><a href="practical-session-3.html#cb170-22" tabindex="-1"></a><span class="do">## Add threshold in colData</span></span>
<span id="cb170-23"><a href="practical-session-3.html#cb170-23" tabindex="-1"></a><span class="fu">colData</span>(sfe)<span class="sc">$</span>qc_lib_size <span class="ot">&lt;-</span> qc_lib_size</span>
<span id="cb170-24"><a href="practical-session-3.html#cb170-24" tabindex="-1"></a><span class="do">## Select expressed genes threshold</span></span>
<span id="cb170-25"><a href="practical-session-3.html#cb170-25" tabindex="-1"></a>qc_detected <span class="ot">&lt;-</span> <span class="fu">colData</span>(sfe)<span class="sc">$</span>detected <span class="sc">&lt;</span> <span class="dv">550</span></span>
<span id="cb170-26"><a href="practical-session-3.html#cb170-26" tabindex="-1"></a><span class="do">## Add threshold in colData</span></span>
<span id="cb170-27"><a href="practical-session-3.html#cb170-27" tabindex="-1"></a><span class="fu">colData</span>(sfe)<span class="sc">$</span>qc_detected <span class="ot">&lt;-</span> qc_detected</span>
<span id="cb170-28"><a href="practical-session-3.html#cb170-28" tabindex="-1"></a><span class="do">## Select mitochondrial percentage threshold</span></span>
<span id="cb170-29"><a href="practical-session-3.html#cb170-29" tabindex="-1"></a>qc_mito <span class="ot">&lt;-</span> <span class="fu">colData</span>(sfe)<span class="sc">$</span>subsets_mito_percent <span class="sc">&gt;</span> <span class="dv">22</span></span>
<span id="cb170-30"><a href="practical-session-3.html#cb170-30" tabindex="-1"></a><span class="do">## Add threshold in colData</span></span>
<span id="cb170-31"><a href="practical-session-3.html#cb170-31" tabindex="-1"></a><span class="fu">colData</span>(sfe)<span class="sc">$</span>qc_mito <span class="ot">&lt;-</span> qc_mito</span>
<span id="cb170-32"><a href="practical-session-3.html#cb170-32" tabindex="-1"></a><span class="do">## Combine together the set of discarded spots</span></span>
<span id="cb170-33"><a href="practical-session-3.html#cb170-33" tabindex="-1"></a>discard <span class="ot">&lt;-</span> qc_lib_size <span class="sc">|</span> qc_detected <span class="sc">|</span> qc_mito</span>
<span id="cb170-34"><a href="practical-session-3.html#cb170-34" tabindex="-1"></a><span class="do">## Store the set in the object</span></span>
<span id="cb170-35"><a href="practical-session-3.html#cb170-35" tabindex="-1"></a><span class="fu">colData</span>(sfe)<span class="sc">$</span>discard <span class="ot">&lt;-</span> discard</span>
<span id="cb170-36"><a href="practical-session-3.html#cb170-36" tabindex="-1"></a><span class="do">## Remove combined set of low-quality spots</span></span>
<span id="cb170-37"><a href="practical-session-3.html#cb170-37" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> sfe[, <span class="sc">!</span><span class="fu">colData</span>(sfe)<span class="sc">$</span>discard]</span>
<span id="cb170-38"><a href="practical-session-3.html#cb170-38" tabindex="-1"></a><span class="co"># ----------------------------------------------- #</span></span>
<span id="cb170-39"><a href="practical-session-3.html#cb170-39" tabindex="-1"></a><span class="do">## FEATURE SELECTION</span></span>
<span id="cb170-40"><a href="practical-session-3.html#cb170-40" tabindex="-1"></a><span class="do">## Calculate library size factors</span></span>
<span id="cb170-41"><a href="practical-session-3.html#cb170-41" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> <span class="fu">computeLibraryFactors</span>(sfe)</span>
<span id="cb170-42"><a href="practical-session-3.html#cb170-42" tabindex="-1"></a><span class="do">## Calculate logcounts using library size factors</span></span>
<span id="cb170-43"><a href="practical-session-3.html#cb170-43" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(sfe)</span>
<span id="cb170-44"><a href="practical-session-3.html#cb170-44" tabindex="-1"></a><span class="do">## Calculate log-counts sample mean</span></span>
<span id="cb170-45"><a href="practical-session-3.html#cb170-45" tabindex="-1"></a><span class="fu">rowData</span>(sfe)[[<span class="st">&quot;JBO019.s_logMean&quot;</span>]] <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">assay</span>(sfe, <span class="st">&quot;logcounts&quot;</span>)) <span class="sc">/</span> <span class="fu">rowData</span>(sfe)[[<span class="st">&quot;JBO019.nLocations&quot;</span>]]</span>
<span id="cb170-46"><a href="practical-session-3.html#cb170-46" tabindex="-1"></a><span class="do">## Set and apply filters</span></span>
<span id="cb170-47"><a href="practical-session-3.html#cb170-47" tabindex="-1"></a>is_zero <span class="ot">&lt;-</span> <span class="fu">rowData</span>(sfe)<span class="sc">$</span>total <span class="sc">==</span> <span class="dv">0</span></span>
<span id="cb170-48"><a href="practical-session-3.html#cb170-48" tabindex="-1"></a>is_logLow <span class="ot">&lt;-</span> <span class="fu">rowData</span>(sfe)[[<span class="st">&quot;JBO019.s_logMean&quot;</span>]] <span class="sc">&lt;=</span> <span class="dv">1</span></span>
<span id="cb170-49"><a href="practical-session-3.html#cb170-49" tabindex="-1"></a>discard_gs <span class="ot">&lt;-</span> is_zero <span class="sc">|</span> is_mito <span class="sc">|</span> is_logLow</span>
<span id="cb170-50"><a href="practical-session-3.html#cb170-50" tabindex="-1"></a><span class="fu">rowData</span>(sfe)<span class="sc">$</span>discard <span class="ot">&lt;-</span> discard_gs</span>
<span id="cb170-51"><a href="practical-session-3.html#cb170-51" tabindex="-1"></a><span class="do">## Remove mitochondrial and other genes</span></span>
<span id="cb170-52"><a href="practical-session-3.html#cb170-52" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> sfe[<span class="sc">!</span><span class="fu">rowData</span>(sfe)<span class="sc">$</span>discard, ]</span>
<span id="cb170-53"><a href="practical-session-3.html#cb170-53" tabindex="-1"></a><span class="do">## Fit mean-variance relationship</span></span>
<span id="cb170-54"><a href="practical-session-3.html#cb170-54" tabindex="-1"></a>dec <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(sfe,</span>
<span id="cb170-55"><a href="practical-session-3.html#cb170-55" tabindex="-1"></a>                    <span class="at">assay.type =</span> <span class="st">&quot;logcounts&quot;</span>)</span>
<span id="cb170-56"><a href="practical-session-3.html#cb170-56" tabindex="-1"></a><span class="do">## Select top HVGs</span></span>
<span id="cb170-57"><a href="practical-session-3.html#cb170-57" tabindex="-1"></a>top_hvgs <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(dec, </span>
<span id="cb170-58"><a href="practical-session-3.html#cb170-58" tabindex="-1"></a>                       <span class="at">var.field =</span> <span class="st">&quot;bio&quot;</span>, </span>
<span id="cb170-59"><a href="practical-session-3.html#cb170-59" tabindex="-1"></a>                       <span class="at">prop =</span> <span class="fl">0.5</span>,</span>
<span id="cb170-60"><a href="practical-session-3.html#cb170-60" tabindex="-1"></a>                       <span class="at">var.threshold =</span> <span class="dv">0</span>,</span>
<span id="cb170-61"><a href="practical-session-3.html#cb170-61" tabindex="-1"></a>                       <span class="at">fdr.threshold =</span> <span class="fl">0.05</span>)</span>
<span id="cb170-62"><a href="practical-session-3.html#cb170-62" tabindex="-1"></a><span class="co"># ----------------------------------------------- #</span></span>
<span id="cb170-63"><a href="practical-session-3.html#cb170-63" tabindex="-1"></a><span class="do">## ADD GEOGRAPHY</span></span>
<span id="cb170-64"><a href="practical-session-3.html#cb170-64" tabindex="-1"></a><span class="do">## Add a neighbour graph using a weighted distance matrix</span></span>
<span id="cb170-65"><a href="practical-session-3.html#cb170-65" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> <span class="fu">addSpatialNeighGraphs</span>(sfe, <span class="st">&quot;JBO019&quot;</span>, <span class="at">type =</span> <span class="st">&quot;knearneigh&quot;</span>, <span class="at">style =</span> <span class="st">&quot;W&quot;</span>, <span class="at">distMod =</span> <span class="st">&quot;raw&quot;</span>, <span class="at">k =</span> <span class="dv">6</span>)</span>
<span id="cb170-66"><a href="practical-session-3.html#cb170-66" tabindex="-1"></a><span class="do">## Calculate a simple distance matrix</span></span>
<span id="cb170-67"><a href="practical-session-3.html#cb170-67" tabindex="-1"></a>sfe <span class="ot">&lt;-</span> <span class="fu">addDistMat</span>(sfe, <span class="at">p =</span> <span class="dv">2</span>)</span></code></pre></div>
</div>
<div id="save-data" class="section level2 hasAnchor" number="3.9">
<h2><span class="header-section-number">3.9</span> Save data<a href="practical-session-3.html#save-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Gollini2015Feb" class="csl-entry">
Gollini, Isabella, Binbin Lu, Martin Charlton, Christopher Brunsdon, and Paul Harris. 2015. <span>“<span class="nocase">GWmodel: An R Package for Exploring Spatial Heterogeneity Using Geographically Weighted Models</span>.”</span> <em>J Stat Soft</em> 63 (February): 1–50. <a href="https://doi.org/10.18637/jss.v063.i17">https://doi.org/10.18637/jss.v063.i17</a>.
</div>
<div id="ref-GUILLIAMS2022379" class="csl-entry">
Guilliams, Martin, Johnny Bonnardel, Birthe Haest, Bart Vanderborght, Camille Wagner, Anneleen Remmerie, Anna Bujko, et al. 2022. <span>“Spatial Proteogenomics Reveals Distinct and Evolutionarily Conserved Hepatic Macrophage Niches.”</span> <em>Cell</em> 185 (2): 379–396.e38. https://doi.org/<a href="https://doi.org/10.1016/j.cell.2021.12.018">https://doi.org/10.1016/j.cell.2021.12.018</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="practical-session-2.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="practical-session-4.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/03-intro-geospatial.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
