<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>References | Spatial transcriptomics data analysis: theory and practice</title>
  <meta name="description" content="This book will guide you through the practical steps of the in-person tutorial IP2 for the ISMB/ECCB 2023 conference in Lyon named: Spatial transcriptomics data analysis: theory and practice." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="References | Spatial transcriptomics data analysis: theory and practice" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This book will guide you through the practical steps of the in-person tutorial IP2 for the ISMB/ECCB 2023 conference in Lyon named: Spatial transcriptomics data analysis: theory and practice." />
  <meta name="github-repo" content="https://github.com/ncl-icbam/ismb-tutorial-2023.git" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="References | Spatial transcriptomics data analysis: theory and practice" />
  
  <meta name="twitter:description" content="This book will guide you through the practical steps of the in-person tutorial IP2 for the ISMB/ECCB 2023 conference in Lyon named: Spatial transcriptomics data analysis: theory and practice." />
  

<meta name="author" content="Eleftherios Zormpas, Dr Simon J. Cockell" />


<meta name="date" content="2023-06-09" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="practical-session-4.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Spatial transcriptomics data analysis: theory and practice</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#abstract"><i class="fa fa-check"></i>Abstract</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#learning-objectives"><i class="fa fa-check"></i>Learning objectives</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="practical-session-1.html"><a href="practical-session-1.html"><i class="fa fa-check"></i><b>1</b> Practical session 1</a>
<ul>
<li class="chapter" data-level="1.1" data-path="practical-session-1.html"><a href="practical-session-1.html#log-into-the-posit-cloud-server"><i class="fa fa-check"></i><b>1.1</b> Log into the Posit Cloud server</a></li>
<li class="chapter" data-level="1.2" data-path="practical-session-1.html"><a href="practical-session-1.html#import-10x-visium-data"><i class="fa fa-check"></i><b>1.2</b> Import 10X Visium data</a></li>
<li class="chapter" data-level="1.3" data-path="practical-session-1.html"><a href="practical-session-1.html#explore-data-types"><i class="fa fa-check"></i><b>1.3</b> Explore data types</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="practical-session-1.html"><a href="practical-session-1.html#spatialexperiment-class"><i class="fa fa-check"></i><b>1.3.1</b> SpatialExperiment class</a></li>
<li class="chapter" data-level="1.3.2" data-path="practical-session-1.html"><a href="practical-session-1.html#inspect-the-object"><i class="fa fa-check"></i><b>1.3.2</b> Inspect the object</a></li>
<li class="chapter" data-level="1.3.3" data-path="practical-session-1.html"><a href="practical-session-1.html#counts-table-and-gene-metadata"><i class="fa fa-check"></i><b>1.3.3</b> Counts table and gene metadata</a></li>
<li class="chapter" data-level="1.3.4" data-path="practical-session-1.html"><a href="practical-session-1.html#coordinates-table-and-spot-metadata"><i class="fa fa-check"></i><b>1.3.4</b> Coordinates table and spot metadata</a></li>
<li class="chapter" data-level="1.3.5" data-path="practical-session-1.html"><a href="practical-session-1.html#image-metadata"><i class="fa fa-check"></i><b>1.3.5</b> Image metadata</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="practical-session-2.html"><a href="practical-session-2.html"><i class="fa fa-check"></i><b>2</b> Practical session 2</a>
<ul>
<li class="chapter" data-level="2.1" data-path="practical-session-2.html"><a href="practical-session-2.html#spot-level-quality-control"><i class="fa fa-check"></i><b>2.1</b> Spot-level Quality Control</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="practical-session-2.html"><a href="practical-session-2.html#plot-tissue-map"><i class="fa fa-check"></i><b>2.1.1</b> Plot tissue map</a></li>
<li class="chapter" data-level="2.1.2" data-path="practical-session-2.html"><a href="practical-session-2.html#calculating-qc-metrics"><i class="fa fa-check"></i><b>2.1.2</b> Calculating QC metrics</a></li>
<li class="chapter" data-level="2.1.3" data-path="practical-session-2.html"><a href="practical-session-2.html#library-size-threshold-plot"><i class="fa fa-check"></i><b>2.1.3</b> Library size threshold plot</a></li>
<li class="chapter" data-level="2.1.4" data-path="practical-session-2.html"><a href="practical-session-2.html#number-of-expressed-genes"><i class="fa fa-check"></i><b>2.1.4</b> Number of expressed genes</a></li>
<li class="chapter" data-level="2.1.5" data-path="practical-session-2.html"><a href="practical-session-2.html#percentage-of-mitochondrial-expression"><i class="fa fa-check"></i><b>2.1.5</b> Percentage of mitochondrial expression</a></li>
<li class="chapter" data-level="2.1.6" data-path="practical-session-2.html"><a href="practical-session-2.html#number-of-cells-per-spot"><i class="fa fa-check"></i><b>2.1.6</b> Number of cells per spot</a></li>
<li class="chapter" data-level="2.1.7" data-path="practical-session-2.html"><a href="practical-session-2.html#remove-low-quality-spots"><i class="fa fa-check"></i><b>2.1.7</b> Remove low-quality spots</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="practical-session-2.html"><a href="practical-session-2.html#normalisation-of-counts"><i class="fa fa-check"></i><b>2.2</b> Normalisation of counts</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background"><i class="fa fa-check"></i><b>2.2.1</b> Background</a></li>
<li class="chapter" data-level="2.2.2" data-path="practical-session-2.html"><a href="practical-session-2.html#log-tranformation-of-counts"><i class="fa fa-check"></i><b>2.2.2</b> Log-tranformation of counts</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="practical-session-2.html"><a href="practical-session-2.html#selecting-genes"><i class="fa fa-check"></i><b>2.3</b> Selecting genes</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background-1"><i class="fa fa-check"></i><b>2.3.1</b> Background</a></li>
<li class="chapter" data-level="2.3.2" data-path="practical-session-2.html"><a href="practical-session-2.html#highly-variable-genes-hvgs"><i class="fa fa-check"></i><b>2.3.2</b> Highly Variable Genes (HVGs)</a></li>
<li class="chapter" data-level="2.3.3" data-path="practical-session-2.html"><a href="practical-session-2.html#spatially-variable-genes-svgs"><i class="fa fa-check"></i><b>2.3.3</b> Spatially variable genes (SVGs)</a></li>
<li class="chapter" data-level="2.3.4" data-path="practical-session-2.html"><a href="practical-session-2.html#integration-of-hvgs-and-svgs"><i class="fa fa-check"></i><b>2.3.4</b> Integration of HVGs and SVGs</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="practical-session-2.html"><a href="practical-session-2.html#dimensionality-reduction"><i class="fa fa-check"></i><b>2.4</b> Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background-2"><i class="fa fa-check"></i><b>2.4.1</b> Background</a></li>
<li class="chapter" data-level="2.4.2" data-path="practical-session-2.html"><a href="practical-session-2.html#pca-principal-component-analysis"><i class="fa fa-check"></i><b>2.4.2</b> PCA: Principal component analysis</a></li>
<li class="chapter" data-level="2.4.3" data-path="practical-session-2.html"><a href="practical-session-2.html#umap-uniform-manifold-approximation-and-projection"><i class="fa fa-check"></i><b>2.4.3</b> UMAP: Uniform Manifold Approximation and Projection</a></li>
<li class="chapter" data-level="2.4.4" data-path="practical-session-2.html"><a href="practical-session-2.html#umap-visualisations"><i class="fa fa-check"></i><b>2.4.4</b> UMAP visualisations</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="practical-session-2.html"><a href="practical-session-2.html#clustering"><i class="fa fa-check"></i><b>2.5</b> Clustering</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background-3"><i class="fa fa-check"></i><b>2.5.1</b> Background</a></li>
<li class="chapter" data-level="2.5.2" data-path="practical-session-2.html"><a href="practical-session-2.html#clustering-on-hvgs"><i class="fa fa-check"></i><b>2.5.2</b> Clustering on HVGs</a></li>
<li class="chapter" data-level="2.5.3" data-path="practical-session-2.html"><a href="practical-session-2.html#hvgs-clustering-visualisations"><i class="fa fa-check"></i><b>2.5.3</b> HVGs clustering visualisations</a></li>
<li class="chapter" data-level="2.5.4" data-path="practical-session-2.html"><a href="practical-session-2.html#spatially-aware-clustering"><i class="fa fa-check"></i><b>2.5.4</b> Spatially-aware clustering</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="practical-session-2.html"><a href="practical-session-2.html#inter-cluster-differentially-expressed-genes-dges"><i class="fa fa-check"></i><b>2.6</b> Inter-cluster differentially expressed genes (DGEs)</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="practical-session-2.html"><a href="practical-session-2.html#background-4"><i class="fa fa-check"></i><b>2.6.1</b> Background</a></li>
<li class="chapter" data-level="2.6.2" data-path="practical-session-2.html"><a href="practical-session-2.html#dges-identification"><i class="fa fa-check"></i><b>2.6.2</b> DGEs identification</a></li>
<li class="chapter" data-level="2.6.3" data-path="practical-session-2.html"><a href="practical-session-2.html#dges-visualisation"><i class="fa fa-check"></i><b>2.6.3</b> DGEs visualisation</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="practical-session-2.html"><a href="practical-session-2.html#other-visualisations"><i class="fa fa-check"></i><b>2.7</b> Other visualisations</a></li>
<li class="chapter" data-level="2.8" data-path="practical-session-2.html"><a href="practical-session-2.html#putting-it-all-together"><i class="fa fa-check"></i><b>2.8</b> Putting it all together</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="practical-session-3.html"><a href="practical-session-3.html"><i class="fa fa-check"></i><b>3</b> Practical session 3</a>
<ul>
<li class="chapter" data-level="3.1" data-path="practical-session-3.html"><a href="practical-session-3.html#background-5"><i class="fa fa-check"></i><b>3.1</b> Background</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="practical-session-3.html"><a href="practical-session-3.html#main-geocomputatinal-data-structures"><i class="fa fa-check"></i><b>3.1.1</b> Main geocomputatinal data structures</a></li>
<li class="chapter" data-level="3.1.2" data-path="practical-session-3.html"><a href="practical-session-3.html#the-sf-objects"><i class="fa fa-check"></i><b>3.1.2</b> The <code>sf</code> objects</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="practical-session-3.html"><a href="practical-session-3.html#data-structures-preparation"><i class="fa fa-check"></i><b>3.2</b> Data structures preparation</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="practical-session-3.html"><a href="practical-session-3.html#load-new-dataset"><i class="fa fa-check"></i><b>3.2.1</b> Load new dataset</a></li>
<li class="chapter" data-level="3.2.2" data-path="practical-session-3.html"><a href="practical-session-3.html#create-point-geometries"><i class="fa fa-check"></i><b>3.2.2</b> Create point geometries</a></li>
<li class="chapter" data-level="3.2.3" data-path="practical-session-3.html"><a href="practical-session-3.html#tesselate-space"><i class="fa fa-check"></i><b>3.2.3</b> Tesselate space</a></li>
<li class="chapter" data-level="3.2.4" data-path="practical-session-3.html"><a href="practical-session-3.html#polygonise-the-tessellation"><i class="fa fa-check"></i><b>3.2.4</b> Polygonise the tessellation</a></li>
<li class="chapter" data-level="3.2.5" data-path="practical-session-3.html"><a href="practical-session-3.html#identify-neighbours"><i class="fa fa-check"></i><b>3.2.5</b> Identify neighbours</a></li>
<li class="chapter" data-level="3.2.6" data-path="practical-session-3.html"><a href="practical-session-3.html#generate-distance-matrices"><i class="fa fa-check"></i><b>3.2.6</b> Generate distance matrices</a></li>
<li class="chapter" data-level="3.2.7" data-path="practical-session-3.html"><a href="practical-session-3.html#putting-it-all-together-1"><i class="fa fa-check"></i><b>3.2.7</b> Putting it all together</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="practical-session-4.html"><a href="practical-session-4.html"><i class="fa fa-check"></i><b>4</b> Practical session 4</a>
<ul>
<li class="chapter" data-level="4.1" data-path="practical-session-4.html"><a href="practical-session-4.html#geographically-weighted-principal-components-analysis-gwpca"><i class="fa fa-check"></i><b>4.1</b> Geographically Weighted Principal Components Analysis (GWPCA)</a></li>
<li class="chapter" data-level="4.2" data-path="practical-session-4.html"><a href="practical-session-4.html#load-packages"><i class="fa fa-check"></i><b>4.2</b> Load packages</a></li>
<li class="chapter" data-level="4.3" data-path="practical-session-4.html"><a href="practical-session-4.html#load-quality-controled-and-normalised-data"><i class="fa fa-check"></i><b>4.3</b> Load Quality Controled and Normalised data</a></li>
<li class="chapter" data-level="4.4" data-path="practical-session-4.html"><a href="practical-session-4.html#data-prearation-and-single-gwpca-run"><i class="fa fa-check"></i><b>4.4</b> Data prearation and single GWPCA run</a></li>
<li class="chapter" data-level="4.5" data-path="practical-session-4.html"><a href="practical-session-4.html#run-gwpca"><i class="fa fa-check"></i><b>4.5</b> Run GWPCA</a></li>
<li class="chapter" data-level="4.6" data-path="practical-session-4.html"><a href="practical-session-4.html#plot-global-pca-results"><i class="fa fa-check"></i><b>4.6</b> Plot global PCA results</a></li>
<li class="chapter" data-level="4.7" data-path="practical-session-4.html"><a href="practical-session-4.html#identify-the-leading-genes-in-each-location"><i class="fa fa-check"></i><b>4.7</b> Identify the leading genes in each location</a></li>
<li class="chapter" data-level="4.8" data-path="practical-session-4.html"><a href="practical-session-4.html#percentage-of-total-variation-ptv"><i class="fa fa-check"></i><b>4.8</b> Percentage of Total Variation (PTV)</a></li>
<li class="chapter" data-level="4.9" data-path="practical-session-4.html"><a href="practical-session-4.html#identify-discrepancies"><i class="fa fa-check"></i><b>4.9</b> Identify discrepancies</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Spatial transcriptomics data analysis: theory and practice</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="references" class="section level1 unnumbered hasAnchor">
<h1>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h1>

<div id="refs" class="references csl-bib-body hanging-indent">
<div class="csl-entry">
Amezquita, Robert A., Aaron T. L. Lun, Etienne Becht, Vince J. Carey, Lindsay N. Carpp, Ludwig Geistlinger, Federico Marini, et al. 2020. <span>“<span class="nocase">Orchestrating single-cell analysis with Bioconductor</span>.”</span> <em>Nat Methods</em> 17 (February): 137–45. <a href="https://doi.org/10.1038/s41592-019-0654-x">https://doi.org/10.1038/s41592-019-0654-x</a>.
</div>
<div class="csl-entry">
Avis, David, and Joe Horton. 1985. <span>“REMARKS ON THE SPHERE OF INFLUENCE GRAPH.”</span> <em>Annals of the New York Academy of Sciences</em> 440 (1): 323–27. https://doi.org/<a href="https://doi.org/10.1111/j.1749-6632.1985.tb14563.x">https://doi.org/10.1111/j.1749-6632.1985.tb14563.x</a>.
</div>
<div class="csl-entry">
Gollini, Isabella, Binbin Lu, Martin Charlton, Christopher Brunsdon, and Paul Harris. 2015. <span>“<span class="nocase">GWmodel: An R Package for Exploring Spatial Heterogeneity Using Geographically Weighted Models</span>.”</span> <em>J Stat Soft</em> 63 (February): 1–50. <a href="https://doi.org/10.18637/jss.v063.i17">https://doi.org/10.18637/jss.v063.i17</a>.
</div>
<div class="csl-entry">
Guilliams, Martin, Johnny Bonnardel, Birthe Haest, Bart Vanderborght, Camille Wagner, Anneleen Remmerie, Anna Bujko, et al. 2022. <span>“Spatial Proteogenomics Reveals Distinct and Evolutionarily Conserved Hepatic Macrophage Niches.”</span> <em>Cell</em> 185 (2): 379–396.e38. https://doi.org/<a href="https://doi.org/10.1016/j.cell.2021.12.018">https://doi.org/10.1016/j.cell.2021.12.018</a>.
</div>
<div class="csl-entry">
Hao, Yuhan, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck, Shiwei Zheng, Andrew Butler, Maddie J. Lee, et al. 2021. <span>“<span class="nocase">Integrated analysis of multimodal single-cell data</span>.”</span> <em>Cell</em> 184 (13): 3573–3587.e29. <a href="https://doi.org/10.1016/j.cell.2021.04.048">https://doi.org/10.1016/j.cell.2021.04.048</a>.
</div>
<div class="csl-entry">
Harris, Paul, Chris Brunsdon, and Martin Charlton. 2011. <span>“<span class="nocase">Geographically weighted principal components analysis</span>.”</span> <em>International Journal of Geographical Information Science</em> 25 (10): 1717–36. <a href="https://doi.org/10.1080/13658816.2011.554838">https://doi.org/10.1080/13658816.2011.554838</a>.
</div>
<div class="csl-entry">
Keogh, Eamonn, and Abdullah Mueen. 2017. <span>“<span class="nocase">Curse of Dimensionality</span>.”</span> In <em><span class="nocase">Encyclopedia of Machine Learning and Data Mining</span></em>, 314–15. Boston, MA, USA: Springer, Boston, MA. <a href="https://doi.org/10.1007/978-1-4899-7687-1_192">https://doi.org/10.1007/978-1-4899-7687-1_192</a>.
</div>
<div class="csl-entry">
Li, Yijun, Stefan Stanojevic, Bing He, Zheng Jing, Qianhui Huang, Jian Kang, and Lana X. Garmire. 2022. <span>“<span class="nocase">Benchmarking Computational Integration Methods for Spatial Transcriptomics Data</span>.”</span> <em>bioRxiv</em>, January, 2021.08.27.457741. <a href="https://doi.org/10.1101/2021.08.27.457741">https://doi.org/10.1101/2021.08.27.457741</a>.
</div>
<div class="csl-entry">
Lun, Aaron T. L., Davis J. McCarthy, and John C. Marioni. 2016. <span>“<span class="nocase">A step-by-step workflow for low-level analysis of single-cell RNA-seq data with Bioconductor</span>.”</span> <em>F1000Research</em> 5 (2122): 2122. <a href="https://doi.org/10.12688/f1000research.9501.2">https://doi.org/10.12688/f1000research.9501.2</a>.
</div>
<div class="csl-entry">
Matula, David W., and Robert R. Sokal. 1980. <span>“<span class="nocase">Properties of Gabriel Graphs Relevant to Geographic Variation Research and the Clustering of Points in the Plane</span>.”</span> <em>Geog Anal</em> 12 (3): 205–22. <a href="https://doi.org/10.1111/j.1538-4632.1980.tb00031.x">https://doi.org/10.1111/j.1538-4632.1980.tb00031.x</a>.
</div>
<div class="csl-entry">
Maynard, Kristen R., Leonardo Collado-Torres, Lukas M. Weber, Cedric Uytingco, Brianna K. Barry, Stephen R. Williams, Joseph L. Catallini, et al. 2021. <span>“<span class="nocase">Transcriptome-scale spatial gene expression in the human dorsolateral prefrontal cortex</span>.”</span> <em>Nat Neurosci</em> 24 (March): 425–36. <a href="https://doi.org/10.1038/s41593-020-00787-0">https://doi.org/10.1038/s41593-020-00787-0</a>.
</div>
<div class="csl-entry">
McCarthy, Davis J., Kieran R. Campbell, Aaron T. L. Lun, and Quin F. Wills. 2017. <span>“<span class="nocase">Scater: pre-processing, quality control, normalization and visualization of single-cell RNA-seq data in R</span>.”</span> <em>Bioinformatics</em> 33 (8): 1179–86. <a href="https://doi.org/10.1093/bioinformatics/btw777">https://doi.org/10.1093/bioinformatics/btw777</a>.
</div>
<div class="csl-entry">
McInnes, Leland, John Healy, and James Melville. 2018. <span>“<span class="nocase">UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction</span>.”</span> <em>arXiv</em>, February. <a href="https://doi.org/10.48550/arXiv.1802.03426">https://doi.org/10.48550/arXiv.1802.03426</a>.
</div>
<div class="csl-entry">
<span>“<span class="nocase">Notes on Continuous Stochastic Phenomena on JSTOR</span>.”</span> 1950. <em>Biometrika</em>. <a href="https://www.jstor.org/stable/2332142">https://www.jstor.org/stable/2332142</a>.
</div>
<div class="csl-entry">
Righelli, Dario, Lukas M. Weber, Helena L. Crowell, Brenda Pardo, Leonardo Collado-Torres, Shila Ghazanfar, Aaron T. L. Lun, Stephanie C. Hicks, and Davide Risso. 2022. <span>“<span class="nocase">SpatialExperiment: infrastructure for spatially-resolved transcriptomics data in R using Bioconductor</span>.”</span> <em>Bioinformatics</em> 38 (11): 3128–31. <a href="https://doi.org/10.1093/bioinformatics/btac299">https://doi.org/10.1093/bioinformatics/btac299</a>.
</div>
<div class="csl-entry">
Sun, Shiquan, Jiaqiang Zhu, and Xiang Zhou. 2020. <span>“<span class="nocase">Statistical analysis of spatial expression patterns for spatially resolved transcriptomic studies</span>.”</span> <em>Nat Methods</em> 17 (February): 193–200. <a href="https://doi.org/10.1038/s41592-019-0701-7">https://doi.org/10.1038/s41592-019-0701-7</a>.
</div>
<div class="csl-entry">
Svensson, Valentine, Sarah A. Teichmann, and Oliver Stegle. 2018. <span>“<span class="nocase">SpatialDE: identification of spatially variable genes</span>.”</span> <em>Nat Methods</em> 15 (May): 343–46. <a href="https://doi.org/10.1038/nmeth.4636">https://doi.org/10.1038/nmeth.4636</a>.
</div>
<div class="csl-entry">
<span>“<span class="nocase">The Contiguity Ratio and Statistical Mapping on JSTOR</span>.”</span> 1954. <em>Incorporated Statistician</em>. <a href="https://www.jstor.org/stable/2986645">https://www.jstor.org/stable/2986645</a>.
</div>
<div class="csl-entry">
Toussaint, Godfried T. 1980. <span>“<span class="nocase">The relative neighbourhood graph of a finite planar set</span>.”</span> <em>Pattern Recognit</em> 12 (4): 261–68. <a href="https://doi.org/10.1016/0031-3203(80)90066-7">https://doi.org/10.1016/0031-3203(80)90066-7</a>.
</div>
<div class="csl-entry">
Weber, Lukas M., and Helena L. Crowell. 2022. <em>Ggspavis: Visualization Functions for Spatially Resolved Transcriptomics Data</em>. <a href="https://github.com/lmweber/ggspavis">https://github.com/lmweber/ggspavis</a>.
</div>
<div class="csl-entry">
Zhu, Jiaqiang, Shiquan Sun, and Xiang Zhou. 2021. <span>“<span class="nocase">SPARK-X: non-parametric modeling enables scalable and robust detection of spatial expression patterns for large spatial transcriptomic studies</span>.”</span> <em>Genome Biol</em> 22 (1): 1–25. <a href="https://doi.org/10.1186/s13059-021-02404-0">https://doi.org/10.1186/s13059-021-02404-0</a>.
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="practical-session-4.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/05-references.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
